% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Whole-Genome Shotgun Sequencing for Pollen Identification},
  pdfauthor={Anya Cutler and Berry Brosi},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
\usepackage{booktabs}
\usepackage{longtable}
\usepackage{array}
\usepackage{multirow}
\usepackage{wrapfig}
\usepackage{float}
\usepackage{colortbl}
\usepackage{pdflscape}
\usepackage{tabu}
\usepackage{threeparttable}
\usepackage{threeparttablex}
\usepackage[normalem]{ulem}
\usepackage{makecell}
\usepackage{xcolor}

\title{Whole-Genome Shotgun Sequencing for Pollen Identification}
\author{Anya Cutler and Berry Brosi}
\date{started 16 August 2018; most recent edits 23 December 2020}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{4}
\tableofcontents
}
\textbf{update Dec 2020}

\begin{itemize}
\tightlist
\item
  this version of the Rmarkdown is updated from previous versions to
  reflect that we are now using merged forward and reverse reads. The
  Rmarkdown should mostly or entirely reflect that.
\end{itemize}

\hypertarget{overview}{%
\section{Overview}\label{overview}}

This analysis is set up to assess how whole-genome shotgun (WGS)
sequencing of known pollen samples performs, both on its own and
relative to amplicon-based metabarcoding (with ITS2 and \emph{rbcL}) in
terms of:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  \textbf{False negatives} (i.e.~true positives)---if a species is
  present in a sample, is that species detected?
\item
  \textbf{False positives}---how many species that are \emph{not}
  present in a sample are erroneously detected?
\item
  \textbf{Quantitative matching}---what is the quantitative
  correspondence between the number of pollen grains going in to a
  sample and the number of sequence reads coming out?
\end{enumerate}

These three questions form the basic structure of the analysis, which is
paralleled by the structure of this file.

The samples that were sequenced with WGS were from constructed ``mock
community'' pollen samples of known composition. We used the exact same
DNA isolations as in Bell et al.~2019 \emph{Molecular Ecology}. This
analysis is largely (though not entirely) based on the analyses in that
paper.

The sequenced pollen samples were constructed to vary in complexity in
three dimensions; we assessed how sample complexity affected the
qualitative outcomes (false positive and false negative reads):

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Question 1: \textbf{species richness} (1-9 species)
\item
  Question 2: \textbf{rarity} (actual proportion of grains this taxon
  has in a sample; from roughly half \& half to \textless{} 1\% of the
  rarer type)
\item
  Question 3: \textbf{taxonomic relatedness} (within genus to across
  broad clades in the seed plants)
\end{enumerate}

\hypertarget{analysis-approach}{%
\section{Analysis approach}\label{analysis-approach}}

\hypertarget{non-independence-of-data}{%
\subsection{non-independence of data}\label{non-independence-of-data}}

Because all analyses included non-independent data (multiple replicates
of the same pollen mixtures; pollen from the same plant species
occurring in multiple mixtures), all of our analyses were conducted with
mixed-effects modeling, using mixture identity and species identity as
crossed random effects (modeled as random intercepts). For the analyses
related to false positive matching, all species truly present in a mix
were pooled together for the ``true positives'', so we did not use
species identity (up to 9 different species) as a random effect in those
analyses.

\hypertarget{comparing-wgs-to-amplicon-data}{%
\subsection{comparing WGS to amplicon
data}\label{comparing-wgs-to-amplicon-data}}

Across our outcomes, in comparing WGS and amplicon performance, we
pooled WGS and amplicon results together into a single data table and
conducted analyses with method (WGS vs.~amplicon) as a fixed effect.

\hypertarget{structure-of-analyses-false-negatives-and-false-positives}{%
\subsection{structure of analyses: false negatives and false
positives}\label{structure-of-analyses-false-negatives-and-false-positives}}

The response variables for our first two outcomes are binomial (yes /
no) in structure. We use binomial-errors mixed-effects models for these
analyses.

For our first outcome of false negatives, we thus need to record---for
each species present in a pollen mixture---whether or not that species
was detected in that sample. To do this, we set up a datafile with each
species truly present within each sample as its own row, which we
subsequently scored as 0 / 1, with a zero for species that were present
in the sample but \emph{not} present in sequencing reads (above the
contamination threshold), and a one for species that \emph{were} present
in the sequencing reads above the threshold.

For our second outcome of false positives, we wished to assess the
proportion of true vs.~false positives. To do this, we aggregated the
data to one row per sample replicate, and summed the read counts of true
positives (combining all species truly present in a particular mix) and
false positives in two separate columns.

\hypertarget{structure-of-analyses-quantitative-matching}{%
\subsection{structure of analyses: quantitative
matching}\label{structure-of-analyses-quantitative-matching}}

To assess the quantitative accuracy of WGS sequencing for our
constructed mixtures, we tested the correlation between the known
proportion of pollen grains in a sample (explanatory variable) and the
proportion of WGS sequencing reads (response variable).

Ideally, we would like for there to be a perfect 1:1 fit between pollen
grains going in and sequence reads coming out. This means a linear
response between the two, with an intercept of 0 and a slope of 1. One
could make an argument that instead of using a linear model, a
binomial-errors model would be more appropriate given that the response
variable is indeed proportional counts (of sequence reads). Moreover,
even with a binomial-errors model, it is possible to get the equivalent
of a line with slope = 1 and intercept = 0.

We considered this issue in detail and decided that a linear model is
better in this context. In particular, a binomial model could for
example predict that for a given taxon, it will have a very very low
detection probability until it reaches say \textasciitilde80\% of the
pollen grains in the sample at which point the logistic curve could
swing up sharply and after \textasciitilde80\% the detection probability
could be very close to 1. If the data matched such a model, it would
return both a very low \(p\)-value and a very high \(R^2\), even though
it would not be giving us what we want. Therefore, we decided to stick
with a linear model.

An alternative possibility is to specify the intercept (0) and slope (1)
and see how well the data do at fitting that line. We also pursued this
approach, by examining the \emph{residuals} of WGS vs.~amplicon data to
that regression line. Smaller residuals are better. One thing to note
about this approach is that it is probably dominated by the proportion
of false positive reads---more false positives will mean that the
proportion of true matches will be further from the 1:1 ratio. We thus
used both of the approaches above to assess quantitative matching.

\hypertarget{data-import}{%
\section{Data import}\label{data-import}}

Three groups of data sets to import:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  sample metadata
\item
  WGS Kraken empirical data
\item
  amplicon data
\end{enumerate}

The \textbf{metadata} are the same for the WGS and amplicon data, as
both were run from the exact same DNA extractions.

For the \textbf{WGS Kraken data}, Jamieson Botsch formatted these
(25-Apr-2018) from the raw Kraken output to combine the two data sources
(we sequenced WGS data both at Emory and at UGA), and to have the family
/ genus / species separated out into columns; see
\texttt{Shotgun\_data\_prep.Rmd}. Unlike the QIIME Illumina data, we do
not need to do any aggregation of read counts as this was done
automatically by Kraken (nice touch); i.e.~all read counts at the
species level are also included in genus-level matches and so on. We do
just a couple of steps of very basic / simple formatting with the data
import here (removal of superfluous numeric column a the beginning of
the data; removal of duplicate rows).

\begin{itemize}
\tightlist
\item
  in addition, at some point in 2020 Robert Petit merged the forward and
  reverse reads from the shotgun data in response to reviews, so that is
  a fairly substantial change from the previous version. Analyses in
  late 2020 reflect his update.
\end{itemize}

The \textbf{amplicon data} are represented by 7 data files: The first
(\texttt{amp\_all}) is already merged with the sample metadata and
already formatted for false-negative analyses (with a separate row for
each species within each sample). There are separate columns for the
presence or absence of that taxon at three different taxonomic levels
(species, genus, and family) and for each marker (ITS2 and \emph{rbcL}).

This \texttt{amp\_all} dataset, however, does not include data on false
positives. Thus, the remaining six datafiles are there to provide data
on how many of the hits were to true positives (known to be present in
the sample) vs.~false positives (not included in the sample). In
contrast to \texttt{amp\_all}, these are formatted as six separate
files, one each for the two markers (ITS2 and \emph{rbcL}) and three
levels of taxonomic matching (species, genus, and family). The format of
these files matches basically what comes from QIIME: taxa are the rows,
and samples are the columns. We re-format these to the tidy format of
having samples as rows later.

Across both the amplicon and the kraken data, one issue is that for
different samples, matching the taxonomy at different levels leads to
different numbers of rows. For example, we have two \emph{Populus}
species in some of the samples. For matching at the species level, that
is two different rows. But for genus and family, it is just one row.
There are some other samples that have two different genera in the same
family (e.g., \emph{Poa} and \emph{Zea}, both in the Poaceae). Thus,
these have to be either 1) stored as completely separate datafiles; or
2) stored in a single datafile (as in \texttt{amp\_all}) but properly
subset to not include duplicate rows for the same genus or family within
samples. In this analysis we have split these out into separate files,
which may not have been the best way to go (certainly not for memory
management / speed; and also it makes the global environment rather
messy). But given the aggregation that we had to do, it was probably the
most straightforward way to go.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{mixes =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"pollen-mixes-proportions.csv"}\NormalTok{)}
\NormalTok{amp_its_family =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_ITS_Family.csv"}\NormalTok{)}
\NormalTok{amp_its_genus =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_ITS_Genus.csv"}\NormalTok{)}
\NormalTok{amp_its_species =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_ITS_Species.csv"}\NormalTok{)}
\NormalTok{amp_rbc_family =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_rbcL_Family.csv"}\NormalTok{)}
\NormalTok{amp_rbc_genus =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_rbcL_Genus.csv"}\NormalTok{)}
\NormalTok{amp_rbc_species =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_rbcL_Species.csv"}\NormalTok{)}
\NormalTok{amp_all =}\StringTok{ }\KeywordTok{read.csv}\NormalTok{(}\StringTok{"Amplicon_all.csv"}\NormalTok{)}

\CommentTok{# 'mixes' comes in with some rows duplicated (because in the spreadsheet, each was assessed with both ITS2 and with rbcL); fix this here:}
\NormalTok{mixes =}\StringTok{ }\KeywordTok{unique}\NormalTok{(mixes)}

\CommentTok{#Create list of kraken2 report files}
\NormalTok{reports =}\StringTok{ }\KeywordTok{list.files}\NormalTok{(}\DataTypeTok{path=}\StringTok{"./kraken2_results"}\NormalTok{, }\DataTypeTok{full.names=}\NormalTok{T)}

\CommentTok{#Read in files as table}
\NormalTok{kraken2=}\KeywordTok{lapply}\NormalTok{(reports, read.table, }\DataTypeTok{header=}\OtherTok{FALSE}\NormalTok{, }\DataTypeTok{sep=}\StringTok{"}\CharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}

\CommentTok{#Create column with file name to get sampleID and bind it to other columns}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{length}\NormalTok{(kraken2))\{kraken2[[i]]<-}\KeywordTok{cbind}\NormalTok{(reports[i],kraken2[[i]])\}}
\NormalTok{kraken2_rbind <-}\StringTok{ }\KeywordTok{do.call}\NormalTok{(}\StringTok{"rbind"}\NormalTok{, kraken2) }

\CommentTok{#Rename column names}
\KeywordTok{colnames}\NormalTok{(kraken2_rbind)[}\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{,}\DecValTok{7}\NormalTok{)]<-}\KeywordTok{c}\NormalTok{(}\StringTok{"sample.id"}\NormalTok{, }\StringTok{"perc.hit"}\NormalTok{, }\StringTok{"hits"}\NormalTok{,}\StringTok{"misses"}\NormalTok{,}\StringTok{"tax"}\NormalTok{,}\StringTok{"tax.id"}\NormalTok{,}\StringTok{"id"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{data-formatting-setup}{%
\section{data formatting / setup}\label{data-formatting-setup}}

\textbf{Overview}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Basic Kraken data prep

  \begin{itemize}
  \tightlist
  \item
    filter to include only matches at family / genus / species taxonomic
    resolutions (not anything coarser, and no intermediate clades)
  \item
    create new columns for \texttt{mix.ID} that matches the column in
    the sample (\texttt{pollen-mixes-proportions.csv}) data, and also
    \texttt{replicate.ID} for replicates within each mix
  \item
    a couple of minor tweaks to make sure data are consistent etc.
  \end{itemize}
\item
  Combine Kraken and sample metadata
\item
  Remove reads below contamination threshold
\item
  Format Kraken data for false negative and quantitative matching
  analyses

  \begin{itemize}
  \tightlist
  \item
    both of these analyses are about true positives (=false negatives)
    and straightforward to do the data formatting in the same step
  \end{itemize}
\item
  Combine amplicon and empirical Kraken data: false negatives
\item
  Format Kraken data for false positive analysis
\item
  Combine amplicon and empirical Kraken data: false positives
\end{enumerate}

\hypertarget{basic-kraken-data-prep}{%
\subsection{Basic Kraken data prep}\label{basic-kraken-data-prep}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  create new column for \texttt{mix.ID} that matches the column in the
  sample metadata (\texttt{pollen-mixes-proportions.csv})\\
\item
  also create a new \texttt{replicate.ID} column (replicate samples of
  each \texttt{mix.ID})
\item
  filter by \texttt{tax} column, including only species / genus / family
  levels
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Create mixID column}
\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID =}\StringTok{ }\KeywordTok{substr}\NormalTok{(kraken2_rbind}\OperatorTok{$}\NormalTok{sample.id, }\DecValTok{19}\NormalTok{, }\DecValTok{24}\NormalTok{)}

\CommentTok{# Want to remove the 6th character if mix.id does not contain the word "mix" or contains 2 dashes. Only want six characters if the mix number is double digits.}
\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\OperatorTok{!}\KeywordTok{grepl}\NormalTok{(}\StringTok{"mix"}\NormalTok{, kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID) }\OperatorTok{|}\StringTok{ }\KeywordTok{str_count}\NormalTok{(kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID, }\DataTypeTok{pattern=}\StringTok{"-"}\NormalTok{)}\OperatorTok{==}\DecValTok{2}\NormalTok{, }\KeywordTok{substr}\NormalTok{(kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID, }\DecValTok{1}\NormalTok{, }\DecValTok{5}\NormalTok{), kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID)}

\CommentTok{#replace dashes in 'mix.ID' with periods so that they match the 'mixes' data and remove underscores}
\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID =}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\StringTok{"-"}\NormalTok{, }\StringTok{"."}\NormalTok{, kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID)}

\CommentTok{#remove periods if mix.ID contains "mix" so it matches the mix data}
\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID =}\StringTok{ }\KeywordTok{gsub}\NormalTok{(}\StringTok{"mix."}\NormalTok{,}\StringTok{"mix"}\NormalTok{, kraken2_rbind}\OperatorTok{$}\NormalTok{mix.ID)}

\CommentTok{# move the new column to be second in order (not at the end where it's hard to see)}
\NormalTok{kraken2_rbind =}\StringTok{ }\NormalTok{kraken2_rbind[, }\KeywordTok{c}\NormalTok{(}\DecValTok{1}\NormalTok{, }\KeywordTok{ncol}\NormalTok{(kraken2_rbind), }\DecValTok{3}\OperatorTok{:}\KeywordTok{ncol}\NormalTok{(kraken2_rbind)}\OperatorTok{-}\DecValTok{1}\NormalTok{)]}

\CommentTok{#Remove any rows that are identified to anything other than family, genus, or species (not higher tax or subgroups)}
\NormalTok{kraken2_rbind =}\StringTok{ }\NormalTok{kraken2_rbind[(kraken2_rbind}\OperatorTok{$}\NormalTok{tax }\OperatorTok{==}\StringTok{ "F"} \OperatorTok{|}\StringTok{ }\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{tax }\OperatorTok{==}\StringTok{ "G"} \OperatorTok{|}\StringTok{ }\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{tax }\OperatorTok{==}\StringTok{ "S"}\NormalTok{),]}

\CommentTok{#Remove leading white space in tax names}
\NormalTok{kraken2_rbind}\OperatorTok{$}\NormalTok{id =}\StringTok{ }\KeywordTok{trimws}\NormalTok{(kraken2_rbind}\OperatorTok{$}\NormalTok{id, }\StringTok{"left"}\NormalTok{)}

\CommentTok{#Rename dataframe so it can run through script}
\NormalTok{krak =}\StringTok{ }\NormalTok{kraken2_rbind}

\CommentTok{#Create list of unique combinations of mix ID and rep ID to properly merge kraken with mixes data at later point}
\NormalTok{krak_uniqueid <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(krak}\OperatorTok{$}\NormalTok{mix.ID)}
\NormalTok{krak_uniqueid <-}\StringTok{ }\KeywordTok{unique.data.frame}\NormalTok{(krak_uniqueid)}
\end{Highlighting}
\end{Shaded}

\hypertarget{combine-kraken-sample-metadata}{%
\subsection{combine Kraken \& sample
metadata}\label{combine-kraken-sample-metadata}}

Create a new aggregated datasheet, based on sample metadata, but which
matches sample data back to the Kraken data so we can run analyses about
probability of matching (both qualitative and quantitative). In
particular this is taking account of sample ``replicates'' (originally
including some that are not true replicates, but rather forward
vs.~reverse reads---this is not the case in the later 2020 data); also
different Illumina lanes in the Emory Genome Center data.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Merge the unique krak IDs with mixes to subset the sample data to only
  the samples in the kraken dataset
\item
  Create a dataset for analysis of false negatives by only including
  taxa that are present in sample data

  \begin{itemize}
  \tightlist
  \item
    for this one, do it separately at each taxonomic level that we are
    assessing, i.e.~species, genus, family
  \end{itemize}
\item
  Create a dataset for analysis of false positives by including all taxa
  in kraken data
\end{enumerate}

use `mix.ID' as the variable to combine by\ldots{}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## Merge sample and replicate IDs with mixes data}
\NormalTok{krak_mixes <-}\StringTok{ }\KeywordTok{merge}\NormalTok{(mixes,krak_uniqueid, }\DataTypeTok{by.x=}\StringTok{"mix.ID"}\NormalTok{, }\DataTypeTok{by.y=}\StringTok{"krak.mix.ID"}\NormalTok{)}

\CommentTok{# create separate datasets for family, genus, and species level, both including and excluding false positives}
\CommentTok{# excluding false positives (focus on true positives / false negatives):}
\CommentTok{#truepos.krak.family =  merge(krak_mixes, filter(krak, tax == "F"), by.x = c("mix.ID", "family"), by.y = c("mix.ID", "id"), all.x=T)}
\NormalTok{truepos.krak.genus =}\StringTok{  }\KeywordTok{merge}\NormalTok{(krak_mixes, }\KeywordTok{filter}\NormalTok{(krak, tax }\OperatorTok{==}\StringTok{ "G"}\NormalTok{), }\DataTypeTok{by.x =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{), }\DataTypeTok{by.y =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"id"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}
\NormalTok{truepos.krak.species =}\StringTok{  }\KeywordTok{merge}\NormalTok{(krak_mixes, }\KeywordTok{filter}\NormalTok{(krak, tax }\OperatorTok{==}\StringTok{ "S"}\NormalTok{), }\DataTypeTok{by.x =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{), }\DataTypeTok{by.y =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"id"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}

\CommentTok{# including false positives:}
\CommentTok{#all.krak.family =  merge(krak_mixes, filter(krak, tax == "F"), by.x = c("mix.ID", "family"), by.y = c("mix.ID", "id"), all.x = T, all.y = T)}
\NormalTok{all.krak.genus =}\StringTok{  }\KeywordTok{merge}\NormalTok{(krak_mixes, }\KeywordTok{filter}\NormalTok{(krak, tax }\OperatorTok{==}\StringTok{ "G"}\NormalTok{), }\DataTypeTok{by.x =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{), }\DataTypeTok{by.y =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"id"}\NormalTok{), }\DataTypeTok{all.x =}\NormalTok{ T, }\DataTypeTok{all.y =}\NormalTok{ T)}
\NormalTok{all.krak.species =}\StringTok{  }\KeywordTok{merge}\NormalTok{(krak_mixes, }\KeywordTok{filter}\NormalTok{(krak, tax }\OperatorTok{==}\StringTok{ "S"}\NormalTok{), }\DataTypeTok{by.x =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{), }\DataTypeTok{by.y =} \KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"id"}\NormalTok{), }\DataTypeTok{all.x =}\NormalTok{ T, }\DataTypeTok{all.y =}\NormalTok{ T)}
\end{Highlighting}
\end{Shaded}

\hypertarget{remove-kraken-reads-below-the-contamination-threshold}{%
\subsection{remove Kraken reads below the contamination
threshold}\label{remove-kraken-reads-below-the-contamination-threshold}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  remove reads below the isolation / PCR negative control thresholds

  \begin{enumerate}
  \def\labelenumii{\alph{enumii})}
  \tightlist
  \item
    ideally, would base these on negative controls \emph{for a given
    Illumina run} (two separate runs here, at UGA and Emory Genome
    center; we will need to keep those separate); but unfortunately
    there are not negative controls for the Emory Genome Center data;
    use the UGA thresholds for all the data (definitely not ideal, but
    working with what we have)
  \item
    this entails removing rows with \(k\)-mer counts below the threshold
  \item
    in the Illumina QIIME amplicon data, we did this separately for
    (entire) taxa below the threshold level, and for reads, because of
    the way that the data were formatted in a taxon-by-sample matrix;
    the Kraken output is already formatted in a way that is closer to
    \texttt{tidy} so we only need to do this step once here.
  \end{enumerate}
\item
  remove negative control rows (once the above is completed) so they
  don't interfere with the analysis
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Establish threshold for maximum number of reads in the negative controls}
\NormalTok{maxy =}\StringTok{ }\KeywordTok{max}\NormalTok{(krak}\OperatorTok{$}\NormalTok{hits[krak}\OperatorTok{$}\NormalTok{mix.ID}\OperatorTok{==}\StringTok{"negat"}\NormalTok{])}

\CommentTok{# Identify rows that fall below this threshold for false negative analysis}
\CommentTok{#truepos.indexy_family = which(truepos.krak.family$hits <= maxy)}
\NormalTok{truepos.indexy_genus =}\StringTok{ }\KeywordTok{which}\NormalTok{(truepos.krak.genus}\OperatorTok{$}\NormalTok{hits }\OperatorTok{<=}\StringTok{ }\NormalTok{maxy)}
\NormalTok{truepos.indexy_species =}\StringTok{ }\KeywordTok{which}\NormalTok{(truepos.krak.species}\OperatorTok{$}\NormalTok{hits }\OperatorTok{<=}\StringTok{ }\NormalTok{maxy)}

\CommentTok{# No rows for false negative analysis with read level below threshold - can continue with analysis }

\CommentTok{# Identify rows that fall below thresshold for false positive analysis}
\CommentTok{#all.indexy_family = which(all.krak.family$hits <= maxy)}
\NormalTok{all.indexy_genus =}\StringTok{ }\KeywordTok{which}\NormalTok{(all.krak.genus}\OperatorTok{$}\NormalTok{hits }\OperatorTok{<=}\StringTok{ }\NormalTok{maxy)}
\NormalTok{all.indexy_species =}\StringTok{ }\KeywordTok{which}\NormalTok{(all.krak.species}\OperatorTok{$}\NormalTok{hits }\OperatorTok{<=}\StringTok{ }\NormalTok{maxy)}

\CommentTok{# Remove these rows}
\CommentTok{#all.krak.family = all.krak.family[-all.indexy_family,]}
\NormalTok{all.krak.genus =}\StringTok{ }\NormalTok{all.krak.genus[}\OperatorTok{-}\NormalTok{all.indexy_genus,]}
\NormalTok{all.krak.species =}\StringTok{ }\NormalTok{all.krak.species[}\OperatorTok{-}\NormalTok{all.indexy_species,]}

\CommentTok{# Remove negative control rows}
\CommentTok{#all.krak.family = filter(all.krak.family, mix.ID!="negat")}
\NormalTok{all.krak.genus =}\StringTok{ }\KeywordTok{filter}\NormalTok{(all.krak.genus, mix.ID}\OperatorTok{!=}\StringTok{"negat"}\NormalTok{)}
\NormalTok{all.krak.species =}\StringTok{ }\KeywordTok{filter}\NormalTok{(all.krak.species, mix.ID}\OperatorTok{!=}\StringTok{"negat"}\NormalTok{)}

\CommentTok{# Clean up}
\KeywordTok{rm}\NormalTok{(maxy, all.indexy_family, all.indexy_genus, all.indexy_species, truepos.indexy_family, truepos.indexy_genus, truepos.indexy_species)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in rm(maxy, all.indexy_family, all.indexy_genus, all.indexy_species, :
## object 'all.indexy_family' not found
\end{verbatim}

\begin{verbatim}
## Warning in rm(maxy, all.indexy_family, all.indexy_genus, all.indexy_species, :
## object 'truepos.indexy_family' not found
\end{verbatim}

\hypertarget{kraken-data-format-for-false-negative-and-quantitative-analyses}{%
\subsection{Kraken data: format for false negative and quantitative
analyses}\label{kraken-data-format-for-false-negative-and-quantitative-analyses}}

(will do false positives in next step; the data formatting at this step
is relatively straightforward; again both of these are focused on true
positives so easy to do them in the same code chunk)

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Convert NAs in true positives / false negatives to 0s
\item
  Convert the ``percentage hits'' to a proportion - which will be the
  response variable in the quantitative models
\item
  Create a qualitative variable that equals 1 if that taxon was detected
  in the analysis (i.e.~if proportion is greater than 0), and 0 if not
  detected (i.e.~if proportion = 0).
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# create quantitative variable by simply dividing the percentage hits (perc.hit) by 100}
\CommentTok{#truepos.krak.family$quant.family = truepos.krak.family$perc.hit/100}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{quant.genus =}\StringTok{ }\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{perc.hit}\OperatorTok{/}\DecValTok{100}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{quant.species =}\StringTok{ }\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{perc.hit}\OperatorTok{/}\DecValTok{100}

\CommentTok{# for taxa that were not detected, NAs are currently present. Need to change this to 0 for quantitative variable before running analysis.}
\CommentTok{#truepos.krak.family$quant.family = ifelse(is.na(truepos.krak.family$quant.family), 0, truepos.krak.family$quant.family)}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{quant.genus =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(truepos.krak.genus}\OperatorTok{$}\NormalTok{quant.genus), }\DecValTok{0}\NormalTok{, truepos.krak.genus}\OperatorTok{$}\NormalTok{quant.genus)}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{quant.species =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(truepos.krak.species}\OperatorTok{$}\NormalTok{quant.species), }\DecValTok{0}\NormalTok{, truepos.krak.species}\OperatorTok{$}\NormalTok{quant.species)}

\CommentTok{#Create qualitative variable based on quantitative variable}
\CommentTok{#truepos.krak.family$qual.family = ifelse(truepos.krak.family$quant.family > 0, 1, 0)}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{qual.genus =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(truepos.krak.genus}\OperatorTok{$}\NormalTok{quant.genus }\OperatorTok{>}\StringTok{ }\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{)}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{qual.species =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(truepos.krak.species}\OperatorTok{$}\NormalTok{quant.species }\OperatorTok{>}\StringTok{ }\DecValTok{0}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{combine-amplicon-and-kraken-data-for-false-negative-analysis}{%
\subsection{combine amplicon and Kraken data for false negative
analysis}\label{combine-amplicon-and-kraken-data-for-false-negative-analysis}}

Using the kraken data formatted for true positive/false negative
analysis (``truepos.krak''), rbind with amplicon data.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Add "K-" to beginning of sample number of kraken dataset and add variable "source" to indicate that data is from kraken dataset}
\CommentTok{#truepos.krak.family$sample <- paste("K-", truepos.krak.family$krak.sample, sep="")}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"K-"}\NormalTok{, truepos.krak.genus}\OperatorTok{$}\NormalTok{krak.sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"K-"}\NormalTok{, truepos.krak.species}\OperatorTok{$}\NormalTok{krak.sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}

\CommentTok{#Add "source" variable to indicate if data is WGS or amplicon}
\CommentTok{#truepos.krak.family$source = "krak"}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{source =}\StringTok{ "krak"}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{source =}\StringTok{ "krak"}
\NormalTok{amp_all}\OperatorTok{$}\NormalTok{source =}\StringTok{ "amp"}

\CommentTok{#Concatenate WGS and amplicon data using mapply so that mismatching variable names of the amplicon data will be ignored}
\CommentTok{# 'fn' appended at the end of the variable names signifies "false negatives"}

\CommentTok{#krakamp_rbc_family_fn <- as.data.frame(mapply(c, truepos.krak.family[,c("mix.ID", "family", "sample", "qual.family", "source")], amp_all[,c("mix.ID","family","sample", "qual.family.rbcL", "source")]))}

\CommentTok{#krakamp_its_family_fn <- as.data.frame(mapply(c, truepos.krak.family[,c("mix.ID", "family", "sample", "qual.family", "source")], amp_all[,c("mix.ID","family","sample", "qual.family.ITS", "source")]))}

\NormalTok{krakamp_rbc_genus_fn <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{mapply}\NormalTok{(c, truepos.krak.genus[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"genus"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus.rbcL"}\NormalTok{, }\StringTok{"source"}\NormalTok{)]))}

\NormalTok{krakamp_its_genus_fn <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{mapply}\NormalTok{(c, truepos.krak.genus[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"genus"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus.ITS"}\NormalTok{, }\StringTok{"source"}\NormalTok{)]))}

\NormalTok{krakamp_rbc_species_fn <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{mapply}\NormalTok{(c, truepos.krak.species[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"species"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species.rbcL"}\NormalTok{, }\StringTok{"source"}\NormalTok{)]))}

\NormalTok{krakamp_its_species_fn <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(}\KeywordTok{mapply}\NormalTok{(c, truepos.krak.species[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"species"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species.ITS"}\NormalTok{, }\StringTok{"source"}\NormalTok{)]))}
\end{Highlighting}
\end{Shaded}

\hypertarget{format-kraken-data-for-false-positive-analysis}{%
\subsection{format kraken data for false positive
analysis}\label{format-kraken-data-for-false-positive-analysis}}

To analyze false positives in kraken data, need to have aggregated
counts of the ``true positive'' and ``false positive'' reads by sample.
These will be our ``success'' and ``failure'' numbers in a binomial
mixed model.

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  If the merged sample data is N/A, this indicates the taxa is a false
  positive. Based on N/A values, create a variable indicating if taxa is
  true or false positive.
\item
  Remove taxa that are in sample data but not kraken data (false
  negatives) - not considered for this analysis

  \begin{itemize}
  \tightlist
  \item
    while perhaps counterintuitive, need to do this because in this
    analysis we are using counts of ``true positive'' reads vs.~``false
    positive'' reads, i.e.~only using reads that exist, not reads that
    should have existed but do not. Those are dealt with in our false
    negative analysis.
  \end{itemize}
\item
  Add ``K'' to the sample names to distinguish from the samples in the
  amplicon data
\item
  Aggregate the counts per sample ID by ``true positive'' and ``false
  positive''
\end{enumerate}

In addition :

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  removed \emph{Zea mays} and mix5 samples (don't match with metadata;
  mix5 has \emph{Zea} in it)
\item
  matched false positive data with metadata so that sample complexity
  analyses can be run down the line
\item
  one consideration is that since we are aggregating all of the reads
  for a particular sample into one row, we have to figure out how to
  deal with rarity / commonness of taxa within a sample.

  \begin{itemize}
  \tightlist
  \item
    to do that, we use the \emph{minimum} \% of any taxon within the
    sample as a measure of rarity
  \end{itemize}
\end{enumerate}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# For false positive analysis, create variable indicating if taxa if "false positive" or "true positive"}
\CommentTok{#all.krak.family$type <- ifelse(is.na(all.krak.family$question.1), "false_pos", "true_pos")}
\NormalTok{all.krak.genus}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(all.krak.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}
\NormalTok{all.krak.species}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(all.krak.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}

\CommentTok{# If hits = N/A, indicates a false negative (in sample data but not in Kraken data). Remove because we are not considering for this analysis}
\CommentTok{# (for the future, probably easier with `filter` in `dplyr` package)}
\CommentTok{#all.krak.family <- all.krak.family[-which(is.na(all.krak.family$sample.id)),]}
\CommentTok{#all.krak.genus <- all.krak.genus[-which(is.na(all.krak.genus$sample.id)),]}
\CommentTok{#all.krak.species <- all.krak.species[-which(is.na(all.krak.species$sample.id)),]}

\CommentTok{#ABOVE COMMENTED OUT BECAUSE THERE WERE NO FALSE NEGATIVES AND CODE WAS CAUSING ISSUES }

\CommentTok{# Add "K_" to sample name}
\CommentTok{#all.krak.family$sample <- paste("K_",all.krak.family$sample, sep="")}
\NormalTok{all.krak.genus}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"K_"}\NormalTok{,all.krak.genus}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}
\NormalTok{all.krak.species}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"K_"}\NormalTok{,all.krak.species}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}

\CommentTok{# Aggregate counts by mix.ID, sample.ID, and type}
\CommentTok{#agg.krak.family <- all.krak.family %>%}
          \CommentTok{#select(mix.ID, family, sample, type, hits) %>%}
          \CommentTok{#group_by(mix.ID, sample, rep.ID, type) %>%}
          \CommentTok{#summarize(total_hits = sum(hits))}
\NormalTok{agg.krak.genus <-}\StringTok{ }\NormalTok{all.krak.genus }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{select}\NormalTok{(mix.ID, genus, sample, type, hits) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{group_by}\NormalTok{(mix.ID, sample, type) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total_hits =} \KeywordTok{sum}\NormalTok{(hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{agg.krak.species <-}\StringTok{ }\NormalTok{all.krak.species }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{select}\NormalTok{(mix.ID, species, sample, type, hits) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{group_by}\NormalTok{(mix.ID, sample, type) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total_hits =} \KeywordTok{sum}\NormalTok{(hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Convert to wide format with one column for true positive hits and one column for false positive hits}
\CommentTok{#agg.krak.family <- spread(agg.krak.family, key=type, value=total_hits)}
\NormalTok{agg.krak.genus <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(agg.krak.genus, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}
\NormalTok{agg.krak.species <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(agg.krak.species, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}

\CommentTok{# If positives are N/A, set to equal 0}
\CommentTok{#agg.krak.family$true_pos = ifelse(is.na(agg.krak.family$true_pos),0, agg.krak.family$true_pos)}
\NormalTok{agg.krak.genus}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(agg.krak.genus}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, agg.krak.genus}\OperatorTok{$}\NormalTok{true_pos)}
\NormalTok{agg.krak.species}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(agg.krak.species}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, agg.krak.species}\OperatorTok{$}\NormalTok{true_pos)}
\CommentTok{#agg.krak.family$false_pos = ifelse(is.na(agg.krak.family$false_pos),0, agg.krak.family$false_pos)}
\NormalTok{agg.krak.genus}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(agg.krak.genus}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, agg.krak.genus}\OperatorTok{$}\NormalTok{false_pos)}
\NormalTok{agg.krak.species}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(agg.krak.species}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, agg.krak.species}\OperatorTok{$}\NormalTok{false_pos)}

\CommentTok{# as set up currently, the false positive datasets ('agg.krak.family', 'agg.krak.genus', 'agg.krak.species')}
\CommentTok{# do not have the needed components to replicate the analysis we used for false negatives}
\CommentTok{# specifically, we need to have data on the facets of samples complexity:}
    \CommentTok{# species richness, taxonomic relatedness, and rarity}
\CommentTok{# and we also need the data subset component ('sub' vs. 'all')}

\CommentTok{# first, remove Zea mays and mix5 (which had Zea in it):}
\CommentTok{#agg.krak.family = filter(agg.krak.family, mix.ID != "Z.may" & mix.ID != "mix5")}
\CommentTok{#agg.krak.genus = filter(agg.krak.genus, mix.ID != "Z.may" & mix.ID != "mix5")}
\CommentTok{#agg.krak.species = filter(agg.krak.species, mix.ID != "Z.may" & mix.ID != "mix5")}

\CommentTok{# now, add the needed metadata}

\CommentTok{# don't want the aggregated data to be repeated for each taxon in a mix}
\CommentTok{# and that is the way the 'mixes' dataframe is set up}
\CommentTok{# in addition, to assess rarity, different taxa have different proportions}
\CommentTok{# thus, summarize 'mixes' to have the *minimum* value of 'pollen.grain.proportion' for each mix}

\NormalTok{mixy =}\StringTok{ }\NormalTok{mixes }
\NormalTok{mixy =}\StringTok{ }\NormalTok{mixy }\OperatorTok{%>%}\StringTok{ }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{family, }\OperatorTok{-}\NormalTok{genus, }\OperatorTok{-}\NormalTok{species) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(mix.ID) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize_all}\NormalTok{(min)}

\CommentTok{# now put them together}
\CommentTok{# via a merge of the false positive datasets with the 'mixes' (sample metadata)}
\CommentTok{# (didn't use 'left_join' because of conversion of factors to character)}
\CommentTok{#agg.krak.family = merge(agg.krak.family, mixy, by = "mix.ID")}
\NormalTok{agg.krak.genus =}\StringTok{ }\KeywordTok{merge}\NormalTok{(agg.krak.genus, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}
\NormalTok{agg.krak.species =}\StringTok{ }\KeywordTok{merge}\NormalTok{(agg.krak.species, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}

\CommentTok{# clean up... or don't}
\CommentTok{# don't remove 'mixy' just yet as we'll employ it again below for the false-positive amplicon data}
\end{Highlighting}
\end{Shaded}

\hypertarget{false-positives-format-amplicon-data}{%
\subsection{false positives: format amplicon
data}\label{false-positives-format-amplicon-data}}

This part is one of the more challenging components of the formatting.
It is based on the six amplicon data files (one for each marker and
taxonomic level), formatted as QIIME matrices such that taxa comprise
the rows and samples comprise the columns.

we need to:

\begin{itemize}
\tightlist
\item
  \texttt{melt} each matrix into a single row for each taxon / mix /
  sample combination, reporting the number of reads for each

  \begin{itemize}
  \tightlist
  \item
    remove combos for which there were zero reads (not strictly
    necessary since we will later just add up all the true and false
    positives, but makes everything more compact)
  \item
    also a good time to remove mix \#5 and also \emph{Zea} samples from
    the amplicon data for consistencey
  \end{itemize}
\item
  merge each melted dataframe with with the sample metadata
\item
  classify reads (``hits'') into false positives and false negatives, on
  a replicate-by-replicate basis
\item
  aggregate (sum) false positive and false negative reads for each
  sample and replicate
\item
  merge sample metadata in with the amplicon data
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Create list of amplicon data frames}
\NormalTok{amp_data <-}\StringTok{ }\KeywordTok{list}\NormalTok{(amp_its_genus, amp_its_species, amp_rbc_genus, amp_rbc_species)}

\CommentTok{# create a function for *melting* the amplicon QIIME matrices, with a row for each...}
\CommentTok{# ...taxon / mix / sample combination, reporting the number of reads for each}
\NormalTok{amp_convert <-}\StringTok{ }\KeywordTok{lapply}\NormalTok{(amp_data, }\ControlFlowTok{function}\NormalTok{(x)\{}
  \CommentTok{#convert first column name to "taxa"}
  \KeywordTok{names}\NormalTok{(x)[}\DecValTok{1}\NormalTok{] <-}\StringTok{ "taxa"}
  \CommentTok{#convert from wide to long format}
\NormalTok{  x <-}\StringTok{ }\KeywordTok{melt}\NormalTok{(x, }\DataTypeTok{id.vars=}\StringTok{"taxa"}\NormalTok{, }\DataTypeTok{value.name=}\StringTok{"hits"}\NormalTok{)}
  \CommentTok{#separate mix ID and sample ID}
\NormalTok{  x[,}\DecValTok{4}\OperatorTok{:}\DecValTok{5}\NormalTok{] <-}\StringTok{ }\KeywordTok{colsplit}\NormalTok{(x}\OperatorTok{$}\NormalTok{variable, }\StringTok{"_"}\NormalTok{, }\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"sample"}\NormalTok{))}
  \CommentTok{#remove variable "id"}
\NormalTok{  x <-}\StringTok{ }\NormalTok{x[,}\OperatorTok{-}\DecValTok{2}\NormalTok{]}
  \CommentTok{# remove any rows for which the "hits" count == 0}
  \CommentTok{# there were no reads, so we don't need to include them}
  \CommentTok{# also a good point to remove Zea mays single-species samples}
  \CommentTok{# and also mix #5 which includes Zea}
\NormalTok{  x <-}\StringTok{ }\KeywordTok{filter}\NormalTok{(x, hits}\OperatorTok{!=}\DecValTok{0} \OperatorTok{&}\StringTok{ }\NormalTok{mix.ID }\OperatorTok{!=}\StringTok{ "Z.mays"} \OperatorTok{&}\StringTok{ }\NormalTok{mix.ID }\OperatorTok{!=}\StringTok{ "mix5"}\NormalTok{)}
\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(x, id.vars = "taxa", value.name = "hits"): The melt generic
## in data.table has been passed a data.frame and will attempt to redirect to the
## relevant reshape2 method; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(x). In the next version, this warning will become
## an error.

## Warning in melt(x, id.vars = "taxa", value.name = "hits"): The melt generic
## in data.table has been passed a data.frame and will attempt to redirect to the
## relevant reshape2 method; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(x). In the next version, this warning will become
## an error.

## Warning in melt(x, id.vars = "taxa", value.name = "hits"): The melt generic
## in data.table has been passed a data.frame and will attempt to redirect to the
## relevant reshape2 method; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(x). In the next version, this warning will become
## an error.

## Warning in melt(x, id.vars = "taxa", value.name = "hits"): The melt generic
## in data.table has been passed a data.frame and will attempt to redirect to the
## relevant reshape2 method; please note that reshape2 is deprecated, and this
## redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(x). In the next version, this warning will become
## an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# apply the melting function to each of the 6 datasets}
\CommentTok{# convert them into data frames at the same time}
\CommentTok{# (in the future, better to do this with `lapply` or similar)}
\CommentTok{#amp_its_family <- as.data.frame(amp_convert[1])}
\NormalTok{amp_its_genus <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_convert[}\DecValTok{1}\NormalTok{])}
\NormalTok{amp_its_species <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_convert[}\DecValTok{2}\NormalTok{])}
\CommentTok{#amp_rbc_family <- as.data.frame(amp_convert[4])}
\NormalTok{amp_rbc_genus <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_convert[}\DecValTok{3}\NormalTok{])}
\NormalTok{amp_rbc_species <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_convert[}\DecValTok{4}\NormalTok{])}

\CommentTok{# merge with sample metadata (`mixes`) by mixID and sampleID}
\CommentTok{#amp_its_family_mix <- merge(amp_its_family, mixes, by.x=c("mix.ID", "taxa"), by.y=c("mix.ID", "family"), all.x=T)}
\NormalTok{amp_its_genus_mix <-}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_genus, mixes, }\DataTypeTok{by.x=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"taxa"}\NormalTok{), }\DataTypeTok{by.y=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}
\NormalTok{amp_its_species_mix <-}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_species, mixes, }\DataTypeTok{by.x=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"taxa"}\NormalTok{), }\DataTypeTok{by.y=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}
\CommentTok{#amp_rbc_family_mix <- merge(amp_rbc_family, mixes, by.x=c("mix.ID", "taxa"), by.y=c("mix.ID", "family"), all.x=T)}
\NormalTok{amp_rbc_genus_mix <-}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_rbc_genus, mixes, }\DataTypeTok{by.x=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"taxa"}\NormalTok{), }\DataTypeTok{by.y=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}
\NormalTok{amp_rbc_species_mix <-}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_rbc_species, mixes, }\DataTypeTok{by.x=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"taxa"}\NormalTok{), }\DataTypeTok{by.y=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T)}

\CommentTok{#Rows with mixes variables that are "NA" are false positive taxa. Create a data frame that indicates if row is true positive or false positive}
\CommentTok{#amp_its_family_mix$type <- ifelse(is.na(amp_its_family_mix$question.1), "false_pos", "true_pos")}
\NormalTok{amp_its_genus_mix}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_genus_mix}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}
\NormalTok{amp_its_species_mix}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_species_mix}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}
\CommentTok{#amp_rbc_family_mix$type <- ifelse(is.na(amp_rbc_family_mix$question.1), "false_pos", "true_pos")}
\NormalTok{amp_rbc_genus_mix}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_genus_mix}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}
\NormalTok{amp_rbc_species_mix}\OperatorTok{$}\NormalTok{type <-}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_species_mix}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{), }\StringTok{"false_pos"}\NormalTok{, }\StringTok{"true_pos"}\NormalTok{)}

\CommentTok{#New list of mixed datasets}
\NormalTok{amp_mixed <-}\StringTok{ }\KeywordTok{list}\NormalTok{(amp_its_genus_mix, amp_its_species_mix, amp_rbc_genus_mix, amp_rbc_species_mix)}


\CommentTok{#Apply function to list}
\NormalTok{amp_summed <-}\StringTok{ }\KeywordTok{lapply}\NormalTok{(amp_mixed, }\ControlFlowTok{function}\NormalTok{(x)\{}
  \CommentTok{#summarize number of hits by mix.ID, sample, and type}
\NormalTok{  x <-}\StringTok{ }\NormalTok{x }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{select}\NormalTok{(mix.ID, taxa, hits, sample, type) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{group_by}\NormalTok{(mix.ID, sample, type) }\OperatorTok{%>%}
\StringTok{          }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total_hits =} \KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(hits)))}
  \CommentTok{#Convert to wide format with one column for true positive hits and one column for false positive hits}
  \CommentTok{#x <- reshape(x, idvar="sample", timevar="type", direction="wide")}
\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#amp_its_family_summ <- as.data.frame(amp_summed[1])}
\NormalTok{amp_its_genus_summ <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_summed[}\DecValTok{1}\NormalTok{])}
\NormalTok{amp_its_species_summ <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_summed[}\DecValTok{2}\NormalTok{])}
\CommentTok{#amp_rbc_family_summ <- as.data.frame(amp_summed[4])}
\NormalTok{amp_rbc_genus_summ <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_summed[}\DecValTok{3}\NormalTok{])}
\NormalTok{amp_rbc_species_summ <-}\StringTok{ }\KeywordTok{as.data.frame}\NormalTok{(amp_summed[}\DecValTok{4}\NormalTok{])}

\CommentTok{# Add "A" to beginning of sample ID - couldn't get this to work in the lapply function}
\CommentTok{#amp_its_family_summ$sample <- paste("A_", amp_its_family_summ$sample, sep="")}
\NormalTok{amp_its_genus_summ}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"A_"}\NormalTok{, amp_its_genus_summ}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}
\NormalTok{amp_its_species_summ}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"A_"}\NormalTok{, amp_its_species_summ}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}
\CommentTok{#amp_rbc_family_summ$sample <- paste("A_", amp_rbc_family_summ$sample, sep="")}
\NormalTok{amp_rbc_genus_summ}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"A_"}\NormalTok{, amp_rbc_genus_summ}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}
\NormalTok{amp_rbc_species_summ}\OperatorTok{$}\NormalTok{sample <-}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"A_"}\NormalTok{, amp_rbc_species_summ}\OperatorTok{$}\NormalTok{sample, }\DataTypeTok{sep=}\StringTok{""}\NormalTok{)}

\CommentTok{# Convert to wide format with one column for true positive hits and one column for false positive hits - couldn't get this to work in the lapply function}
\CommentTok{#amp_its_family_reshape <- spread(amp_its_family_summ, key=type, value=total_hits)}
\NormalTok{amp_its_genus_reshape <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(amp_its_genus_summ, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}
\NormalTok{amp_its_species_reshape <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(amp_its_species_summ, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}
\CommentTok{#amp_rbc_family_reshape <- spread(amp_rbc_family_summ, key=type, value=total_hits)}
\NormalTok{amp_rbc_genus_reshape <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(amp_rbc_genus_summ, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}
\NormalTok{amp_rbc_species_reshape <-}\StringTok{ }\KeywordTok{spread}\NormalTok{(amp_rbc_species_summ, }\DataTypeTok{key=}\NormalTok{type, }\DataTypeTok{value=}\NormalTok{total_hits)}

\CommentTok{#If positives are N/A, set to equal 0}
\CommentTok{#amp_its_family_reshape$true_pos = ifelse(is.na(amp_its_family_reshape$true_pos),0, amp_its_family_reshape$true_pos)}
\NormalTok{amp_its_genus_reshape}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_genus_reshape}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, amp_its_genus_reshape}\OperatorTok{$}\NormalTok{true_pos)}
\NormalTok{amp_its_species_reshape}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_species_reshape}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, amp_its_species_reshape}\OperatorTok{$}\NormalTok{true_pos)}
\CommentTok{#amp_rbc_family_reshape$true_pos = ifelse(is.na(amp_rbc_family_reshape$true_pos),0, amp_rbc_family_reshape$true_pos)}
\NormalTok{amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{true_pos)}
\NormalTok{amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{true_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{true_pos),}\DecValTok{0}\NormalTok{, amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{true_pos)}

\CommentTok{#amp_its_family_reshape$false_pos = ifelse(is.na(amp_its_family_reshape$false_pos),0, amp_its_family_reshape$false_pos)}
\NormalTok{amp_its_genus_reshape}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_genus_reshape}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, amp_its_genus_reshape}\OperatorTok{$}\NormalTok{false_pos)}
\NormalTok{amp_its_species_reshape}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_its_species_reshape}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, amp_its_species_reshape}\OperatorTok{$}\NormalTok{false_pos)}
\CommentTok{#amp_rbc_family_reshape$false_pos = ifelse(is.na(amp_rbc_family_reshape$false_pos),0, amp_rbc_family_reshape$false_pos)}
\NormalTok{amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{false_pos)}
\NormalTok{amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{false_pos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{is.na}\NormalTok{(amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{false_pos),}\DecValTok{0}\NormalTok{, amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{false_pos)}


\CommentTok{# merge these data with the sample metadata}
\CommentTok{# use 'mixy' from above which is the sample metadata summarized by sample}
\CommentTok{# (not by taxon within sample, as for 'mixes')}
\CommentTok{#amp_its_family_reshape = merge(amp_its_family_reshape, mixy, by = "mix.ID")}
\NormalTok{amp_its_genus_reshape =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_genus_reshape, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}
\NormalTok{amp_its_species_reshape =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_species_reshape, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}
\CommentTok{#amp_rbc_family_reshape = merge(amp_rbc_family_reshape, mixy, by = "mix.ID")}
\NormalTok{amp_rbc_genus_reshape =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_rbc_genus_reshape, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}
\NormalTok{amp_rbc_species_reshape =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_rbc_species_reshape, mixy, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{)}

\CommentTok{# clean up}
\KeywordTok{rm}\NormalTok{(mixy)}
\end{Highlighting}
\end{Shaded}

\hypertarget{combine-amplicon-and-kraken-data-for-false-positive-analysis}{%
\subsection{combine amplicon and Kraken data for false positive
analysis}\label{combine-amplicon-and-kraken-data-for-false-positive-analysis}}

Combine the WGS with amplicon data by creating a dataframe showing the
number of true positive and the number of false positive reads by
source, mix ID, and sample ID. Straightforward. Updated Dec 2020 to also
take out \emph{Zea} (this had previously been done in a subsequent
step).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Add column "source" to amplicon and kraken data}
\CommentTok{#amp_its_family_reshape$source = "amp"}
\NormalTok{amp_its_genus_reshape}\OperatorTok{$}\NormalTok{source =}\StringTok{ "amp"}
\NormalTok{amp_its_species_reshape}\OperatorTok{$}\NormalTok{source =}\StringTok{ "amp"}
\CommentTok{#amp_rbc_family_reshape$source = "amp"}
\NormalTok{amp_rbc_genus_reshape}\OperatorTok{$}\NormalTok{source =}\StringTok{ "amp"}
\NormalTok{amp_rbc_species_reshape}\OperatorTok{$}\NormalTok{source =}\StringTok{ "amp"}

\CommentTok{#agg.krak.family$source = "krak"}
\NormalTok{agg.krak.genus}\OperatorTok{$}\NormalTok{source =}\StringTok{ "krak"}
\NormalTok{agg.krak.species}\OperatorTok{$}\NormalTok{source =}\StringTok{ "krak"}

\CommentTok{# Merge kraken and amplicon data}
\CommentTok{#krakamp_its_family = rbind(subset(amp_its_family_reshape, mix.ID %in% agg.krak.family$mix.ID), as.data.frame(agg.krak.family))}
\NormalTok{krakamp_its_genus =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{subset}\NormalTok{(amp_its_genus_reshape, mix.ID }\OperatorTok{%in%}\StringTok{ }\NormalTok{agg.krak.genus}\OperatorTok{$}\NormalTok{mix.ID), }\KeywordTok{as.data.frame}\NormalTok{(agg.krak.genus))}
\NormalTok{krakamp_its_species =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{subset}\NormalTok{(amp_its_species_reshape, mix.ID }\OperatorTok{%in%}\StringTok{ }\NormalTok{agg.krak.species}\OperatorTok{$}\NormalTok{mix.ID), }\KeywordTok{as.data.frame}\NormalTok{(agg.krak.species))}
\CommentTok{#krakamp_rbc_family = rbind(subset(amp_rbc_family_reshape, mix.ID %in% agg.krak.family$mix.ID), as.data.frame(agg.krak.family))}
\NormalTok{krakamp_rbc_genus =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{subset}\NormalTok{(amp_rbc_genus_reshape, mix.ID }\OperatorTok{%in%}\StringTok{ }\NormalTok{agg.krak.genus}\OperatorTok{$}\NormalTok{mix.ID), }\KeywordTok{as.data.frame}\NormalTok{(agg.krak.genus))}
\NormalTok{krakamp_rbc_species =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{subset}\NormalTok{(amp_rbc_species_reshape, mix.ID }\OperatorTok{%in%}\StringTok{ }\NormalTok{agg.krak.species}\OperatorTok{$}\NormalTok{mix.ID), }\KeywordTok{as.data.frame}\NormalTok{(agg.krak.species))}

\CommentTok{# Take out Zea}
\CommentTok{#truepos.krak.family = filter(truepos.krak.family, genus!= "Zea") }
\NormalTok{truepos.krak.genus =}\StringTok{ }\KeywordTok{filter}\NormalTok{(truepos.krak.genus, genus}\OperatorTok{!=}\StringTok{"Zea"}\NormalTok{)}
\NormalTok{truepos.krak.species =}\StringTok{ }\KeywordTok{filter}\NormalTok{(truepos.krak.species, genus}\OperatorTok{!=}\StringTok{"Zea"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\hypertarget{analysis}{%
\section{Analysis}\label{analysis}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  false negatives

  \begin{itemize}
  \tightlist
  \item
    WGS data alone, as a function of sample complexity
  \item
    WGS data compared to amplicon data
  \end{itemize}
\item
  false positives

  \begin{itemize}
  \tightlist
  \item
    WGS data alone, as a function of sample complexity
  \item
    WGS data compared to amplicon data
  \end{itemize}
\item
  quantitative matching

  \begin{itemize}
  \tightlist
  \item
    WGS data alone (sample input compared to sample output)
  \item
    WGS data compared to amplicon data
  \end{itemize}
\end{enumerate}

\hypertarget{false-negatives}{%
\subsection{false negatives}\label{false-negatives}}

To assess if species richness, relatedness, and pollen grain proportion
have a significant effect on the ability of the empirical WGS to
qualitatively detect the presence/absence of a species, use a binomial
mixed effects model with species as a random effect and rep.ID nested in
sample nested in mix.ID as a random effect.

For consistency, take out \emph{Zea mays} and format data tables to have
the same name as those in the mixed amplicon analysis.

\hypertarget{sample-complexity-false-negatives}{%
\subsubsection{sample complexity: false
negatives}\label{sample-complexity-false-negatives}}

\begin{itemize}
\tightlist
\item
  the analysis approach here is based on binomial-errors mixed-effects
  models
\item
  the \textbf{fixed effects} are sample complexity; we are assessing one
  facet of sample complexity per model

  \begin{itemize}
  \tightlist
  \item
    Question 1: \textbf{species richness} (1-9 species)
  \item
    Question 2: \textbf{rarity} (actual proportion of grains this taxon
    has in a sample; from roughly half \& half to \textless{} 1\% of the
    rarer type)
  \item
    Question 3: \textbf{taxonomic relatedness} (within genus to across
    broad clades in the seed plants)
  \end{itemize}
\item
  we separately assessed models that considered different
  \textbf{subsets of the data}:

  \begin{itemize}
  \tightlist
  \item
    just the mixes specifically set up for each of the questions above
    (smaller subset of the data)
  \item
    all of the mixes (thus more data)
  \end{itemize}
\item
  we also ran separate models for matching at each of the
  \textbf{taxonomic levels:}

  \begin{itemize}
  \tightlist
  \item
    family, genus, and species
  \end{itemize}
\item
  the \textbf{random effects} are nested, with the highest level as
  \texttt{mix.ID} (i.e.~the identity of the pollen mixture); then
  \texttt{sample} replicate within each mix and finally \texttt{rep.ID}
  (forward / reverse read within a sample, or Illumina lane for samples
  split across lanes)

  \begin{itemize}
  \tightlist
  \item
    we modeled these as random intercepts
  \item
    the random effects were consistent and equivalent for all models
  \end{itemize}
\item
  there are 3 facets of sample complexity times 2 data subsets times 3
  taxonomic levels = \textbf{18 total models}

  \begin{itemize}
  \tightlist
  \item
    set this up with a nested \texttt{for} loop; there are almost
    certainly cleaner and more elegant ways to do this (e.g.~probably
    better to use \texttt{formula} objects and to vectorize)
  \end{itemize}
\end{itemize}

'============================================

\begin{itemize}
\tightlist
\item
  \textbf{update Dec 2020: NO ANALYSES POSSIBLE}

  \begin{itemize}
  \tightlist
  \item
    we found true positives for \textbf{all} samples in the shotgun data
    using the Kraken analysis pipeline
  \item
    (the one exception was \emph{Broussonetia papifera}, which
    apparently was a sequencing failure with very very low overall read
    rates; but even for that species, we did actually get a match, it is
    just that the read rates were so low that they did not pass our
    contamination threshold and were removed)
  \item
    thus, statistical analyses are not possible since essentially 0\% of
    samples had false negatives / 100\% true positives
  \item
    because of that, we set the below code chunk to \texttt{eval\ =\ F}
    (i.e., not run)
  \end{itemize}
\end{itemize}

'============================================

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set up the three factors by which we are running the models}
\NormalTok{taxon =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"species"}\NormalTok{, }\StringTok{"genus"}\NormalTok{)}
\NormalTok{datasubset =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"sub"}\NormalTok{, }\StringTok{"all"}\NormalTok{) }\CommentTok{# whether we are using the designated subset of data designed for the question, or all data}
\NormalTok{question =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"spp.rich"}\NormalTok{, }\StringTok{"relatedness"}\NormalTok{, }\StringTok{"pollen.grain.proportion"}\NormalTok{)}

\CommentTok{# calculate total number of models}
\NormalTok{total =}\StringTok{ }\KeywordTok{length}\NormalTok{(taxon)}\OperatorTok{*}\KeywordTok{length}\NormalTok{(datasubset)}\OperatorTok{*}\KeywordTok{length}\NormalTok{(question)}

\CommentTok{# first set up a table for the results with a number of entries equal to the 'total' variable above (18):}
\NormalTok{results.table =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{question =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{data.subset  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{model.name =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,total), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,total), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total))}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# EXAMPLE FORMULA:}
\CommentTok{# Krak.Q1.species.all = glmer(qual.species.rbcL ~ spp.rich + (1|mix.ID) + (1|species), family = binomial, data = truepos.krak.species, control = glmerControl(optimizer="bobyqa"))}

\CommentTok{# 'for' loop:}

\ControlFlowTok{for}\NormalTok{(q }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{) \{ }\CommentTok{# 'question': response variables for Q1 / Q2 / Q3}
    \ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{)\{ }\CommentTok{# 'taxon': species, genus}
      \ControlFlowTok{for}\NormalTok{(l }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{) \{ }\CommentTok{# 'datasubset': sub or all}
        \CommentTok{# first, name the analysis:}
\NormalTok{        namer =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"Krak.Q"}\NormalTok{, q, }\StringTok{"."}\NormalTok{, taxon[k], }\StringTok{"."}\NormalTok{, datasubset[l], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{        data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"truepos.krak."}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \CommentTok{# third, set up the data subset}
\NormalTok{        subster =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"data.sub = filter("}\NormalTok{, data.to.use, }\StringTok{", question.1 == "}\NormalTok{, q, }\StringTok{" | question.2 == "}\NormalTok{, q, }\StringTok{" | question.3 == "}\NormalTok{,q, }\StringTok{")"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ subster)) }\CommentTok{# probably not the most efficient thing ever... }
        \CommentTok{# fourth, set whether or not data subset is used (vs. all data)}
        \ControlFlowTok{if}\NormalTok{(datasubset[l]}\OperatorTok{==}\StringTok{"sub"}\NormalTok{) \{data.to.use =}\StringTok{ "data.sub"}\NormalTok{\} }\CommentTok{# i.e., doesn't change if all data are to be used}
        \CommentTok{# fifth, set up mixed-effects model: }
\NormalTok{        mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = try(suppressWarnings(glmer(qual."}\NormalTok{, taxon[k],}
          \StringTok{" ~ "}\NormalTok{,  question[q], }\StringTok{" + (1|mix.ID) + (1|"}\NormalTok{, taxon[k], }\StringTok{"), family = binomial, }
\StringTok{          data = "}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{))))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \CommentTok{# sixth, evaluate the mixed-effects model}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}
            \CommentTok{# # eighth, print summary of model [SKIP FOR NOW]}
            \CommentTok{# summarizer = paste("print(summary(", namer, "))", sep = "")}
            \CommentTok{# eval(parse(text = summarizer)) # print summary of the mixed-effects model}
        
        \CommentTok{## extract p-value}
        \CommentTok{# (this would probably be easier using the 'broom.mixed' package?)}
        \CommentTok{# example: coef(summary(Q3.genus.ITS.all))[2,4]}
\NormalTok{        pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
        
        \CommentTok{# extract convergence failures}
\NormalTok{        converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{        converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{        converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
        
        \CommentTok{# record results in table}
\NormalTok{        results.table[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{question[q]}
\NormalTok{        results.table[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
\NormalTok{        results.table[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\NormalTok{datasubset[l]}
\NormalTok{        results.table[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\NormalTok{namer}
\NormalTok{        results.table[tracker,}\DecValTok{5}\NormalTok{] =}\StringTok{ }\NormalTok{pval}
\NormalTok{        results.table[tracker,}\DecValTok{6}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{        results.table[tracker,}\DecValTok{7}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
        
        \CommentTok{# advance tracker}
\NormalTok{        tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{      \}}
\NormalTok{    \}}
\NormalTok{  \}}


\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package and `kable_styling` is from the `kableExtra` package }
\KeywordTok{kable}\NormalTok{(results.table) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\hypertarget{amplicon-vs.-shotgun-comparison-false-negatives}{%
\subsubsection{Amplicon vs.~Shotgun comparison: false
negatives}\label{amplicon-vs.-shotgun-comparison-false-negatives}}

This analysis uses binomial mixed models to assess if using amplicon
vs.~WGS has an effect on the ability to qualitatively detect true
positives vs.~false negatives. This analysis is similar to the
qualitative mixed models to assess the effect of species richness,
pollen grain proportion, etc. on the ability to qualitatively detect the
correct taxa, but instead using ``source'' as the (sole) fixed effect.

\hypertarget{by-marker-rbcl-vs-its2amplicon-vs.-shotgun-false-negatives}{%
\paragraph{\texorpdfstring{by marker (\emph{rbcL} vs ITS2)---Amplicon
vs.~Shotgun false
negatives}{by marker (rbcL vs ITS2)---Amplicon vs.~Shotgun false negatives}}\label{by-marker-rbcl-vs-its2amplicon-vs.-shotgun-false-negatives}}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{taxon =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"species"}\NormalTok{, }\StringTok{"genus"}\NormalTok{)}
\NormalTok{marker =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"its"}\NormalTok{, }\StringTok{"rbc"}\NormalTok{)}


\CommentTok{# first set up a table for the results:}
\NormalTok{rowz =}\StringTok{ }\KeywordTok{length}\NormalTok{(taxon)}\OperatorTok{*}\KeywordTok{length}\NormalTok{(marker)}

\NormalTok{krakamp.fn.results.table =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz), }\DataTypeTok{marker  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz), }\DataTypeTok{model.name =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,rowz), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,rowz), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz))}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# # EXAMPLE FORMULA}
\CommentTok{# its.family = glmer(qual.family ~ source + (1|mix.ID/sample) + (1|family), family = binomial, data = krakamp_its_family_fn, control = glmerControl(optimizer="bobyqa"))}

\CommentTok{# response variables relating to each of the three questions (column names in data)}

\ControlFlowTok{for}\NormalTok{(q }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{) \{ }\CommentTok{# 'marker': its or rbc}
    \ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{)\{ }\CommentTok{# 'taxon': species, genus, family}
        \CommentTok{# first, name the analysis (i.e. name the result object):}
\NormalTok{        namer =}\StringTok{ }\KeywordTok{paste}\NormalTok{(marker[q], }\StringTok{"."}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{        data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"krakamp_"}\NormalTok{, marker[q], }\StringTok{"_"}\NormalTok{, taxon[k],}\StringTok{"_fn"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# third, set up mixed-effects model: }
\NormalTok{        mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = suppressWarnings(glmer(qual."}\NormalTok{, taxon[k],}
          \StringTok{" ~ source + (1|mix.ID/sample) + (1|"}\NormalTok{, taxon[k], }\StringTok{"), family = binomial, }
\StringTok{          data = "}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{)))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# fourth, evaluate the mixed-effects model}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}
            \CommentTok{#fifth, print summary of model [SKIP FOR NOW]}
            \CommentTok{#summarizer = paste("print(summary(", namer, "))", sep = "")}
            \CommentTok{#eval(parse(text = summarizer)) # print summary of the mixed-effects model}
        
        \CommentTok{# fifth, extract the results}
        \CommentTok{## extract p-value}
        \CommentTok{# example: coef(summary(its.family))[2,4]}
\NormalTok{        pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
        
        \CommentTok{# extract convergence failures}
\NormalTok{        converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{        converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{        converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
        
        \CommentTok{# record results in table}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{marker[q]}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\NormalTok{namer}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\NormalTok{pval}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{5}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{        krakamp.fn.results.table[tracker,}\DecValTok{6}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
        
        \CommentTok{# advance tracker}
\NormalTok{        tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{      \}}
\NormalTok{    \}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Error in mkRespMod(fr, family = family): response must be numeric or factor
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package and `kable_styling` is from the `kableExtra` package }
\KeywordTok{kable}\NormalTok{(krakamp.fn.results.table) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|l|l|r|r|l}
\hline
taxon & marker & model.name & p.val & n & warning.msg\\
\hline
NA & NA & NA & 1.000001 & 9999 & NA\\
\hline
NA & NA & NA & 1.000001 & 9999 & NA\\
\hline
NA & NA & NA & 1.000001 & 9999 & NA\\
\hline
NA & NA & NA & 1.000001 & 9999 & NA\\
\hline
\end{tabular}
\end{table}

\textbf{Results interpretation updated Dec 2020}

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} shotgun / Kraken outperforms amplicon
  approach in terms of identifying true positives
\item
  not surprising since the shotgun approach generated a true positive
  match for essentially every taxon in our samples
\item
  this is true irrespective of amplicon marker (\emph{rbcL} or ITS2) or
  taxonomic resolution (genus vs.~species)
\end{itemize}

\hypertarget{markers-combinedamplicon-vs.-shotgun-false-negatives}{%
\paragraph{markers combined---Amplicon vs.~Shotgun false
negatives}\label{markers-combinedamplicon-vs.-shotgun-false-negatives}}

This is a repeat of the analysis above, but with ITS2 and \emph{rbcL}
combined

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Create new qualitative variable in amplicon dataset that equals 1 if ITS and/or rbcL equals 1, 0 if else}

\CommentTok{#amp_all$qual.family = ifelse(amp_all$qual.family.ITS == 1 | amp_all$qual.family.rbcL == 1, 1, 0)}
\NormalTok{amp_all}\OperatorTok{$}\NormalTok{qual.genus =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(amp_all}\OperatorTok{$}\NormalTok{qual.genus.ITS }\OperatorTok{==}\StringTok{ }\DecValTok{1} \OperatorTok{|}\StringTok{ }\NormalTok{amp_all}\OperatorTok{$}\NormalTok{qual.genus.rbcL }\OperatorTok{==}\StringTok{ }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{)}
\NormalTok{amp_all}\OperatorTok{$}\NormalTok{qual.species =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(amp_all}\OperatorTok{$}\NormalTok{qual.species.ITS }\OperatorTok{==}\StringTok{ }\DecValTok{1} \OperatorTok{|}\StringTok{ }\NormalTok{amp_all}\OperatorTok{$}\NormalTok{qual.species.rbcL }\OperatorTok{==}\StringTok{ }\DecValTok{1}\NormalTok{, }\DecValTok{1}\NormalTok{, }\DecValTok{0}\NormalTok{)}

\CommentTok{#Combine kraken and amplicon data}
\CommentTok{#krakamp_family_fn <- rbind(truepos.krak.family[,c("mix.ID", "family", "sample", "qual.family", "source")], amp_all[,c("mix.ID","family","sample", "qual.family", "source")])}

\NormalTok{krakamp_genus_fn <-}\StringTok{ }\KeywordTok{rbind}\NormalTok{(truepos.krak.genus[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"genus"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.genus"}\NormalTok{, }\StringTok{"source"}\NormalTok{)])}

\NormalTok{krakamp_species_fn <-}\StringTok{ }\KeywordTok{rbind}\NormalTok{(truepos.krak.species[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species"}\NormalTok{, }\StringTok{"source"}\NormalTok{)], amp_all[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"species"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"qual.species"}\NormalTok{, }\StringTok{"source"}\NormalTok{)])}


\NormalTok{taxon =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"species"}\NormalTok{, }\StringTok{"genus"}\NormalTok{)}

\CommentTok{# first set up a table for the results:}
\CommentTok{# I will set this up with 18 entries}
\NormalTok{krakamp.fn.results.combined =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{model.name =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,}\DecValTok{2}\NormalTok{))}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# # EXAMPLE FORMULA}
\CommentTok{# its.family = glmer(qual.family ~ source + (1|mix.ID/sample) + (1|family), family = binomial, data = krakamp_family_fn, control = glmerControl(optimizer="bobyqa"))}

\CommentTok{# response variables relating to each of the three questions (column names in data)}

    \ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{)\{ }\CommentTok{# 'taxon': species, genus}
        \CommentTok{# first, name the analysis:}
\NormalTok{        namer =}\StringTok{ }\NormalTok{taxon[k]}
        
        \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{        data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"krakamp_"}\NormalTok{, taxon[k],}\StringTok{"_fn"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# third, set up mixed-effects model: }
\NormalTok{        mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = suppressWarnings(glmer(qual."}\NormalTok{, taxon[k],}
          \StringTok{" ~ source + (1|mix.ID/sample) + (1|"}\NormalTok{, taxon[k], }\StringTok{"), family = binomial, }
\StringTok{          data = "}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{)))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \CommentTok{# fourth, evaluate the mixed-effects model}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}

        \CommentTok{# fifth, extract model information}
        \CommentTok{## extract p-value}
        \CommentTok{# example: coef(summary(its.family))[2,4]}
\NormalTok{        pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
        
        \CommentTok{# extract convergence failures}
\NormalTok{        converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{        converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{        converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
        
        \CommentTok{# record results in table}
\NormalTok{        krakamp.fn.results.combined[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
\NormalTok{        krakamp.fn.results.combined[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{namer}
\NormalTok{        krakamp.fn.results.combined[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\NormalTok{pval}
\NormalTok{        krakamp.fn.results.combined[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{        krakamp.fn.results.combined[tracker,}\DecValTok{5}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
        
        \CommentTok{# advance tracker}
\NormalTok{        tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{      \}}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## boundary (singular) fit: see ?isSingular
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package and `kable_styling` is from the `kableExtra` package }
\KeywordTok{kable}\NormalTok{(krakamp.fn.results.combined) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|l|r|r|l}
\hline
taxon & model.name & p.val & n & warning.msg\\
\hline
species & species & 0.0027829 & 294 & \\
\hline
genus & genus & 0.4333710 & 294 & \\
\hline
\end{tabular}
\end{table}

\textbf{updated results Dec 2020}

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} shotgun approach works better than the two
  amplicon markers combined, when focusing on species-level resolution;
  but there is no statistically significant difference when focusing on
  the genus level
\item
  makes sense in light of the results above
\item
  at the genus level, using both amplicon markers gets you true
  positives essentially all of the time, so it isn't surprising that
  there wasn't a difference
\end{itemize}

\hypertarget{false-positives}{%
\subsection{false positives}\label{false-positives}}

\hypertarget{sample-complexity-false-positives}{%
\subsubsection{sample complexity: false
positives}\label{sample-complexity-false-positives}}

this analysis has the same structure as for the false negatives analysis
of sample complexity, again examining 3 taxonomic levels (family / genus
/ species) for three dimensions of sample complexity (species richness /
taxonomic relatedness / rarity).

\begin{itemize}
\tightlist
\item
  \textbf{update Dec 2020:} we had intended to also run this for each of
  two datasets (the subsample specifically designed for the sample
  complexity component of interest and the entire dataset), but did not
  do so, I (Berry) think that when we merged the forward and reverse
  reads we ended up with too few samples to do the subsets (even when
  accounting previously for the lack of independence of forward and
  reverse reads)
\item
  take home is that all analyses are only from all of the data, not the
  subset
\end{itemize}

Some difference from the false negatives analysis:

\begin{itemize}
\tightlist
\item
  the response variable needs to change

  \begin{itemize}
  \tightlist
  \item
    formerly (for false negatives) it was a 0/1 column which had info on
    whether or not each taxon within a mixture was present or not
  \item
    in this analysis, the response variable is two columns, the count of
    true positive reads and the count of false positive reads, which are
    bound together with \texttt{cbind}
  \end{itemize}
\item
  the random effect for taxon is removed, since data are not aggregated
  taxonomically (instead they are aggregated by sample, i.e.~replicate
  within each mix)
\item
  since the data are aggregated, for pollen grain proportion the
  \emph{minimum} proportion (i.e.~the proportion of the smallest
  quantity of pollen in any given mix) was used as the explanatory
  variable
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# set up the three factors by which we are running the models}
\NormalTok{taxon =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"species"}\NormalTok{, }\StringTok{"genus"}\NormalTok{)}
\CommentTok{# datasubset = c("sub", "all") # whether we are using the designated subset of data designed for the question, or all data}
\CommentTok{# taking this out, Dec 2020, because data subsets are too small to be meaningful and may be throwing other analyses off}
\NormalTok{question =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"spp.rich"}\NormalTok{, }\StringTok{"relatedness"}\NormalTok{, }\StringTok{"pollen.grain.proportion"}\NormalTok{)}

\CommentTok{# calculate total number of models}
\NormalTok{total =}\StringTok{ }\KeywordTok{length}\NormalTok{(taxon)}\OperatorTok{*}\KeywordTok{length}\NormalTok{(question) }\CommentTok{# was previously also *length(datasubset)}

\CommentTok{# first set up a table for the results with a number of entries equal to the 'total' variable above (18):}
\NormalTok{results.table.falsepos =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{question =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{model.name =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,total), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,total), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,total))}
\CommentTok{# previously also included 'data.subset  = rep(NA,total)'}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# EXAMPLE FORMULA:}
\CommentTok{# Krak.Q1.species.all = glmer(qual.species.rbcL ~ spp.rich + (1|mix.ID/sample) + (1|species), family = binomial, data = agg.krak.species, control = glmerControl(optimizer="bobyqa"))}

\CommentTok{# 'for' loop:}

\ControlFlowTok{for}\NormalTok{(q }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{) \{ }\CommentTok{# 'question': response variables for Q1 / Q2 / Q3}
  \ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{)\{ }\CommentTok{# 'taxon': species, genus}
\CommentTok{#    for(l in 1:2) \{ # 'datasubset': sub or all}
      \CommentTok{# first, name the analysis:}
\NormalTok{      namer =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"Krak.falsepos.Q"}\NormalTok{, q, }\StringTok{"."}\NormalTok{, taxon[k], }\StringTok{"."}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{) }\CommentTok{# had also included 'datasubset[l],'}
      \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{      data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"agg.krak."}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
      \CommentTok{# third, set up the data subset}
\NormalTok{      subster =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"data.sub = filter("}\NormalTok{, data.to.use, }\StringTok{", question.1 == "}\NormalTok{, q, }\StringTok{" | question.2 == "}\NormalTok{, q, }\StringTok{" | question.3 == "}\NormalTok{, q, }\StringTok{")"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
      \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ subster)) }\CommentTok{# probably not the most efficient thing ever... }
      \CommentTok{# fourth, set whether or not data subset is used (vs. all data)}
      \CommentTok{# if(datasubset[l]=="sub") \{data.to.use = "data.sub"\} # i.e., doesn't change if all data are to be used}
        \CommentTok{# above taken out when datasubsets removed}
      \CommentTok{# fifth, set up mixed-effects model: }
\NormalTok{      mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = suppressWarnings(glmer(cbind(false_pos, true_pos) ~ "}\NormalTok{,  question[q], }\StringTok{" + (1|mix.ID), family = binomial, }
\StringTok{                    data = "}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{, optCtrl = list(maxfun = 10000))))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
      
      \CommentTok{# }\AlertTok{NOTE}\CommentTok{ (update Dec 2020): random effects specification changed from}
      \CommentTok{# (1|mix.ID/sample) to}
      \CommentTok{# (1|mix.ID) given that there is only one replicate per sample now}
      \CommentTok{# this is because of merging forward and reverse reads}
      \CommentTok{# this did not change anything at all results-wise (exact same *p*-values)}
      
      \CommentTok{# sixth, evaluate the mixed-effects model}
      \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}
      \CommentTok{# # eighth, print summary of model [SKIP FOR NOW]}
      \CommentTok{# summarizer = paste("print(summary(", namer, "))", sep = "")}
      \CommentTok{# eval(parse(text = summarizer)) # print summary of the mixed-effects model}
      
      \CommentTok{## extract p-value}
      \CommentTok{# (this would probably be easier using the 'broom.mixed' package?)}
      \CommentTok{# example: coef(summary(Q3.genus.ITS.all))[2,4]}
\NormalTok{      pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
      \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
      
      \CommentTok{# extract convergence failures}
\NormalTok{      converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{      converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{      converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
      
      \CommentTok{# record results in table}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{question[q]}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
      \CommentTok{# results.table.falsepos[tracker,3] = datasubset[l]}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\NormalTok{namer }\CommentTok{# was '[tracker, 4]'}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\NormalTok{pval }\CommentTok{# was '[tracker, 5]' (etc for 6 and 7)}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{5}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{      results.table.falsepos[tracker,}\DecValTok{6}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
      
      \CommentTok{# advance tracker}
\NormalTok{      tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{    \}}
\NormalTok{  \}}
\CommentTok{# \} # (was here for datasubset)}

\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package and `kable_styling` is from the `kableExtra` package }
\KeywordTok{kable}\NormalTok{(results.table.falsepos) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|l|l|r|r|l}
\hline
question & taxon & model.name & p.val & n & warning.msg\\
\hline
spp.rich & species & Krak.falsepos.Q1.species. & 0.8957633 & 17 & \\
\hline
spp.rich & genus & Krak.falsepos.Q1.genus. & 0.3810926 & 17 & \\
\hline
relatedness & species & Krak.falsepos.Q2.species. & 0.5412559 & 17 & \\
\hline
relatedness & genus & Krak.falsepos.Q2.genus. & 0.0988828 & 17 & \\
\hline
pollen.grain.proportion & species & Krak.falsepos.Q3.species. & 0.6783320 & 17 & \\
\hline
pollen.grain.proportion & genus & Krak.falsepos.Q3.genus. & 0.0887988 & 17 & \\
\hline
\end{tabular}
\end{table}

\textbf{Results interpretation: updated Dec 2020}

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} we do not see any statistically
  significant relationships between any of the sample complexity
  components (species richness, relatedness, or minimum pollen grain
  proportions) and increasing false positive reads, assessed at either
  the genus or the species level

  \begin{itemize}
  \tightlist
  \item
    two of the results were marginally significant (with marginality set
    at \(0.05 < p < 0.10\))
  \item
    I SUGGEST NOT DISCUSSING THESE MARGINAL RESULTS---not worth it
    relative to our more substantive results, and ultimately they are
    not statistically significant
  \item
    both were for genus-level data
  \item
    pollen grain relatedness: while marginal, if you look at figure J
    you can see that while the CIs are tiny, there doesn't really seem
    to be any \emph{meaningful} difference among relatedness categories
  \item
    pollen grain proportion: while marginal, again, tiny CIs and a
    pattern that doesn't seem particularly meaningful, with slightly
    higher false positive rates in the middle range of the minimum
    proportions (hard to interpret---we were expecting more false
    positives when minimum proportions got lower)
  \end{itemize}
\item
  again, we did not analyze the data based on the ``subsample'' of data
  designed for each question, instead we used all data, otherwise we
  were left with sample sizes e.g.~of 4 and 2 for many of the questions
\item
  in contrast to previous results, we did not get any warning messages /
  models seemed to fit well; this could be in part due to the fact that
  we did not have within-sample random effects due to the merging of the
  forward and reverse reads
\end{itemize}

\hypertarget{amplicon-vs.-shotgun-comparison-false-positives}{%
\subsubsection{Amplicon vs.~Shotgun comparison, false
positives}\label{amplicon-vs.-shotgun-comparison-false-positives}}

This analysis uses binomial mixed models to assess if using amplicon
vs.~WGS approaches return different proportions of false positive
results (relative to true positives).

\hypertarget{amplicon-markers-considered-separately-rbcl-and-its2}{%
\paragraph{\texorpdfstring{Amplicon markers considered separately
(\emph{rbcL} and
ITS2)}{Amplicon markers considered separately (rbcL and ITS2)}}\label{amplicon-markers-considered-separately-rbcl-and-its2}}

First we will split out the analysis by amplicon marker (\emph{rbcL}
vs.~ITS2) and by taxonomic level of identification (family / genus /
species); next we will combine both amplicons but continue to assess
matches at all three taxonomic levels.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{taxon =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"species"}\NormalTok{, }\StringTok{"genus"}\NormalTok{)}
\NormalTok{marker =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"its"}\NormalTok{, }\StringTok{"rbc"}\NormalTok{)}

\CommentTok{# set up a data frame for the results: }
\CommentTok{# I will set this up with 6 rows (3 taxonomy levels x 2 markers)}
\NormalTok{rowz =}\StringTok{ }\KeywordTok{length}\NormalTok{(taxon)}\OperatorTok{*}\KeywordTok{length}\NormalTok{(marker)}
\NormalTok{krakamp.results.table =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz), }\DataTypeTok{marker  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,rowz), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,rowz), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,rowz))}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# # EXAMPLE FORMULA}
\CommentTok{# glmer(cbind(true_pos,false_pos) ~ source + (1|mix.ID/sample), family = binomial, data = krakamp_its_species, control = glmerControl(optimizer="bobyqa"))}


\ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{) \{ }\CommentTok{# 'taxon': species, genus, family}
    \ControlFlowTok{for}\NormalTok{(l }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{)\{ }\CommentTok{# 'marker': ITS2 or rbcL}
        \CommentTok{# first, name the analysis:}
\NormalTok{        namer =}\StringTok{ }\KeywordTok{paste}\NormalTok{(taxon[k], }\StringTok{"."}\NormalTok{, marker[l], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{        data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"krakamp_"}\NormalTok{, marker[l], }\StringTok{"_"}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# third, set up mixed-effects model: }
\NormalTok{        mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = suppressWarnings(glmer(cbind(true_pos,false_pos) ~ source + (1|mix.ID/sample), family = binomial, data ="}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{)))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# fourth, evaluate the mixed-effects model}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}
        
        \CommentTok{## extract p-value}
\NormalTok{        pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
        
        \CommentTok{# extract convergence failures}
\NormalTok{        converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{        converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{        converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
        
        \CommentTok{# record results in table}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{marker[l]}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\NormalTok{pval}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{5}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
        
        \CommentTok{# advance tracker}
\NormalTok{        tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{      \}}
\NormalTok{    \}}
  
\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package, which I required at the very top of this document. }
\KeywordTok{kable}\NormalTok{(krakamp.results.table) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|l|r|r|l}
\hline
taxon & marker & p.val & n & warning.msg\\
\hline
species & its & 0.9102740 & 54 & \\
\hline
species & rbc & 0.5067176 & 54 & \\
\hline
genus & its & 0.0000454 & 54 & \\
\hline
genus & rbc & 0.6917513 & 54 & \\
\hline
\end{tabular}
\end{table}

\textbf{updated results Dec 2020}

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} somewhat surprisingly, with the new
  analysis these results have now changed and for the most part we don't
  see a difference between shotgun- and amplicon-based approaches in
  terms of false positives
\item
  the one exception is for genus-level resolution, when shotgun data are
  compared to amplicon data from ITS2
\item
  based on Figure 6, we are seeing statistically higher rates of false
  positives for the shotgun data relative to genus-level amplicon data
  using ITS2
\end{itemize}

\hypertarget{amplicon-markers-combined}{%
\paragraph{amplicon markers combined}\label{amplicon-markers-combined}}

for this analysis we need to format the data a little differently; the
formatted amplicon data are in different data frames based on the marker
(ITS2 vs.~\emph{rbcL}). Need to combine those two to generate a single
amplicon data frame; this has to happen at each taxonomic level. While
seemingly easy with \texttt{rbind}, we have to take the additional step
of aggregating for each mix.id and sample.id the two amplicon markers
(add the false positives and false negatives together). Still, this is
very straightforward with \texttt{dplyr}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# first combine the two amplicon datasets, aggregating counts for the same mix and sample IDs}
\CommentTok{# this has to happen at each taxonomic level}
\CommentTok{#ampkrak.family = rbind(amp_its_family_reshape, amp_rbc_family_reshape) %>% group_by(mix.ID, sample, rep.ID, source) %>% summarize(false_pos = sum(false_pos), true_pos = sum(true_pos))}

\NormalTok{ampkrak.genus =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(amp_its_genus_reshape, amp_rbc_genus_reshape) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(mix.ID, sample, source) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{false_pos =} \KeywordTok{sum}\NormalTok{(false_pos), }\DataTypeTok{true_pos =} \KeywordTok{sum}\NormalTok{(true_pos))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{ampkrak.species =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(amp_its_species_reshape, amp_rbc_species_reshape) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(mix.ID, sample, source) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{false_pos =} \KeywordTok{sum}\NormalTok{(false_pos), }\DataTypeTok{true_pos =} \KeywordTok{sum}\NormalTok{(true_pos))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'mix.ID', 'sample' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# then combine amplicon data with kraken data}
\CommentTok{#ampkrak.family = rbind(data.frame(ampkrak.family), data.frame(select(agg.krak.family, mix.ID, sample, rep.ID, source, false_pos, true_pos)))}
\NormalTok{ampkrak.genus =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{data.frame}\NormalTok{(ampkrak.genus), }\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{select}\NormalTok{(agg.krak.genus, mix.ID, sample, source, false_pos, true_pos)))}
\NormalTok{ampkrak.species =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(}\KeywordTok{data.frame}\NormalTok{(ampkrak.species), }\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{select}\NormalTok{(agg.krak.species, mix.ID, sample, source, false_pos, true_pos)))}

\CommentTok{# set up a data frame for the results with 3 rows (by taxonomic level)}
\NormalTok{krakamp.results.table =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\DataTypeTok{taxon  =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{p.val  =} \KeywordTok{rep}\NormalTok{(}\FloatTok{1.000001}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{n  =} \KeywordTok{rep}\NormalTok{(}\DecValTok{9999}\NormalTok{,}\DecValTok{2}\NormalTok{), }\DataTypeTok{warning.msg =} \KeywordTok{rep}\NormalTok{(}\OtherTok{NA}\NormalTok{,}\DecValTok{2}\NormalTok{))}

\CommentTok{# keep track of which row of the table to record in:}
\NormalTok{tracker =}\StringTok{ }\DecValTok{1}

\CommentTok{# # EXAMPLE FORMULA}
\CommentTok{# glmer(cbind(true_pos,false_pos) ~ source + (1|mix.ID/sample), family = binomial, data = krakamp_its_species, control = glmerControl(optimizer="bobyqa"))}


\ControlFlowTok{for}\NormalTok{(k }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\DecValTok{2}\NormalTok{) \{ }\CommentTok{# 'taxon': species, genus, family}
        \CommentTok{# first, name the analysis:}
\NormalTok{        namer =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"krakamp.GLMM"}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{"."}\NormalTok{)}
        
        \CommentTok{# second, set which taxonomic data to use:}
\NormalTok{        data.to.use =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"ampkrak"}\NormalTok{, taxon[k], }\DataTypeTok{sep =} \StringTok{"."}\NormalTok{)}
        
        \CommentTok{# third, set up mixed-effects model: }
\NormalTok{        mixed =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{" = suppressWarnings(glmer(cbind(true_pos,false_pos) ~ source + (1|mix.ID/sample), family = binomial, data = "}\NormalTok{, data.to.use, }\StringTok{", control = glmerControl(optimizer=}\CharTok{\textbackslash{}"}\StringTok{bobyqa}\CharTok{\textbackslash{}"}\StringTok{)))"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        
        \CommentTok{# fourth, evaluate the mixed-effects model}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ mixed))}
        
        \CommentTok{## extract p-value}
\NormalTok{        pvaller =}\StringTok{ }\KeywordTok{paste}\NormalTok{(}\StringTok{"pval <- coef(summary("}\NormalTok{, namer, }\StringTok{"))[2,4]"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
        \KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ pvaller))}
        
        \CommentTok{# extract convergence failures}
\NormalTok{        converger =}\StringTok{ }\KeywordTok{paste}\NormalTok{(namer, }\StringTok{"@optinfo$conv$lme4$code"}\NormalTok{, }\DataTypeTok{sep =} \StringTok{""}\NormalTok{)}
\NormalTok{        converg =}\StringTok{ }\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ converger))}
\NormalTok{        converg.return =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\KeywordTok{length}\NormalTok{(converg)}\OperatorTok{==}\DecValTok{1}\NormalTok{, }\StringTok{"ERROR!!"}\NormalTok{, }\StringTok{""}\NormalTok{)}
        
        \CommentTok{# record results in table}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{1}\NormalTok{] =}\StringTok{ }\NormalTok{taxon[k]}
        \CommentTok{# krakamp.results.table[tracker,2] = marker[l]}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{2}\NormalTok{] =}\StringTok{ }\NormalTok{pval}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{3}\NormalTok{] =}\StringTok{ }\KeywordTok{nrow}\NormalTok{(}\KeywordTok{eval}\NormalTok{(}\KeywordTok{parse}\NormalTok{(}\DataTypeTok{text =}\NormalTok{ data.to.use)))}
\NormalTok{        krakamp.results.table[tracker,}\DecValTok{4}\NormalTok{] =}\StringTok{ }\NormalTok{converg.return}
        
        \CommentTok{# advance tracker}
\NormalTok{        tracker =}\StringTok{ }\NormalTok{tracker }\OperatorTok{+}\StringTok{ }\DecValTok{1}
\NormalTok{    \}}

  
\CommentTok{# display results table}
\CommentTok{# note that the 'kable' function is part of the 'knitr' package, which I required at the very top of this document. }
\KeywordTok{kable}\NormalTok{(krakamp.results.table) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|r|r|l}
\hline
taxon & p.val & n & warning.msg\\
\hline
species & 0.9694023 & 106 & \\
\hline
genus & 0.0993345 & 106 & \\
\hline
\end{tabular}
\end{table}

*updated results Dec 2020**

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} consistently with the results for the
  amplicon markers split out, with the new analysis these results have
  now changed; we don't see a statistically significant difference
  between shotgun- and amplicon-based approaches in terms of false
  positives
\item
  there is now a \emph{marginally} significant result for genus-level
  resolution
\item
  based on Figure 6, we are seeing statistically higher rates of false
  positives for the shotgun data relative to genus-level amplicon data
\end{itemize}

\hypertarget{quantitative-matching}{%
\subsection{quantitative matching}\label{quantitative-matching}}

To assess if pollen grain proportion is correlated to read proportion in
the empirical WGS data, use mixed-effects models with mix.ID and species
as random effects. Using the same dataset as for the false negatives
analysis, i.e.~data are formatted to have each taxon (truly present)
within each sample representing a row.

There are two analyses:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  WGS data only: quantitative matching
\item
  amplicon + WGS data: comparison on which gives us better quantitative
  matching
\end{enumerate}

\hypertarget{wgs-data-only-quantitative-matching}{%
\subsubsection{WGS data only: quantitative
matching}\label{wgs-data-only-quantitative-matching}}

Here the question is: do we see a statistical relationship between the
proportion of pollen grains input into a sample and the proportion of
read counts that are output? As described in the \textbf{Overview}
section at the beginning, we do this with a linear mixed-effects model
as we are looking for a linear relationship between the input pollen
grain proportions, and the proportions of output sample reads. A
logistic relationship, as would likely be predicted with a
binomial-errors model, is not helpful to an end-user here even though
the data are ultimately binomial in nature.

We tried specifying these models with intercept = 0. Unfortunately, we
can't estimate an \(R^2\) with a specification of zero intercept, at
least with the \texttt{r2beta} function (not sure why). Either way, we
ultimately didn't do this even though we could get the models to work.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# linear mixed-effects model}

\CommentTok{# ultimately want 3 analyses: \{species, genus, and family\}}
\CommentTok{# within each, we will do a separate analysis without the nested random effect of the 'replicate ID', and with a zero intercept. We will do these to make them more directly comparable to the amplicon analyses below. }

\CommentTok{# species}
\NormalTok{quant.species =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.species }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{species), }\DataTypeTok{data =}\NormalTok{ truepos.krak.species)}
\CommentTok{#quant.species.nonest = lmer(quant.species ~ pollen.grain.proportion + (1|mix.ID) + (1|species), data = truepos.krak.species)}

\CommentTok{# genus}
\NormalTok{quant.genus =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.genus }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{genus), }\DataTypeTok{data =}\NormalTok{ truepos.krak.genus)}
\CommentTok{#quant.genus.nonest = lmer(quant.genus ~ pollen.grain.proportion + (1|mix.ID) + (1|genus), data = truepos.krak.genus)}

\CommentTok{#family}
\CommentTok{#quant.family = lmer(quant.family ~ pollen.grain.proportion + (1|mix.ID/krak.rep.ID) + (1|family), data = truepos.krak.family)}
\CommentTok{#quant.family.nonest = lmer(quant.family ~ pollen.grain.proportion + (1|mix.ID) + (1|family), data = truepos.krak.family)}

\CommentTok{#=========================================}
\CommentTok{# r-squared calculation at the family level}
\CommentTok{#r2.family = r2beta(quant.family)}
\CommentTok{#r2.family.nonest = r2beta(quant.family.nonest)}

\CommentTok{# r-squared calculation at the genus level}
\NormalTok{r2.genus =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.genus)}
\CommentTok{#r2.genus.nonest = r2beta(quant.genus.nonest)}

\CommentTok{# r-squared calculation at the species level}
\NormalTok{r2.species =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.species)}
\CommentTok{#r2.species.nonest = r2beta(quant.species.nonest)}

\CommentTok{#Merge the slope ("Estimate"), p-value ("Pr...t..") from the mixed models with the r-squared value for each test}
\CommentTok{#coefs_quant.family <- cbind((data.frame(coef(summary(quant.family)))["pollen.grain.proportion",c("Estimate","Pr...t..")]), r2.family[2,6])}
\CommentTok{#coefs_quant.family.nonest <- cbind((data.frame(coef(summary(quant.family.nonest)))["pollen.grain.proportion",c("Estimate","Pr...t..")]), r2.family.nonest[2,6])}

\NormalTok{coefs_quant.genus <-}\StringTok{ }\KeywordTok{cbind}\NormalTok{((}\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{coef}\NormalTok{(}\KeywordTok{summary}\NormalTok{(quant.genus)))[}\StringTok{"pollen.grain.proportion"}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\StringTok{"Estimate"}\NormalTok{,}\StringTok{"Pr...t.."}\NormalTok{)]), r2.genus[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{]) }
\CommentTok{#coefs_quant.genus.nonest <- cbind((data.frame(coef(summary(quant.genus.nonest)))["pollen.grain.proportion",c("Estimate","Pr...t..")]), r2.genus.nonest[2,6]) }

\NormalTok{coefs_quant.species <-}\StringTok{ }\KeywordTok{cbind}\NormalTok{((}\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{coef}\NormalTok{(}\KeywordTok{summary}\NormalTok{(quant.species)))[}\StringTok{"pollen.grain.proportion"}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\StringTok{"Estimate"}\NormalTok{,}\StringTok{"Pr...t.."}\NormalTok{)]), r2.species[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{]) }
\CommentTok{#coefs_quant.species.nonest <- cbind((data.frame(coef(summary(quant.species.nonest)))["pollen.grain.proportion",c("Estimate","Pr...t..")]), r2.species.nonest[2,6])}

\CommentTok{#Rename row names }
\CommentTok{#row.names(coefs_quant.family) = "Family"}
\KeywordTok{row.names}\NormalTok{(coefs_quant.genus) =}\StringTok{ "Genus"}
\KeywordTok{row.names}\NormalTok{(coefs_quant.species) =}\StringTok{ "Species"}

\CommentTok{# row.names(coefs_quant.family.nonest) = "Family"}
\CommentTok{# row.names(coefs_quant.genus.nonest) = "Genus"}
\CommentTok{# row.names(coefs_quant.species.nonest) = "Species"}

\CommentTok{# rename column names }
\CommentTok{#colnames(coefs_quant.family) = c("Slope", "p-value", "R2")}
\KeywordTok{colnames}\NormalTok{(coefs_quant.genus) =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"Slope"}\NormalTok{, }\StringTok{"p-value"}\NormalTok{, }\StringTok{"R2"}\NormalTok{)}
\KeywordTok{colnames}\NormalTok{(coefs_quant.species) =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"Slope"}\NormalTok{, }\StringTok{"p-value"}\NormalTok{, }\StringTok{"R2"}\NormalTok{)}

\CommentTok{# colnames(coefs_quant.family.nonest) = c("Slope", "p-value", "R2")}
\CommentTok{# colnames(coefs_quant.genus.nonest) = c("Slope", "p-value", "R2")}
\CommentTok{# colnames(coefs_quant.species.nonest) = c("Slope", "p-value", "R2")}

\CommentTok{#Merge summary of coefficients into one dataset}
\NormalTok{coefs_summ <-}\StringTok{ }\KeywordTok{rbind}\NormalTok{(coefs_quant.genus, coefs_quant.species)}
\CommentTok{# coefs_summ.nonest <- rbind(coefs_quant.family.nonest, coefs_quant.genus.nonest, coefs_quant.species.nonest)}

\CommentTok{#Display table of summarized coefficients}
\KeywordTok{kable}\NormalTok{(coefs_summ) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|r|r|r}
\hline
  & Slope & p-value & R2\\
\hline
Genus & 0.4559603 & 9e-07 & 0.6201248\\
\hline
Species & 0.4550948 & 4e-07 & 0.6022033\\
\hline
\end{tabular}
\end{table}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# kable(coefs_summ.nonest) %>% }
\CommentTok{#   kable_styling(bootstrap_options = "striped", full_width = F)}
\end{Highlighting}
\end{Shaded}

Pollen grain proportion was highly significantly related to read
proportion across all taxonomic levels, with relatively strong \(R^2\)
values of 0.60 and 0.62.

\begin{itemize}
\tightlist
\item
  this result stayed consistent with the updates in Dec 2020
\end{itemize}

\hypertarget{amplicon-vs.-wgs-quantitative-matching}{%
\subsubsection{Amplicon vs.~WGS: quantitative
matching}\label{amplicon-vs.-wgs-quantitative-matching}}

The basic idea here is to assess if WGS vs.~amplicon data has better
quantitative matching. This is at first glance more challenging than for
the qualitative metrics, in which the directionality of the data tells
the story completely: more true positives = good, and more false
positives = bad. But for the quantitative analyses, having a higher
proportion of matches than there are input proportion of pollen grains
is bad, and having a lower proportion of matches than the input
proportion of pollen grains is also bad.

We used two complementary approaches. The first compares the residuals
of WGS vs.~amplicon to the ``true'' model (slope = 1, intercept = 0).
The second compares the \(R^2\) values of unconstrained linear models
between WGS and amplicon approaches.

\textbf{FIRST} A quantitative approach to this is to analyze the
\emph{residuals} of the input vs.~output quantitative model, and see if
there are differences in the \emph{mean absolute values of the
residuals} between WGS and amplicon approaches. Here, directionality has
a meaning: smaller absolute residual values are better. There are at
least two different approaches one could take: first is just to fit a
linear model to the data (as we have done above); second is to fit a
linear model with intercept = 0 and slope = 1, and use the residuals
from that model.

The latter approach here seems better---ultimately we want the output
proportion of sequence reads to match as closely to possible the input
proportion of pollen grains (i.e.~a linear relationship with intercept =
0 and slope = 1). To find the residuals of this relationship, we don't
need to run a model at all; because the expected value of the output
proportion is the input proportion, we can just calculate the absolute
value of (\texttt{input\ -\ output}) in a new column. Then we can use
that residual value as the response variable in a new linear
mixed-effects model, with the sole fixed effect being the data
\texttt{source} (i.e.~amplicon vs.~WGS), but allowing us to use the
random intercepts we specified in previous models (mix identity and
taxonomic identity).

One thing to note about this approach is that we already know that the
WGS data have more false positives, and that is likely to pull them
further away from the ``true'' line with slope = 1 and intercept = 0.

\textbf{SECOND} Alternatively, another approach is to just directly
compare the WGS and amplicon results in terms of \(R^2\) values for the
linear models. Higher \(R^2\) values are better. This is somewhat
qualitative, however: e.g.~while \(R^2 = 0.77\) is better than
\(R^2 = 0.70\), are those two values meaningfullly different? For
example, it would be easy for a couple of values that were strongly off
from the prediction to greatly alter the \(R^2\) values. But if the
results are strongly different between WGS and amplicon approaches, that
will tell us something.

\hypertarget{data-prep-for-quant-matching-amplicon-vs.-wgs}{%
\paragraph{data prep for quant matching amplicon
vs.~WGS}\label{data-prep-for-quant-matching-amplicon-vs.-wgs}}

We need to do is to get the amplicon data formatted in a similar way to
the WGS data for this analysis (with a percentage of hits for each taxon
truly present in a sample, as well as the proportion of pollen grains of
that taxon in that sample). For this, we will combine our original
\texttt{amp\_all} data with e.g.~\texttt{truepos.krak.species} (and the
corresponding files for genus and species)

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# # first get a `rep.ID` variable into the `amp_all` dataset}
\CommentTok{# # so that there is a parallel with the kraken data}
\CommentTok{# amp_all$rep.ID = 1}

\CommentTok{# pull out only the relevant variables from the kraken data, for easier merging down the line:}
\CommentTok{#quant.krak.family = select(truepos.krak.family, mix.ID, family, pollen.grain.proportion, quant.family, source)}
\NormalTok{quant.krak.genus =}\StringTok{ }\KeywordTok{select}\NormalTok{(truepos.krak.genus, mix.ID, genus, pollen.grain.proportion, quant.genus, source)}
\NormalTok{quant.krak.species =}\StringTok{ }\KeywordTok{select}\NormalTok{(truepos.krak.species, mix.ID, species, pollen.grain.proportion, quant.species, source)}

\CommentTok{# amplicon: pull out relevant variables, for easier `rbind` down the line}
\CommentTok{# rename `sample` to `rep.ID` since both represent sub-samples within a `mix.ID`}
\CommentTok{# ...but `sample` only meaningful for amplicon and `rep.ID` only meaningful for WGS}
\CommentTok{#quant.amp.family.its = select(amp_all, mix.ID, rep.ID = sample, family, pollen.grain.proportion, quant.family = quant.family.ITS, source)}
\CommentTok{#quant.amp.family.rbcl = select(amp_all, mix.ID, rep.ID = sample, family, pollen.grain.proportion, quant.family = quant.family.rbcL, source)}
\NormalTok{quant.amp.genus.its =}\StringTok{ }\KeywordTok{select}\NormalTok{(amp_all, mix.ID, genus, pollen.grain.proportion, }\DataTypeTok{quant.genus =}\NormalTok{ quant.genus.ITS, source)}
\NormalTok{quant.amp.genus.rbcl =}\StringTok{ }\KeywordTok{select}\NormalTok{(amp_all, mix.ID, genus, pollen.grain.proportion, }\DataTypeTok{quant.genus =}\NormalTok{ quant.genus.rbcL, source)}
\NormalTok{quant.amp.species.its =}\StringTok{ }\KeywordTok{select}\NormalTok{(amp_all, mix.ID, species, pollen.grain.proportion, }\DataTypeTok{quant.species =}\NormalTok{ quant.species.ITS, source)}
\NormalTok{quant.amp.species.rbcl =}\StringTok{ }\KeywordTok{select}\NormalTok{(amp_all, mix.ID, species, pollen.grain.proportion, }\DataTypeTok{quant.species =}\NormalTok{ quant.species.rbcL, source)}

\CommentTok{# rbind 'em together: }
\CommentTok{#quant.krakamp.family.its = rbind(quant.amp.family.its, quant.krak.family)}
\CommentTok{#quant.krakamp.family.rbcl = rbind(quant.amp.family.rbcl, quant.krak.family)}
\NormalTok{quant.krakamp.genus.its =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.genus.its, quant.krak.genus)}
\NormalTok{quant.krakamp.genus.rbcl =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.genus.rbcl, quant.krak.genus)}
\NormalTok{quant.krakamp.species.its =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.species.its, quant.krak.species)}
\NormalTok{quant.krakamp.species.rbcl =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.species.rbcl, quant.krak.species)}
\end{Highlighting}
\end{Shaded}

\hypertarget{residual-based-analysis-for-quant-matching-amplicon-vs.-wgs}{%
\paragraph{residual-based analysis for quant matching amplicon
vs.~WGS}\label{residual-based-analysis-for-quant-matching-amplicon-vs.-wgs}}

\begin{itemize}
\tightlist
\item
  calculate residuals (absolute value of difference between input pollen
  grain proportion and output proportion of sequence reads), as a new
  column
\item
  then run a linear mixed-effects model, with \texttt{source} as the
  sole fixed effect
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# calculate residuals}
\CommentTok{#quant.krakamp.family.its = quant.krakamp.family.its %>% mutate(resid = abs(pollen.grain.proportion - quant.family))}
\CommentTok{#quant.krakamp.family.rbcl = quant.krakamp.family.rbcl %>% mutate(resid = abs(pollen.grain.proportion - quant.family))}
\NormalTok{quant.krakamp.genus.its =}\StringTok{ }\NormalTok{quant.krakamp.genus.its }\OperatorTok{%>%}\StringTok{ }\KeywordTok{mutate}\NormalTok{(}\DataTypeTok{resid =} \KeywordTok{abs}\NormalTok{(pollen.grain.proportion }\OperatorTok{-}\StringTok{ }\NormalTok{quant.genus))}
\NormalTok{quant.krakamp.genus.rbcl =}\StringTok{ }\NormalTok{quant.krakamp.genus.rbcl }\OperatorTok{%>%}\StringTok{ }\KeywordTok{mutate}\NormalTok{(}\DataTypeTok{resid =} \KeywordTok{abs}\NormalTok{(pollen.grain.proportion }\OperatorTok{-}\StringTok{ }\NormalTok{quant.genus))}
\NormalTok{quant.krakamp.species.its =}\StringTok{ }\NormalTok{quant.krakamp.species.its }\OperatorTok{%>%}\StringTok{ }\KeywordTok{mutate}\NormalTok{(}\DataTypeTok{resid =} \KeywordTok{abs}\NormalTok{(pollen.grain.proportion }\OperatorTok{-}\StringTok{ }\NormalTok{quant.species))}
\NormalTok{quant.krakamp.species.rbcl =}\StringTok{ }\NormalTok{quant.krakamp.species.rbcl }\OperatorTok{%>%}\StringTok{ }\KeywordTok{mutate}\NormalTok{(}\DataTypeTok{resid =} \KeywordTok{abs}\NormalTok{(pollen.grain.proportion }\OperatorTok{-}\StringTok{ }\NormalTok{quant.species))}

\CommentTok{# run models:}
\CommentTok{#quant.krakamp.family.its.lmer = lmer(resid ~ source + (1|mix.ID/rep.ID) + (1|family), data = quant.krakamp.family.its)}
\CommentTok{#quant.krakamp.family.rbcl.lmer = lmer(resid ~ source + (1|mix.ID/rep.ID) + (1|family), data = quant.krakamp.family.rbcl)}
\NormalTok{quant.krakamp.genus.its.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(resid }\OperatorTok{~}\StringTok{ }\NormalTok{source }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{genus), }\DataTypeTok{data =}\NormalTok{ quant.krakamp.genus.its)}
\NormalTok{quant.krakamp.genus.rbcl.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(resid }\OperatorTok{~}\StringTok{ }\NormalTok{source }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{genus), }\DataTypeTok{data =}\NormalTok{ quant.krakamp.genus.rbcl)}
\NormalTok{quant.krakamp.species.its.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(resid }\OperatorTok{~}\StringTok{ }\NormalTok{source }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{species), }\DataTypeTok{data =}\NormalTok{ quant.krakamp.species.its)}
\NormalTok{quant.krakamp.species.rbcl.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(resid }\OperatorTok{~}\StringTok{ }\NormalTok{source }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{species), }\DataTypeTok{data =}\NormalTok{ quant.krakamp.species.rbcl)}

\CommentTok{# build a table of results, using `broom.mixed`}
\CommentTok{#quant.krakamp.table = tidy(quant.krakamp.family.its.lmer)[2,c(3,4,8)]}
\CommentTok{#quant.krakamp.table = rbind(quant.krakamp.table, tidy(quant.krakamp.family.rbcl.lmer)[2,c(3,4,8)])}
\NormalTok{quant.krakamp.table =}\StringTok{ }\KeywordTok{tidy}\NormalTok{(quant.krakamp.genus.its.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)]}
\NormalTok{quant.krakamp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.krakamp.table, }\KeywordTok{tidy}\NormalTok{(quant.krakamp.genus.rbcl.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}
\NormalTok{quant.krakamp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.krakamp.table, }\KeywordTok{tidy}\NormalTok{(quant.krakamp.species.its.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}
\NormalTok{quant.krakamp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.krakamp.table, }\KeywordTok{tidy}\NormalTok{(quant.krakamp.species.rbcl.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}

\NormalTok{quant.krakamp.table}\OperatorTok{$}\NormalTok{source =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"genus.its"}\NormalTok{, }\StringTok{"genus.rbcl"}\NormalTok{, }\StringTok{"species.its"}\NormalTok{, }\StringTok{"species.rbcl"}\NormalTok{)}

\KeywordTok{kable}\NormalTok{(quant.krakamp.table) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|r|r|l}
\hline
term & estimate & p.value & source\\
\hline
sourcekrak & -0.0272715 & 0.1027510 & genus.its\\
\hline
sourcekrak & -0.0608579 & 0.0252742 & genus.rbcl\\
\hline
sourcekrak & -0.0683416 & 0.0026338 & species.its\\
\hline
sourcekrak & -0.0226632 & 0.4738227 & species.rbcl\\
\hline
\end{tabular}
\end{table}

*updated results Dec 2020**

\begin{itemize}
\tightlist
\item
  \textbf{take-home message:} in contrast to the previous results, we
  now see that for two of the four models, we have a statistically
  significant difference
\item
  in those two models, amplicon data perform better / shotgun data
  perform worse

  \begin{itemize}
  \tightlist
  \item
    previously, amplicon outperformed shotgun for all four
  \item
    directionality of trends is that amplicon does better, only
    significant for 2/4
  \end{itemize}
\item
  interestingly (weirdly?), the two statistically significant results
  are for different markers and different levels of taxonomic resolution
  (!?)---\emph{rbcL} at the genus level, and ITS2 at the species level
\end{itemize}

\hypertarget{linear-model-based-analysis-for-quant-matching-amplicon-vs.-wgs}{%
\paragraph{linear-model based analysis for quant matching amplicon
vs.~WGS}\label{linear-model-based-analysis-for-quant-matching-amplicon-vs.-wgs}}

Run two separate linear mixed-effects models, one for WGS and one for
amplicon. We already did the WGS ones above, so here we just have to do
the ones for amplicon. Then we can ask two different things: 1) which
has a better \(R^2\), i.e.~even if the slope is not 1 and the intercept
is not zero, which one would give us a better relative estimate of
pollen grain proportions; and 2) which one has a slope closer to one
(which is really ultimately an analogue of the residuals analysis).
While it would be sensible to constrain the intercept to be zero in both
cases, again for the WGS I tried this and was not able to get \(R^2\)
estimates.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# run models:}
\CommentTok{#quant.amp.family.its.lmer = lmer(quant.family ~ pollen.grain.proportion + (1|mix.ID) + (1|family), data = quant.amp.family.its)}
\CommentTok{#quant.amp.family.rbcl.lmer = lmer(quant.family ~ pollen.grain.proportion + (1|mix.ID) + (1|family), data = quant.amp.family.rbcl)}
\NormalTok{quant.amp.genus.rbcl.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.genus }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{genus), }\DataTypeTok{data =}\NormalTok{ quant.amp.genus.rbcl)}
\NormalTok{quant.amp.genus.its.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.genus }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{genus), }\DataTypeTok{data =}\NormalTok{ quant.amp.genus.its)}
\NormalTok{quant.amp.species.rbcl.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.species }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{species), }\DataTypeTok{data =}\NormalTok{ quant.amp.species.rbcl)}
\NormalTok{quant.amp.species.its.lmer =}\StringTok{ }\KeywordTok{lmer}\NormalTok{(quant.species }\OperatorTok{~}\StringTok{ }\NormalTok{pollen.grain.proportion }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{mix.ID) }\OperatorTok{+}\StringTok{ }\NormalTok{(}\DecValTok{1}\OperatorTok{|}\NormalTok{species), }\DataTypeTok{data =}\NormalTok{ quant.amp.species.its)}

\CommentTok{# r-squared calculations}
\CommentTok{#r2.family.its.amp = r2beta(quant.amp.family.its.lmer)}
\NormalTok{r2.genus.its.amp =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.amp.genus.its.lmer)}
\NormalTok{r2.species.its.amp =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.amp.species.its.lmer)}
\CommentTok{#r2.family.rbcl.amp = r2beta(quant.amp.family.rbcl.lmer)}
\NormalTok{r2.genus.rbcl.amp =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.amp.genus.rbcl.lmer)}
\NormalTok{r2.species.rbcl.amp =}\StringTok{ }\KeywordTok{r2beta}\NormalTok{(quant.amp.species.rbcl.lmer)}

\CommentTok{# build a table of results, using `broom.mixed`}
\CommentTok{#quant.amp.table = tidy(quant.amp.family.its.lmer)[2,c(3,4,8)]}
\CommentTok{#quant.amp.table = rbind(quant.amp.table, tidy(quant.amp.family.rbcl.lmer)[2,c(3,4,8)])}
\NormalTok{quant.amp.table =}\StringTok{ }\KeywordTok{tidy}\NormalTok{(quant.amp.genus.its.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)]}
\NormalTok{quant.amp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.table, }\KeywordTok{tidy}\NormalTok{(quant.amp.genus.rbcl.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}
\NormalTok{quant.amp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.table, }\KeywordTok{tidy}\NormalTok{(quant.amp.species.its.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}
\NormalTok{quant.amp.table =}\StringTok{ }\KeywordTok{rbind}\NormalTok{(quant.amp.table, }\KeywordTok{tidy}\NormalTok{(quant.amp.species.rbcl.lmer)[}\DecValTok{2}\NormalTok{,}\KeywordTok{c}\NormalTok{(}\DecValTok{3}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{8}\NormalTok{)])}

\CommentTok{# add in source column}
\NormalTok{quant.amp.table}\OperatorTok{$}\NormalTok{source =}\StringTok{ }\KeywordTok{c}\NormalTok{(}\StringTok{"genus.its"}\NormalTok{, }\StringTok{"genus.rbcl"}\NormalTok{, }\StringTok{"species.its"}\NormalTok{, }\StringTok{"species.rbcl"}\NormalTok{)}
\NormalTok{quant.amp.table}\OperatorTok{$}\NormalTok{rsquared =}\StringTok{ }\KeywordTok{c}\NormalTok{(r2.genus.its.amp[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{], r2.genus.rbcl.amp[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{], r2.species.its.amp[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{], r2.species.rbcl.amp[}\DecValTok{2}\NormalTok{,}\DecValTok{6}\NormalTok{])}

\CommentTok{# show off the table!}
\KeywordTok{kable}\NormalTok{(quant.amp.table) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{kable_styling}\NormalTok{(}\DataTypeTok{bootstrap_options =} \StringTok{"striped"}\NormalTok{, }\DataTypeTok{full_width =}\NormalTok{ F)}
\end{Highlighting}
\end{Shaded}

\begin{table}[H]
\centering
\begin{tabular}{l|r|r|l|r}
\hline
term & estimate & p.value & source & rsquared\\
\hline
pollen.grain.proportion & 0.4899757 & 0 & genus.its & 0.2562560\\
\hline
pollen.grain.proportion & 0.5463068 & 0 & genus.rbcl & 0.3973826\\
\hline
pollen.grain.proportion & 0.2421154 & 0 & species.its & 0.0897513\\
\hline
pollen.grain.proportion & 0.3901673 & 0 & species.rbcl & 0.2959296\\
\hline
\end{tabular}
\end{table}

\textbf{update Dec 2020: these results did not change at all}

Comparing this table to the table from the WGS only quantitative
matching, we can see that the \(R^2\) values are are substantially
higher for WGS / kraken as compared to the amplicon data. Consistently
with the residual-based analysis, the slopes of the relationships for
the amplicon data are much larger / much closer to one than for the WGS
data.

This is consistent with:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  WGS being more quantitative (higher \(R^2\)---better explanatory power
  in a linear relationship)
\item
  WGS having lower slopes, and thus higher residuals from a relationship
  where intercept = 0 and slope = 1, which could be driven in large part
  by the high rate of false positives in the WGS data (which greatly
  limit quantitative matching)
\end{enumerate}

\hypertarget{figures}{%
\section{Figures}\label{figures}}

The numbering of figures in this section is not ultimately parallel to
their numbering in the manuscript; we left it in the order they were
developed because the naming in the code would have to change, which
would be a substantial amount of work for relatively cosmetic purposes.
Figures added later in the process are denoted with alphabetic rather
than numeric names; e.g., X, Y, and Z for false negatives and I, J, and
K for false positives.

\hypertarget{figure-1-true-positives-by-taxon}{%
\subsection{Figure 1: true positives by
taxon}\label{figure-1-true-positives-by-taxon}}

This figure shows the mean proportion and binomial CI of correct (true
positive) taxonomic matches by species and genus. Fig. 1a is by family,
Fig. 1b is by genus, and Fig. 1c is by species.

The binomial CI is calculated using the binom.confint function, which
requires a vector of the number of success (\(x\)) and a vector of the
number of independent trials (\(n\)).

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Create variable "presence" to get total number of independent tests}
\CommentTok{#truepos.krak.family$presence = 1}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{presence =}\StringTok{ }\DecValTok{1}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{presence =}\StringTok{ }\DecValTok{1}


\CommentTok{#Total the number of taxa correctly identified (x) when grouped by taxa}
\CommentTok{#family_x = truepos.krak.family %>%}
 \CommentTok{# group_by(family) %>%}
  \CommentTok{#summarise(x = sum(qual.family))}

\NormalTok{genus_x =}\StringTok{ }\NormalTok{truepos.krak.genus }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(genus) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarise}\NormalTok{(}\DataTypeTok{x =} \KeywordTok{sum}\NormalTok{(qual.genus))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{species_x =}\StringTok{ }\NormalTok{truepos.krak.species }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(species) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarise}\NormalTok{(}\DataTypeTok{x =} \KeywordTok{sum}\NormalTok{(qual.species))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Get the total number of independent tests (n) when grouped by taxa}
\CommentTok{#family_n = truepos.krak.family %>% }
 \CommentTok{# group_by(family) %>%}
\CommentTok{#  summarise(n = sum(presence))}

\NormalTok{genus_n =}\StringTok{ }\NormalTok{truepos.krak.genus }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{group_by}\NormalTok{(genus) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarise}\NormalTok{(}\DataTypeTok{n =} \KeywordTok{sum}\NormalTok{(presence))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{species_n =}\StringTok{ }\NormalTok{truepos.krak.species }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{group_by}\NormalTok{(species) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarise}\NormalTok{(}\DataTypeTok{n =} \KeywordTok{sum}\NormalTok{(presence))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Calculate binomial confidence interval using x and n}
\CommentTok{#binomci_family = binom.confint(family_x$x, family_n$n, method="exact")}
\NormalTok{binomci_species =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(species_x}\OperatorTok{$}\NormalTok{x, species_n}\OperatorTok{$}\NormalTok{n, }\DataTypeTok{method=}\StringTok{"exact"}\NormalTok{)}
\NormalTok{binomci_genus =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(genus_x}\OperatorTok{$}\NormalTok{x, genus_n}\OperatorTok{$}\NormalTok{n, }\DataTypeTok{method=}\StringTok{"exact"}\NormalTok{)}

\CommentTok{#Add the taxa names, rename column}
\CommentTok{#binomci_family = cbind(family_x$family, binomci_family)}
\NormalTok{binomci_genus =}\StringTok{ }\KeywordTok{cbind}\NormalTok{(genus_x}\OperatorTok{$}\NormalTok{genus, binomci_genus)}
\NormalTok{binomci_species =}\StringTok{ }\KeywordTok{cbind}\NormalTok{(species_x}\OperatorTok{$}\NormalTok{species, binomci_species)}

\CommentTok{#colnames(binomci_family)[1] = "family"}
\KeywordTok{colnames}\NormalTok{(binomci_genus)[}\DecValTok{1}\NormalTok{] =}\StringTok{ "genus"}
\KeywordTok{colnames}\NormalTok{(binomci_species)[}\DecValTok{1}\NormalTok{] =}\StringTok{ "species"}


\CommentTok{#Figure 1A: Family level}
\CommentTok{#fig1a=binomci_family%>%}
\CommentTok{#  ggplot(aes(family,mean))+}
\CommentTok{#  geom_point(position=position_dodge(width=0.3), size=4, alpha=0.6)+}
\CommentTok{#  geom_errorbar(width=0.6, aes(family,ymin=lower, ymax=upper), alpha=0.6, position=position_dodge(width=0.3))+}
\CommentTok{#  xlab("family")+}
\CommentTok{#  ylab("proportion of correct matches")+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw()}
\CommentTok{# fig1a = fig1a + theme(axis.text.x = element_text(angle = 45, hjust = 1))}

\CommentTok{#Figure 1B: Genus level}
\NormalTok{fig1b=binomci_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(genus,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\DataTypeTok{position=}\KeywordTok{position_dodge}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.3}\NormalTok{), }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.6}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.6}\NormalTok{, }\KeywordTok{aes}\NormalTok{(genus,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{alpha=}\FloatTok{0.6}\NormalTok{, }\DataTypeTok{position=}\KeywordTok{position_dodge}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.3}\NormalTok{))}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"genus"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{fig1b =}\StringTok{ }\NormalTok{fig1b }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Figure 1C: Species level}
\NormalTok{fig1c=binomci_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(species,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\DataTypeTok{position=}\KeywordTok{position_dodge}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.3}\NormalTok{), }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.6}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.6}\NormalTok{, }\KeywordTok{aes}\NormalTok{(species,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{alpha=}\FloatTok{0.6}\NormalTok{, }\DataTypeTok{position=}\KeywordTok{position_dodge}\NormalTok{(}\DataTypeTok{width=}\FloatTok{0.3}\NormalTok{))}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"species"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{fig1c =}\StringTok{ }\NormalTok{fig1c }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

  \CommentTok{#Figure 1 panel}

\CommentTok{#suppressMessages(ggsave("WGS_fig1a.pdf", plot=fig1a, device="pdf", height=5, width=7, units="in"))}
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig1b.pdf"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig1b, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{width =} \DecValTok{7}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig1c.pdf"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig1c, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{width =} \DecValTok{7}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\NormalTok{panel.fig1 =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig1a, }
\NormalTok{  fig1b,fig1c, }\DataTypeTok{ncol=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig1_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{panel.fig1, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{height=}\DecValTok{11}\NormalTok{, }\DataTypeTok{width =} \FloatTok{8.5}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{fig1_combined.jpg}
\caption{Figure 1}
\end{figure}

\hypertarget{figure-2-true-positives-vs.-species-richness}{%
\subsection{Figure 2: true positives vs.~species
richness}\label{figure-2-true-positives-vs.-species-richness}}

This figure shows the proportion of correct taxonomic matches by species
richness for \emph{rbcL} and ITS2 at the species, genus, and family
levels of matching. To correctly format the data, the mean and binomial
CI of correct taxonomic matches needs to be summarized 3 times: once for
sample, once for mix, and once for level of species richness. Otherwise
samples with multiple species will get over-represented in terms of the
overall means. The integers are tracked at each step and summarized with
\texttt{dplyr}. The data are summarized with \texttt{dplyr} twice, once
per mix and then once per sample. Then within each level of the factor
of interest, the mean and binomial confidence intervals are calculated
using \texttt{binom.confint} from the \texttt{binom} package.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Summarize the agg dataset first by species richness, mix.ID, and sample. Also calculate pool size for each sample ('n()'), as well as sum # of positive identifications}

\CommentTok{#CI.family.1 = truepos.krak.family %>% }
\CommentTok{#select(mix.ID, sample, spp.rich, qual.family, question.1, question.2, question.3) %>% group_by(spp.rich, mix.ID, sample) %>%}
\CommentTok{#summarize(pool.size = n(),}
\CommentTok{#  family = sum(qual.family),}
\CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
\CommentTok{#  question.1 = mean(question.1),}
\CommentTok{#  question.2 = mean(question.2),}
\CommentTok{#  question.3 = mean(question.3))}

\NormalTok{CI.genus}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.genus }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, sample, spp.rich, qual.genus, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich, mix.ID, sample) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(qual.genus),}
\CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
  \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
  \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
  \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.species}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.species }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, sample, spp.rich, qual.species, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich, mix.ID, sample) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(qual.species),}
\CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
  \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
  \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
  \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# second step: group_by(spp.rich, mix.ID, pool.size)}

\CommentTok{#CI.family.2 = CI.family.1 %>% group_by(spp.rich, mix.ID, pool.size) %>% }
\CommentTok{#  summarize(pool.num = n(),}
\CommentTok{#  family = sum(family), }
\CommentTok{#  question.1 = mean(question.1),}
\CommentTok{#  question.2 = mean(question.2),}
\CommentTok{#  question.3 = mean(question.3)) # %>% }
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\CommentTok{#  CI.family.2$n2 = CI.family.2$pool.size * CI.family.2$pool.num}
  
\NormalTok{CI.genus}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(genus), }
  \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
  \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
  \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{  CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}

\NormalTok{CI.species}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(species), }
  \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
  \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
  \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{  CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}
  
\CommentTok{#Format variables to be factors.  Restrict to mixtures where questions are equal to 1. **Not sure why this is done?  }
  
\CommentTok{#dat.family <- CI.family.2}
\CommentTok{#dat.family$question.1=as.factor(dat.family$question.1)}
\CommentTok{#dat.family$question.2=as.factor(dat.family$question.2)}
\CommentTok{#dat.family$question.3=as.factor(dat.family$question.3)}
\CommentTok{#dat.family$spp.rich=as.factor(dat.family$spp.rich)}
\CommentTok{#dat.family=filter(dat.family,question.1=="1"|question.2=="1"|question.3=="1")}
\CommentTok{#fig2_family=dat.family%>%}
\CommentTok{#  select(spp.rich, mix.ID, n2, family)}

\NormalTok{dat.genus <-}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{spp.rich=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{spp.rich)}
\CommentTok{#dat.genus=filter(dat.genus,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{fig2_genus=dat.genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(spp.rich, mix.ID, n2, genus)}

\NormalTok{dat.species <-}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{spp.rich=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{spp.rich)}
\CommentTok{#dat.species=filter(dat.species,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{fig2_species=dat.species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(spp.rich, mix.ID, n2, species)}

\CommentTok{#Convert from wide to long}

\CommentTok{#fig2_long_family = melt(fig2_family, id.vars = c("spp.rich","mix.ID", "n2"))}
\NormalTok{fig2_long_genus =}\StringTok{ }\KeywordTok{melt}\NormalTok{(fig2_genus, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"spp.rich"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(fig2_genus, id.vars = c("spp.rich", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(fig2_genus). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fig2_long_species =}\StringTok{ }\KeywordTok{melt}\NormalTok{(fig2_species, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"spp.rich"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(fig2_species, id.vars = c("spp.rich", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(fig2_species). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Subset to species richness}

\CommentTok{#spp.rich_family = fig2_long_family%>%}
\CommentTok{#  filter(substr(variable,1,7)=="family")%>%}
\CommentTok{#  rename(x=value, n=n2)}
\NormalTok{spp.rich_genus =}\StringTok{ }\NormalTok{fig2_long_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}
\NormalTok{spp.rich_species =}\StringTok{ }\NormalTok{fig2_long_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}

\CommentTok{#group by species richness and marker so that all mixtures with richness = 2 are grouped together}

\CommentTok{#spp.rich_family = spp.rich_family %>%}
\CommentTok{#  group_by(spp.rich, variable) %>%}
\CommentTok{#  summarize(n=mean(n),x=mean(x)) %>%}
\CommentTok{#  ungroup()}

\NormalTok{spp.rich_genus =}\StringTok{ }\NormalTok{spp.rich_genus }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(spp.rich, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{spp.rich_species =}\StringTok{ }\NormalTok{spp.rich_species }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(spp.rich, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'spp.rich' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Calculate mean and binomial CI}

\CommentTok{#bin.family = binom.confint(spp.rich_family$x, spp.rich_family$n,methods="exact") %>%}
\CommentTok{#  select(-method)}

\NormalTok{bin.genus =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(spp.rich_genus}\OperatorTok{$}\NormalTok{x, spp.rich_genus}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}

\NormalTok{bin.species =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(spp.rich_species}\OperatorTok{$}\NormalTok{x, spp.rich_species}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}

\CommentTok{#Merge mean and binomial CI with mixture info}

\CommentTok{#family.all = merge(bin.family, spp.rich_family, by = c("x", "n"))}
\CommentTok{#family.all = unique(family.all)}

\NormalTok{genus.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.genus, spp.rich_genus, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{genus.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(genus.all)}

\NormalTok{species.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.species, spp.rich_species, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{species.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(species.all)}

\CommentTok{#Figure 2A: Family level}
\CommentTok{#fig2a=family.all%>%}
\CommentTok{#  filter(variable=="family")%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(spp.rich,mean))+}
\CommentTok{#  geom_errorbar(aes(spp.rich,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("species richness")+}
\CommentTok{#  ylab("proportion of correct matches")+}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches")+}
\CommentTok{#   theme_bw()}

\CommentTok{#Figure 2B: Genus level}
\NormalTok{fig2b=genus.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"species richness"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{   }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure 2A: Species level}
\NormalTok{fig2c=species.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"species richness"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{   }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Create panel for Figure 2}
\NormalTok{fig2=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig2a,}
\NormalTok{  fig2b,fig2c, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig2_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig2, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{fig2_combined.jpg}
\caption{Figure 2}
\end{figure}

\hypertarget{figure-x-true-positives-vs.-relatedness}{%
\subsection{Figure X: true positives
vs.~relatedness}\label{figure-x-true-positives-vs.-relatedness}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Summarize the agg dataset first by relatedness, mix.ID, and sample. Also calculate pool size for each sample ('n()'), as well as sum # of positive identifications}

\CommentTok{#CI.family.1 = truepos.krak.family %>% }
\CommentTok{#  select(mix.ID, sample, relatedness, qual.family, question.1, question.2, question.3) %>% group_by(relatedness, mix.ID, sample) %>%}
\CommentTok{#  summarize(pool.size = n(),}
\CommentTok{#            family = sum(qual.family),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
\CommentTok{#            question.1 = mean(question.1),}
\CommentTok{#            question.2 = mean(question.2),}
\CommentTok{#            question.3 = mean(question.3))}

\NormalTok{CI.genus}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.genus }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, sample, relatedness, qual.genus, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness, mix.ID, sample) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(qual.genus),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.species}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.species }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, sample, relatedness, qual.species, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness, mix.ID, sample) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(qual.species),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# second step: group_by(relatedness, mix.ID, pool.size)}

\CommentTok{#CI.family.2 = CI.family.1 %>% group_by(relatedness, mix.ID, pool.size) %>% }
\CommentTok{#  summarize(pool.num = n(),}
\CommentTok{#            family = sum(family), }
\CommentTok{#            question.1 = mean(question.1),}
\CommentTok{#            question.2 = mean(question.2),}
\CommentTok{#            question.3 = mean(question.3)) # %>% }

\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{CI.family}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.family}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.family}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Error in eval(expr, envir, enclos): object 'CI.family.2' not found
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.genus}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(genus), }
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}

\NormalTok{CI.species}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(species), }
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}

\CommentTok{#Format variables to be factors.  Restrict to mixtures where questions are equal to 1. **Not sure why this is done?  }

\CommentTok{#dat.family <- CI.family.2}
\CommentTok{#dat.family$question.1=as.factor(dat.family$question.1)}
\CommentTok{#dat.family$question.2=as.factor(dat.family$question.2)}
\CommentTok{#dat.family$question.3=as.factor(dat.family$question.3)}
\CommentTok{#dat.family$relatedness=as.factor(dat.family$relatedness)}
\CommentTok{#dat.family=filter(dat.family,question.1=="1"|question.2=="1"|question.3=="1")}
\CommentTok{#figX_family=dat.family%>%}
\CommentTok{#  select(relatedness, mix.ID, n2, family)}

\NormalTok{dat.genus <-}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{relatedness=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{relatedness)}
\CommentTok{#dat.genus=filter(dat.genus,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{figX_genus=dat.genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(relatedness, mix.ID, n2, genus)}

\NormalTok{dat.species <-}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{relatedness=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{relatedness)}
\CommentTok{#dat.species=filter(dat.species,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{figX_species=dat.species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(relatedness, mix.ID, n2, species)}

\CommentTok{#Convert from wide to long}
\CommentTok{#figX_long_family = melt(figX_family, id.vars = c("relatedness","mix.ID", "n2"))}
\NormalTok{figX_long_genus =}\StringTok{ }\KeywordTok{melt}\NormalTok{(figX_genus, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(figX_genus, id.vars = c("relatedness", "mix.ID", "n2")):
## The melt generic in data.table has been passed a grouped_df and will attempt
## to redirect to the relevant reshape2 method; please note that reshape2 is
## deprecated, and this redirection is now deprecated as well. To continue using
## melt methods from reshape2 while both libraries are attached, e.g. melt.list,
## you can prepend the namespace like reshape2::melt(figX_genus). In the next
## version, this warning will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{figX_long_species =}\StringTok{ }\KeywordTok{melt}\NormalTok{(figX_species, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(figX_species, id.vars = c("relatedness", "mix.ID", "n2")):
## The melt generic in data.table has been passed a grouped_df and will attempt
## to redirect to the relevant reshape2 method; please note that reshape2 is
## deprecated, and this redirection is now deprecated as well. To continue using
## melt methods from reshape2 while both libraries are attached, e.g. melt.list,
## you can prepend the namespace like reshape2::melt(figX_species). In the next
## version, this warning will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Subset to relatedness}
\CommentTok{#relatedness_family = figX_long_family%>%}
\CommentTok{#  filter(substr(variable,1,7)=="family")%>%}
\CommentTok{#  rename(x=value, n=n2)}
\NormalTok{relatedness_genus =}\StringTok{ }\NormalTok{figX_long_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}
\NormalTok{relatedness_species =}\StringTok{ }\NormalTok{figX_long_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}

\CommentTok{#group by relatedness and marker}
\CommentTok{#relatedness_family = relatedness_family %>%}
\CommentTok{#  group_by(relatedness, variable) %>%}
\CommentTok{#  summarize(n=mean(n),x=mean(x)) %>%}
\CommentTok{#  ungroup()}

\NormalTok{relatedness_genus =}\StringTok{ }\NormalTok{relatedness_genus }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(relatedness, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{relatedness_species =}\StringTok{ }\NormalTok{relatedness_species }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(relatedness, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'relatedness' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Calculate mean and binomial CI}
\CommentTok{#bin.family = binom.confint(relatedness_family$x, relatedness_family$n,methods="exact") %>%}
  \CommentTok{#select(-method)}

\CommentTok{#bin.genus = binom.confint(relatedness_genus$x, relatedness_genus$n,methods="exact") %>%}
  \CommentTok{#select(-method)}

\CommentTok{#bin.species = binom.confint(relatedness_species$x, relatedness_species$n,methods="exact") %>%}
  \CommentTok{#select(-method)}

\CommentTok{#Merge mean and binomial CI with mixture info}
\CommentTok{#family.all = merge(bin.family, relatedness_family, by = c("x", "n"))}
\CommentTok{#family.all = unique(family.all)}

\NormalTok{genus.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.genus, relatedness_genus, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{genus.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(genus.all)}

\NormalTok{species.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.species, relatedness_species, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{species.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(species.all)}

\CommentTok{#Figure XA: Family level}
\CommentTok{#figXa=family.all%>%}
\CommentTok{#  filter(variable=="family")%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(relatedness,mean))+}
\CommentTok{#  geom_errorbar(aes(relatedness,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("relatedness")+}
\CommentTok{#  ylab("proportion of correct matches")+}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  scale_x_discrete(labels=c("same genus", "same family", "same order", "dif. orders")) +}
\CommentTok{#  theme_bw()}
\CommentTok{# figXa = figXa + theme(axis.text.x = element_text(angle = 45, hjust = 1))}

\CommentTok{#Figure XB: Genus level}
\NormalTok{figXb=genus.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_x_discrete}\NormalTok{(}\DataTypeTok{labels=}\KeywordTok{c}\NormalTok{(}\StringTok{"same genus"}\NormalTok{, }\StringTok{"same family"}\NormalTok{, }\StringTok{"same order"}\NormalTok{, }\StringTok{"dif. orders"}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figXb =}\StringTok{ }\NormalTok{figXb }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Figure XA: Species level}
\NormalTok{figXc=species.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_x_discrete}\NormalTok{(}\DataTypeTok{labels=}\KeywordTok{c}\NormalTok{(}\StringTok{"same genus"}\NormalTok{, }\StringTok{"same family"}\NormalTok{, }\StringTok{"same order"}\NormalTok{, }\StringTok{"dif. orders"}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figXc =}\StringTok{ }\NormalTok{figXc }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Create panel for Figure X}
\NormalTok{figX=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figXa,}
\NormalTok{  figXb,figXc, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figX_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figX, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figX_combined.jpg}
\caption{Figure X}
\end{figure}

\hypertarget{figure-y-true-positives-vs.-rarity-binned}{%
\subsection{Figure Y: true positives vs.~rarity
(binned)}\label{figure-y-true-positives-vs.-rarity-binned}}

The first step here is to create a categorical \texttt{rarity} variable
based on binning \texttt{pollen.grain.proportion} into four equal bins
(1-25\%; 26-50\%; 51-75\%; 76-100\%). We do that using the \texttt{cut}
function. By creating a categorical variable, we can produce a figure
that is highly parallel to those for the other two facets of sample
complexity (species richness and relatedness)

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# this figure mirrors the other two questions (species richness and rarity) but}
\CommentTok{# rarity (input pollen grain proportion). Since rarity is continuous and the others are}
\CommentTok{# categorical we will 'bin' the rarity metrics into 4 bins:}
\CommentTok{# 1-25%; 26-50%; 51-75%; 76-100%}

\CommentTok{# use the `cut` function to create the `rarity` variable by binning `pollen.grain.proportion`:}
\CommentTok{#truepos.krak.family$rarity = cut(truepos.krak.family$pollen.grain.proportion, breaks = c(0, 0.25, 0.5, 0.75, 1), }
\CommentTok{#      labels = c("<25%", "25-50%", "50-75%", ">75%"))}
\NormalTok{truepos.krak.genus}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{cut}\NormalTok{(truepos.krak.genus}\OperatorTok{$}\NormalTok{pollen.grain.proportion, }\DataTypeTok{breaks =} \KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\FloatTok{0.25}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\FloatTok{0.75}\NormalTok{, }\DecValTok{1}\NormalTok{), }
      \DataTypeTok{labels =} \KeywordTok{c}\NormalTok{(}\StringTok{"<25%"}\NormalTok{, }\StringTok{"25-50%"}\NormalTok{, }\StringTok{"50-75%"}\NormalTok{, }\StringTok{">75%"}\NormalTok{))}
\NormalTok{truepos.krak.species}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{cut}\NormalTok{(truepos.krak.species}\OperatorTok{$}\NormalTok{pollen.grain.proportion, }\DataTypeTok{breaks =} \KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\FloatTok{0.25}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\FloatTok{0.75}\NormalTok{, }\DecValTok{1}\NormalTok{), }
      \DataTypeTok{labels =} \KeywordTok{c}\NormalTok{(}\StringTok{"<25%"}\NormalTok{, }\StringTok{"25-50%"}\NormalTok{, }\StringTok{"50-75%"}\NormalTok{, }\StringTok{">75%"}\NormalTok{))}

\CommentTok{#Summarize the agg dataset first by rarity, mix.ID, and sample. Also calculate pool size for each sample ('n()'), as well as sum # of positive identifications}

\CommentTok{#CI.family.1 = truepos.krak.family %>% }
\CommentTok{#  select(mix.ID, sample, rarity, qual.family, question.1, question.2, question.3) %>% group_by(rarity, mix.ID, sample) %>%}
\CommentTok{#  summarize(pool.size = n(),}
\CommentTok{#            family = sum(qual.family),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
\CommentTok{#            question.1 = mean(question.1),}
\CommentTok{#            question.2 = mean(question.2),}
\CommentTok{#            question.3 = mean(question.3))}

\NormalTok{CI.genus}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.genus }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, sample, rarity, qual.genus, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity, mix.ID, sample) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(qual.genus),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.species}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{truepos.krak.species }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, sample, rarity, qual.species, question}\FloatTok{.1}\NormalTok{, question}\FloatTok{.2}\NormalTok{, question}\FloatTok{.3}\NormalTok{) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity, mix.ID, sample) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(qual.species),}
            \CommentTok{# add question.1, question.2, question.3 because those equal to 1 are filtered out in Jamie's figure 3 code}
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# second step: group_by(rarity, mix.ID, pool.size)}

\CommentTok{#CI.family.2 = CI.family.1 %>% group_by(rarity, mix.ID, pool.size) %>% }
\CommentTok{#  summarize(pool.num = n(),}
\CommentTok{#            family = sum(family), }
\CommentTok{#            question.1 = mean(question.1),}
\CommentTok{#            question.2 = mean(question.2),}
\CommentTok{#            question.3 = mean(question.3)) # %>% }

\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\CommentTok{#CI.family.2$n2 = CI.family.2$pool.size * CI.family.2$pool.num}

\NormalTok{CI.genus}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(genus), }
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}

\NormalTok{CI.species}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
            \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(species), }
            \DataTypeTok{question.1 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.1}\NormalTok{),}
            \DataTypeTok{question.2 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.2}\NormalTok{),}
            \DataTypeTok{question.3 =} \KeywordTok{mean}\NormalTok{(question}\FloatTok{.3}\NormalTok{)) }\CommentTok{# %>% }
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}

\CommentTok{#Format variables to be factors.  Restrict to mixtures where questions are equal to 1. **Not sure why this is done?  }

\CommentTok{#dat.family <- CI.family.2}
\CommentTok{#dat.family$question.1=as.factor(dat.family$question.1)}
\CommentTok{#dat.family$question.2=as.factor(dat.family$question.2)}
\CommentTok{#dat.family$question.3=as.factor(dat.family$question.3)}
\CommentTok{#dat.family$rarity=as.factor(dat.family$rarity)}
\CommentTok{#dat.family=filter(dat.family,question.1=="1"|question.2=="1"|question.3=="1")}
\CommentTok{#figY_family=dat.family%>%}
\CommentTok{#  select(rarity, mix.ID, n2, family)}

\NormalTok{dat.genus <-}\StringTok{ }\NormalTok{CI.genus}\FloatTok{.2}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.genus}\OperatorTok{$}\NormalTok{rarity=}\KeywordTok{as.factor}\NormalTok{(dat.genus}\OperatorTok{$}\NormalTok{rarity)}
\CommentTok{#dat.genus=filter(dat.genus,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{figY_genus=dat.genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(rarity, mix.ID, n2, genus)}

\NormalTok{dat.species <-}\StringTok{ }\NormalTok{CI.species}\FloatTok{.2}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.1}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.2}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{question}\FloatTok{.3}\NormalTok{)}
\NormalTok{dat.species}\OperatorTok{$}\NormalTok{rarity=}\KeywordTok{as.factor}\NormalTok{(dat.species}\OperatorTok{$}\NormalTok{rarity)}
\CommentTok{#dat.species=filter(dat.species,question.1=="1"|question.2=="1"|question.3=="1")}
\NormalTok{figY_species=dat.species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(rarity, mix.ID, n2, species)}

\CommentTok{#Convert from wide to long}

\CommentTok{#figY_long_family = melt(figY_family, id.vars = c("rarity","mix.ID", "n2"))}
\NormalTok{figY_long_genus =}\StringTok{ }\KeywordTok{melt}\NormalTok{(figY_genus, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"rarity"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(figY_genus, id.vars = c("rarity", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(figY_genus). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{figY_long_species =}\StringTok{ }\KeywordTok{melt}\NormalTok{(figY_species, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"rarity"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(figY_species, id.vars = c("rarity", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(figY_species). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Subset to rarity}

\CommentTok{#rarity_family = figY_long_family%>%}
\CommentTok{#  filter(substr(variable,1,7)=="family")%>%}
\CommentTok{#  rename(x=value, n=n2)}
\NormalTok{rarity_genus =}\StringTok{ }\NormalTok{figY_long_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}
\NormalTok{rarity_species =}\StringTok{ }\NormalTok{figY_long_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(}\KeywordTok{substr}\NormalTok{(variable,}\DecValTok{1}\NormalTok{,}\DecValTok{7}\NormalTok{)}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}

\CommentTok{#group by rarity and marker so that all mixtures with richness = 2 are grouped together}

\CommentTok{#rarity_family = rarity_family %>%}
\CommentTok{#  group_by(rarity, variable) %>%}
\CommentTok{#  summarize(n=mean(n),x=mean(x)) %>%}
\CommentTok{#  ungroup()}

\NormalTok{rarity_genus =}\StringTok{ }\NormalTok{rarity_genus }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(rarity, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{rarity_species =}\StringTok{ }\NormalTok{rarity_species }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(rarity, variable) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'rarity' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Calculate mean and binomial CI}

\CommentTok{#bin.family = binom.confint(rarity_family$x, rarity_family$n,methods="exact") %>%}
\CommentTok{#  select(-method)}

\NormalTok{bin.genus =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(rarity_genus}\OperatorTok{$}\NormalTok{x, rarity_genus}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}

\NormalTok{bin.species =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(rarity_species}\OperatorTok{$}\NormalTok{x, rarity_species}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}

\CommentTok{#Merge mean and binomial CI with mixture info}

\CommentTok{#family.all = merge(bin.family, rarity_family, by = c("x", "n"))}
\CommentTok{#family.all = unique(family.all)}

\NormalTok{genus.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.genus, rarity_genus, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{genus.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(genus.all)}

\NormalTok{species.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(bin.species, rarity_species, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{species.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(species.all)}

\CommentTok{#Figure YA: Family level}
\CommentTok{#figYa=family.all%>%}
\CommentTok{#  filter(variable=="family")%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(rarity,mean))+}
\CommentTok{#  geom_errorbar(aes(rarity,ymin=lower, ymax=upper), width=0.2) +}
\CommentTok{#  xlab("rarity") +}
\CommentTok{#  ylab("proportion of correct matches") +}
\CommentTok{#  ylim(0,1) +}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw()}
\CommentTok{#figYa = figYa + theme(axis.text.x = element_text(angle = 45, hjust = 1))}

\CommentTok{#Figure YB: Genus level}
\NormalTok{figYb=genus.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"genus"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"rarity"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figYb =}\StringTok{ }\NormalTok{figYb }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Figure YC: Species level}
\NormalTok{figYc=species.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{filter}\NormalTok{(variable}\OperatorTok{==}\StringTok{"species"}\NormalTok{)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"rarity"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figYc =}\StringTok{ }\NormalTok{figYc }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Create panel for Figure Y}
\NormalTok{figY=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figYa,}
\NormalTok{  figYb,figYc, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figY_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figY, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figY_combined.jpg}
\caption{Figure Y}
\end{figure}

\hypertarget{figure-s1-true-positives-combined}{%
\subsection{Figure S1: true positives
combined}\label{figure-s1-true-positives-combined}}

Here we are putting together a figure comprised of panels we have
already built in the sections for Figs 2, X, and Y. This figure has 9
panels, with a row for each of the 3 facets of sample complexity and a
column for each of the 3 levels of taxonomic matching. Very
straightforward.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# re-label panels: }
\CommentTok{#figXa = figXa + labs(title = NULL, subtitle = "D: Family matches")}
\NormalTok{figXb =}\StringTok{ }\NormalTok{figXb }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"E: Genus matches"}\NormalTok{)}
\NormalTok{figXc =}\StringTok{ }\NormalTok{figXc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"F: Species matches"}\NormalTok{)}
\CommentTok{#figYa = figYa + labs(title = NULL, subtitle = "H: Family matches")}
\NormalTok{figYb =}\StringTok{ }\NormalTok{figYb }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"I: Genus matches"}\NormalTok{)}
\NormalTok{figYc =}\StringTok{ }\NormalTok{figYc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"J: Species matches"}\NormalTok{)}

\CommentTok{#Create panel for Figure S1}
\NormalTok{figS1 =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig2a, }
\NormalTok{  fig2b, fig2c, }\CommentTok{#figXa, }
\NormalTok{  figXb, figXc, }\CommentTok{#figYa, }
\NormalTok{  figYb, figYc, }\DataTypeTok{nrow=}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figS1_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figS1, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{height =} \DecValTok{250}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figS1_combined.jpg}
\caption{Figure S1}
\end{figure}

\hypertarget{figure-z-true-positives-by-all-3-complexity-variables-species-level-only}{%
\subsection{Figure Z: true positives by all 3 complexity variables;
species level
only}\label{figure-z-true-positives-by-all-3-complexity-variables-species-level-only}}

Here we are putting together a figure comprised of panels we have
already built in the sections for Figs 2, X, and Y. One relatively minor
tweak is altering the figure header for each panel. Currently they all
say ``species'' because they were pooled into figures with varying
levels of taxonomic matching. We need to instead change the headers to
reflect the explanatory variable (element of sample complexity) that
each panel is focused on.

This code chunk comes \emph{after} the Supplemental Figure so that we
change those headers only after the previous headers are used as they
are supposed to be (in the figure where each panel needs to be
identified by its level of taxonomic matching).

A second minor change is that because the \(x\)-axis labels are
different for each panel, with some of them being longer, the panels
themselves end up being different heights. To get around this, setting
the orientation of this plot as vertical rather than horizontal. This
may also work better in a journal format with the figure being easier to
compare to the text to the left of it.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# change panel headers}
\NormalTok{fig2c =}\StringTok{ }\NormalTok{fig2c }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"A: Species Richness"}\NormalTok{)}
\NormalTok{figXc =}\StringTok{ }\NormalTok{figXc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Relatedness"}\NormalTok{)}
\NormalTok{figYc =}\StringTok{ }\NormalTok{figYc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Rarity"}\NormalTok{)}

\CommentTok{#Create panel for Figure Z}
\NormalTok{figZ =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(fig2c,figXc,figYc, }\DataTypeTok{nrow=}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figZ_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figZ, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\NormalTok{(}\DecValTok{169}\OperatorTok{/}\DecValTok{3}\NormalTok{), }\DataTypeTok{height =} \DecValTok{250}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figZ_combined.jpg}
\caption{Figure Z}
\end{figure}

\hypertarget{figure-3-true-positives-vs.-rarity-by-taxon}{%
\subsection{Figure 3: true positives vs.~rarity by
taxon}\label{figure-3-true-positives-vs.-rarity-by-taxon}}

\textbf{don't use this figure---instead use fig Y}

This figure shows the relationship between the proportion of pollen
grains belonging to a particular taxon with the probability of detection
(presence/absence) of that taxon in the sequencing reads. Each color
belongs to a particular taxon, and each taxon has its own trendline as
determined by logistic regression.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#fig3_sub_family=truepos.krak.family%>%}
\CommentTok{#  select(mix.ID, family, spp.rich, pollen.grain.proportion, qual.family)}

\CommentTok{#fig3_sub_family$spp.rich=as.factor(fig3_sub_family$spp.rich)}

\NormalTok{fig3_sub_genus=truepos.krak.genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, genus, spp.rich, pollen.grain.proportion, qual.genus)}

\NormalTok{fig3_sub_genus}\OperatorTok{$}\NormalTok{spp.rich=}\KeywordTok{as.factor}\NormalTok{(fig3_sub_genus}\OperatorTok{$}\NormalTok{spp.rich)}

\NormalTok{fig3_sub_species=truepos.krak.species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{select}\NormalTok{(mix.ID, species, spp.rich, pollen.grain.proportion, qual.species)}

\NormalTok{fig3_sub_genus}\OperatorTok{$}\NormalTok{spp.rich=}\KeywordTok{as.factor}\NormalTok{(fig3_sub_genus}\OperatorTok{$}\NormalTok{spp.rich)}

\CommentTok{#Figure 3a: Family level}
\CommentTok{#fig3a=fig3_sub_family%>%}
\CommentTok{#  group_by(family)%>%}
\CommentTok{#  ggplot(aes(pollen.grain.proportion, qual.family, color = family, shape = family))+}
\CommentTok{#  geom_jitter( size=4, alpha=0.4, height=0.05 )+}
\CommentTok{#    geom_smooth(method = "glm", method.args = list(family = "binomial"), se = FALSE, alpha=0.2) +}
\CommentTok{#  ggtitle("A")+}
\CommentTok{#  xlab("proportion of pollen grains in the sample") +}
\CommentTok{#  ylab("probability of detection") +}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw() +}
\CommentTok{#  theme(legend.key = element_rect(size = 5),}
\CommentTok{#        legend.key.size = unit(1.5, 'lines'))}

\CommentTok{#Figure 3b: Genus level}
\NormalTok{fig3b=fig3_sub_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(genus)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(pollen.grain.proportion, qual.genus, }\DataTypeTok{color =}\NormalTok{ genus, }\DataTypeTok{shape =}\NormalTok{ genus))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_jitter}\NormalTok{( }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.4}\NormalTok{, }\DataTypeTok{height=}\FloatTok{0.05}\NormalTok{ )}\OperatorTok{+}
\StringTok{    }\KeywordTok{geom_smooth}\NormalTok{(}\DataTypeTok{method =} \StringTok{"glm"}\NormalTok{, }\DataTypeTok{method.args =} \KeywordTok{list}\NormalTok{(}\DataTypeTok{family =} \StringTok{"binomial"}\NormalTok{), }\DataTypeTok{se =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.2}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"B"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"proportion of pollen grains in the sample"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"probability of detection"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{() }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{legend.key =} \KeywordTok{element_rect}\NormalTok{(}\DataTypeTok{size =} \DecValTok{5}\NormalTok{),}
        \DataTypeTok{legend.key.size =} \KeywordTok{unit}\NormalTok{(}\FloatTok{1.5}\NormalTok{, }\StringTok{'lines'}\NormalTok{))}

\CommentTok{#Figure 3c: Species level}
\NormalTok{fig3c=fig3_sub_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{group_by}\NormalTok{(species)}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{(}\KeywordTok{aes}\NormalTok{(pollen.grain.proportion, qual.species, }\DataTypeTok{color =}\NormalTok{ species, }\DataTypeTok{shape =}\NormalTok{ species))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_jitter}\NormalTok{( }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.4}\NormalTok{, }\DataTypeTok{height=}\FloatTok{0.05}\NormalTok{ )}\OperatorTok{+}
\StringTok{    }\KeywordTok{geom_smooth}\NormalTok{(}\DataTypeTok{method =} \StringTok{"glm"}\NormalTok{, }\DataTypeTok{method.args =} \KeywordTok{list}\NormalTok{(}\DataTypeTok{family =} \StringTok{"binomial"}\NormalTok{), }\DataTypeTok{se =} \OtherTok{FALSE}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.2}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ggtitle}\NormalTok{(}\StringTok{"C"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"proportion of pollen grains  in the sample"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"probability of detection"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{() }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{legend.key =} \KeywordTok{element_rect}\NormalTok{(}\DataTypeTok{size =} \DecValTok{5}\NormalTok{),}
        \DataTypeTok{legend.key.size =} \KeywordTok{unit}\NormalTok{(}\FloatTok{1.5}\NormalTok{, }\StringTok{'lines'}\NormalTok{))}

\CommentTok{#Create panel for figure 3}
\NormalTok{fig3=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig3a,}
\NormalTok{  fig3b,fig3c, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `geom_smooth()` using formula 'y ~ x'
\end{verbatim}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 7. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{verbatim}
## `geom_smooth()` using formula 'y ~ x'
\end{verbatim}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 8. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#suppressMessages(ggsave("WGS_fig3a.pdf", plot=fig3a, device="pdf", width=7, height=5, units="in"))}
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig3b.pdf"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig3b, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{7}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 7. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig3c.pdf"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig3c, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{7}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 8. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{panel.fig3 =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig3a, }
\NormalTok{  fig3b, fig3c, }\DataTypeTok{ncol=}\DecValTok{1}\NormalTok{) }\CommentTok{#  heights = rep(50,3)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `geom_smooth()` using formula 'y ~ x'
\end{verbatim}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 7. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{verbatim}
## `geom_smooth()` using formula 'y ~ x'
\end{verbatim}

\begin{verbatim}
## Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred
\end{verbatim}

\begin{verbatim}
## Warning: The shape palette can deal with a maximum of 6 discrete values because
## more than 6 becomes difficult to discriminate; you have 8. Consider
## specifying shapes manually if you must have them.
\end{verbatim}

\begin{verbatim}
## Warning: Removed 5 rows containing missing values (geom_point).
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig3_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{panel.fig3, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{height=}\FloatTok{8.5}\NormalTok{, }\DataTypeTok{width =} \DecValTok{11}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{fig3_combined.jpg}
\caption{Figure 3}
\end{figure}

\textbf{AGAIN, DO NOT USE THIS FIGURE}

\hypertarget{figure-4-quantitative-matching-by-taxon}{%
\subsection{Figure 4: quantitative matching by
taxon}\label{figure-4-quantitative-matching-by-taxon}}

This figure shows the quantitative relationship between the proportion
of pollen grains and the proportion of reads matching to a particular
taxa. There are plots for the species, genus, and family level.

Fortunately, these figures do not require formatting of the dataset
beyond what has already been done. The code is fairly self-explanatory.

While it may seem non-sensical that all or nearly all of the points are
falling out below the ``true'' line with slope = 1 and intercept = 0,
that is because there are many false positive reads.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Figure 4A: Family level}
\CommentTok{#fig4a=truepos.krak.family%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(pollen.grain.proportion, quant.family, color=family, shape=family), size=4, alpha=0.4, position = "jitter")+}
\CommentTok{#  xlab("proportion of pollen grains in sample")+}
\CommentTok{#  ylab("proportion of reads")+}
\CommentTok{#  geom_abline(size=0.2, alpha=1) + #line w/ slope 1}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  scale_shape_manual(values=c(15,16,17,18,0,1,2,5,6)) + }
\CommentTok{#  theme_bw()}

\CommentTok{#Figure 4B: Genus level}
\NormalTok{fig4b=truepos.krak.genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(pollen.grain.proportion, quant.genus, }\DataTypeTok{color=}\NormalTok{genus, }\DataTypeTok{shape=}\NormalTok{genus), }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.4}\NormalTok{, }\DataTypeTok{position =} \StringTok{"jitter"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"proportion of pollen grains in sample"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of reads"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_abline}\NormalTok{(}\DataTypeTok{size=}\FloatTok{0.2}\NormalTok{, }\DataTypeTok{alpha=}\DecValTok{1}\NormalTok{) }\OperatorTok{+}\StringTok{ }\CommentTok{#line w/ slope 1}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_shape_manual}\NormalTok{(}\DataTypeTok{values=}\KeywordTok{c}\NormalTok{(}\DecValTok{15}\NormalTok{,}\DecValTok{16}\NormalTok{,}\DecValTok{17}\NormalTok{,}\DecValTok{18}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure 4C: Species level}
\NormalTok{fig4c=truepos.krak.species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{() }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(pollen.grain.proportion, quant.species, }\DataTypeTok{color=}\NormalTok{species, }\DataTypeTok{shape=}\NormalTok{species), }\DataTypeTok{size=}\DecValTok{4}\NormalTok{, }\DataTypeTok{alpha=}\FloatTok{0.4}\NormalTok{, }\DataTypeTok{position =} \StringTok{"jitter"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"proportion of pollen grains in sample"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of reads"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_abline}\NormalTok{(}\DataTypeTok{size=}\FloatTok{0.2}\NormalTok{, }\DataTypeTok{alpha=}\DecValTok{1}\NormalTok{) }\OperatorTok{+}\StringTok{ }\CommentTok{#line w/ slope 1}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_shape_manual}\NormalTok{(}\DataTypeTok{values=}\KeywordTok{c}\NormalTok{(}\DecValTok{15}\NormalTok{,}\DecValTok{16}\NormalTok{,}\DecValTok{17}\NormalTok{,}\DecValTok{18}\NormalTok{,}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{,}\DecValTok{5}\NormalTok{,}\DecValTok{6}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure 4 Panel}
\CommentTok{#ggsave("WGS_fig4a.pdf", fig4a, device="pdf", width=7, height=5, units="in")}
\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig4b.pdf"}\NormalTok{, fig4b, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{7}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{)}
\KeywordTok{ggsave}\NormalTok{(}\StringTok{"WGS_fig4c.pdf"}\NormalTok{, fig4c, }\DataTypeTok{device=}\StringTok{"pdf"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{7}\NormalTok{, }\DataTypeTok{height=}\DecValTok{5}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{)}

\CommentTok{# save panel}
\NormalTok{panel.fig4 =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig4a, }
\NormalTok{  fig4b, fig4c,}\DataTypeTok{ncol=}\DecValTok{1}\NormalTok{) }\CommentTok{#  heights = rep(50,3)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig4_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{panel.fig4, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{height=}\DecValTok{11}\NormalTok{, }\DataTypeTok{width =} \DecValTok{11}\NormalTok{, }\DataTypeTok{units=}\StringTok{"in"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{fig4_combined.jpg}
\caption{Figure 4}
\end{figure}

\#\#figure 5: true positives by source

This figure shows the proportion of correct taxonomic matchess by source
(amplicon vs.~WGS) at the species, genus, and family level. To correctly
format the data, the mean and binomial CI of correct taxonomic matches
needs to be summarized 3 times: once for sample, once for mix, and once
for source. Otherwise samples with multiple species will get
over-represented in terms of the overall means. The integers are tracked
at each step and summarized with \texttt{dplyr}. The data is summarized
with \texttt{dplyr} twice, once per mix and then once per sample. Then
within each level of the factor of interest, the mean and binomial
confidence intervals are calculated using \texttt{binom.confint}.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Summarize the agg dataset first by source, mix.ID, and sample. Also calculate pool size for each sample ('n()'), as well as sum # of positive identifications}

\CommentTok{#CI.krakamp.family.fn.1 = krakamp_family_fn %>% }
\CommentTok{#select(mix.ID, sample, qual.family, source) %>% group_by(source, mix.ID, sample) %>%}
\CommentTok{#summarize(pool.size = n(),}
\CommentTok{#  family = sum(as.numeric(qual.family)))}

\NormalTok{CI.krakamp.genus.fn}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{krakamp_genus_fn }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, sample, qual.genus, source) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID, sample) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(qual.genus)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.krakamp.species.fn}\FloatTok{.1}\NormalTok{ =}\StringTok{ }\NormalTok{krakamp_species_fn }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, sample, qual.species, source) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID, sample) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.size =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(qual.species)))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# second step: group_by(source, mix.ID, pool.size)}

\CommentTok{#CI.krakamp.family.fn.2 = CI.krakamp.family.fn.1 %>% group_by(source, mix.ID, pool.size) %>% }
\CommentTok{#  summarize(pool.num = n(),}
\CommentTok{#  family = sum(family))}
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\CommentTok{#  CI.krakamp.family.fn.2$n2 = CI.krakamp.family.fn.2$pool.size * CI.krakamp.family.fn.2$pool.num}
  \CommentTok{#format source to be factor}
\CommentTok{#  CI.krakamp.family.fn.2$source=as.factor(CI.krakamp.family.fn.2$source)}
  
\CommentTok{#  fig5_family=CI.krakamp.family.fn.2%>%}
\CommentTok{#    select(source, mix.ID, n2, family)}
  
\NormalTok{CI.krakamp.genus.fn}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.krakamp.genus.fn}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{genus =} \KeywordTok{sum}\NormalTok{(genus))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{  CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}
  \CommentTok{#format source to be factor}
\NormalTok{  CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{source=}\KeywordTok{as.factor}\NormalTok{(CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{source)}
  
\NormalTok{  fig5_genus=CI.krakamp.genus.fn}\FloatTok{.2}\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(source, mix.ID, n2, genus)}

\NormalTok{CI.krakamp.species.fn}\FloatTok{.2}\NormalTok{ =}\StringTok{ }\NormalTok{CI.krakamp.species.fn}\FloatTok{.1} \OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID, pool.size) }\OperatorTok{%>%}\StringTok{ }
\StringTok{  }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{pool.num =} \KeywordTok{n}\NormalTok{(),}
  \DataTypeTok{species =} \KeywordTok{sum}\NormalTok{(species))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source', 'mix.ID' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{# multiply pool size & number to get total number of possibilities of matches}
\NormalTok{  CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{n2 =}\StringTok{ }\NormalTok{CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.size }\OperatorTok{*}\StringTok{ }\NormalTok{CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{pool.num}
  \CommentTok{#format source to be factor}
\NormalTok{  CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{source=}\KeywordTok{as.factor}\NormalTok{(CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{$}\NormalTok{source)}
  
\NormalTok{  fig5_species=CI.krakamp.species.fn}\FloatTok{.2}\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(source, mix.ID, n2, species)}
  
  \CommentTok{#Convert from wide to long}
 \CommentTok{# fig5_long_family = melt(fig5_family, id.vars = c("source","mix.ID", "n2"))}
\CommentTok{#  fig5_long_family = fig5_long_family%>%}
\CommentTok{#  rename(x=value, n=n2)}
  
\NormalTok{  fig5_long_genus =}\StringTok{ }\KeywordTok{melt}\NormalTok{(fig5_genus, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"source"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(fig5_genus, id.vars = c("source", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(fig5_genus). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{  fig5_long_genus =}\StringTok{ }\NormalTok{fig5_long_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}
  
\NormalTok{  fig5_long_species =}\StringTok{ }\KeywordTok{melt}\NormalTok{(fig5_species, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"source"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"n2"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(fig5_species, id.vars = c("source", "mix.ID", "n2")): The melt
## generic in data.table has been passed a grouped_df and will attempt to redirect
## to the relevant reshape2 method; please note that reshape2 is deprecated, and
## this redirection is now deprecated as well. To continue using melt methods from
## reshape2 while both libraries are attached, e.g. melt.list, you can prepend the
## namespace like reshape2::melt(fig5_species). In the next version, this warning
## will become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{  fig5_long_species =}\StringTok{ }\NormalTok{fig5_long_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{n2)}
  
  \CommentTok{#group by source}
\CommentTok{#  fig5_long_family = fig5_long_family %>%}
\CommentTok{#    group_by(source) %>%}
\CommentTok{#    summarize(n=mean(n),x=mean(x)) %>%}
\CommentTok{#    ungroup()}
  
\NormalTok{  fig5_long_genus =}\StringTok{ }\NormalTok{fig5_long_genus }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{group_by}\NormalTok{(source) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{  fig5_long_species =}\StringTok{ }\NormalTok{fig5_long_species }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{group_by}\NormalTok{(source) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
  \CommentTok{#Calculate mean and binomial CI}

 \CommentTok{# fig5_bin.family = binom.confint(fig5_long_family$x, fig5_long_family$n,methods="exact") %>%}
\CommentTok{#    select(-method)}
  
\NormalTok{  fig5_bin.genus =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(fig5_long_genus}\OperatorTok{$}\NormalTok{x, fig5_long_genus}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}
  
\NormalTok{  fig5_bin.species =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(fig5_long_species}\OperatorTok{$}\NormalTok{x, fig5_long_species}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}
  
  \CommentTok{#merge datasets}
  
\CommentTok{#  fig5_family.all = merge(fig5_bin.family, fig5_long_family, by = c("x", "n"))}
\CommentTok{#  fig5_family.all = unique(fig5_family.all)}
  
\NormalTok{  fig5_genus.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(fig5_bin.genus, fig5_long_genus, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{  fig5_genus.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(fig5_genus.all)}
  
\NormalTok{  fig5_species.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(fig5_bin.species, fig5_long_species, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{  fig5_species.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(fig5_species.all)}
  
\CommentTok{#Figure 5A: Family level}
\CommentTok{#fig5a=fig5_family.all%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(source,mean))+}
\CommentTok{#  geom_errorbar(aes(source,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("source")+}
\CommentTok{#  ylab("proportion of correct matches")+}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw()}

\CommentTok{#Figure 5B: Genus level}
\NormalTok{fig5b=fig5_genus.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"source"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure 5C: Species level}
\NormalTok{fig5c=fig5_species.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"source"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of correct matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}


\CommentTok{#Create panel for Figure 5}
\NormalTok{fig5=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig5a,}
\NormalTok{  fig5b,fig5c, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig5_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig5, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{fig5_combined.jpg}
\caption{Figure 5}
\end{figure}

\hypertarget{figure-6-false-negatives-by-source}{%
\subsection{Figure 6: false negatives by
source}\label{figure-6-false-negatives-by-source}}

This figure shows the proportion of hits that were false positives. The
``false positive'' vs.~``false negative'' variables created previously
are converted to proportion of correct matches. The mean and binomial CI
of false positive counts are aggregated once by sample, once by mix, and
once by source. Data summary and confidence intervals are calculated as
in figure 5.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Create variable called "falsepos" that is equivalent to "type"}

\CommentTok{#all.krak.family$falsepos = all.krak.family$type}
\NormalTok{all.krak.genus}\OperatorTok{$}\NormalTok{falsepos =}\StringTok{ }\NormalTok{all.krak.genus}\OperatorTok{$}\NormalTok{type}
\NormalTok{all.krak.species}\OperatorTok{$}\NormalTok{falsepos =}\StringTok{ }\NormalTok{all.krak.species}\OperatorTok{$}\NormalTok{type}

\CommentTok{#Merge ITS and rbcL data}
\CommentTok{#amp_family_allmix = merge(amp_its_family_mix, amp_rbc_family_mix, by=c("mix.ID","taxa","sample"), all.x=T, all.y=T)}
\NormalTok{amp_genus_allmix =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_genus_mix, amp_rbc_genus_mix, }\DataTypeTok{by=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"taxa"}\NormalTok{,}\StringTok{"sample"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T, }\DataTypeTok{all.y=}\NormalTok{T)}
\NormalTok{amp_species_allmix =}\StringTok{ }\KeywordTok{merge}\NormalTok{(amp_its_species_mix, amp_rbc_species_mix, }\DataTypeTok{by=}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"taxa"}\NormalTok{,}\StringTok{"sample"}\NormalTok{), }\DataTypeTok{all.x=}\NormalTok{T, }\DataTypeTok{all.y=}\NormalTok{T)}

\CommentTok{#Create hits variable that sums the hits for rbcL and ITS}
\CommentTok{#amp_family_allmix$hits = rowSums(amp_family_allmix[,c("hits.x","hits.y")], na.rm=T)}
\NormalTok{amp_genus_allmix}\OperatorTok{$}\NormalTok{hits =}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(amp_genus_allmix[,}\KeywordTok{c}\NormalTok{(}\StringTok{"hits.x"}\NormalTok{,}\StringTok{"hits.y"}\NormalTok{)], }\DataTypeTok{na.rm=}\NormalTok{T)}
\NormalTok{amp_species_allmix}\OperatorTok{$}\NormalTok{hits =}\StringTok{ }\KeywordTok{rowSums}\NormalTok{(amp_species_allmix[,}\KeywordTok{c}\NormalTok{(}\StringTok{"hits.x"}\NormalTok{,}\StringTok{"hits.y"}\NormalTok{)], }\DataTypeTok{na.rm=}\NormalTok{T)}

\CommentTok{#Create false positive variable}
\CommentTok{#amp_family_allmix$falsepos = ifelse(!is.na(amp_family_allmix$type.x),amp_family_allmix$type.x,amp_family_allmix$type.y)}
\NormalTok{amp_genus_allmix}\OperatorTok{$}\NormalTok{falsepos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\OperatorTok{!}\KeywordTok{is.na}\NormalTok{(amp_genus_allmix}\OperatorTok{$}\NormalTok{type.x),amp_genus_allmix}\OperatorTok{$}\NormalTok{type.x,amp_genus_allmix}\OperatorTok{$}\NormalTok{type.y)}
\NormalTok{amp_species_allmix}\OperatorTok{$}\NormalTok{falsepos =}\StringTok{ }\KeywordTok{ifelse}\NormalTok{(}\OperatorTok{!}\KeywordTok{is.na}\NormalTok{(amp_species_allmix}\OperatorTok{$}\NormalTok{type.x),amp_species_allmix}\OperatorTok{$}\NormalTok{type.x,amp_species_allmix}\OperatorTok{$}\NormalTok{type.y)}

\CommentTok{#Rename variable "taxa"}
\CommentTok{#setnames(amp_family_allmix, "taxa", "family")}
\KeywordTok{setnames}\NormalTok{(amp_genus_allmix, }\StringTok{"taxa"}\NormalTok{,}\StringTok{"genus"}\NormalTok{)}
\KeywordTok{setnames}\NormalTok{(amp_species_allmix, }\StringTok{"taxa"}\NormalTok{, }\StringTok{"species"}\NormalTok{)}

\CommentTok{#all.krak.family$source = "K"}
\CommentTok{#amp_family_allmix$source = "A"}
\NormalTok{all.krak.genus}\OperatorTok{$}\NormalTok{source =}\StringTok{ "K"}
\NormalTok{amp_genus_allmix}\OperatorTok{$}\NormalTok{source =}\StringTok{ "A"}
\NormalTok{all.krak.species}\OperatorTok{$}\NormalTok{source =}\StringTok{ "K"}
\NormalTok{amp_species_allmix}\OperatorTok{$}\NormalTok{source =}\StringTok{ "A"}

\CommentTok{#Merge amplicon and kraken data }
\CommentTok{#krakamp_family_fp <- rbind(all.krak.family[,c("mix.ID", "family", "sample", "falsepos", "source", "hits")], amp_family_allmix[,c("mix.ID","family","sample", "falsepos", "source", "hits")])}
\NormalTok{krakamp_genus_fp <-}\StringTok{ }\KeywordTok{rbind}\NormalTok{(all.krak.genus[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"genus"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"falsepos"}\NormalTok{, }\StringTok{"source"}\NormalTok{, }\StringTok{"hits"}\NormalTok{)], amp_genus_allmix[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"genus"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"falsepos"}\NormalTok{, }\StringTok{"source"}\NormalTok{, }\StringTok{"hits"}\NormalTok{)])}
\NormalTok{krakamp_species_fp <-}\StringTok{ }\KeywordTok{rbind}\NormalTok{(all.krak.species[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"species"}\NormalTok{, }\StringTok{"sample"}\NormalTok{, }\StringTok{"falsepos"}\NormalTok{, }\StringTok{"source"}\NormalTok{, }\StringTok{"hits"}\NormalTok{)], amp_species_allmix[,}\KeywordTok{c}\NormalTok{(}\StringTok{"mix.ID"}\NormalTok{,}\StringTok{"species"}\NormalTok{,}\StringTok{"sample"}\NormalTok{, }\StringTok{"falsepos"}\NormalTok{, }\StringTok{"source"}\NormalTok{, }\StringTok{"hits"}\NormalTok{)])}

\CommentTok{#Summarize the agg dataset first by source and mix ID}

\CommentTok{#CI.krakamp.family.fp = krakamp_family_fp %>% }
\CommentTok{#select(mix.ID, falsepos, source, hits) %>% group_by(source, mix.ID) %>%}
\CommentTok{#summarize(total.hits = sum(hits),}
\CommentTok{#          fp.hits = sum(hits[falsepos=="false_pos"]))}

\NormalTok{CI.krakamp.genus.fp =}\StringTok{ }\NormalTok{krakamp_genus_fp }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, falsepos, source, hits) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(hits),}
          \DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(hits[falsepos}\OperatorTok{==}\StringTok{"false_pos"}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{CI.krakamp.species.fp =}\StringTok{ }\NormalTok{krakamp_species_fp }\OperatorTok{%>%}\StringTok{ }
\KeywordTok{select}\NormalTok{(mix.ID, falsepos, source, hits) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(source, mix.ID) }\OperatorTok{%>%}
\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(hits),}
          \DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(hits[falsepos}\OperatorTok{==}\StringTok{"false_pos"}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` regrouping output by 'source' (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{#Convert from wide to long}
\CommentTok{#fig6_long_family = melt(CI.krakamp.family.fp, id.vars = c("source","mix.ID", "total.hits"))}
\CommentTok{#  fig6_long_family = fig6_long_family%>%}
\CommentTok{#  rename(x=value, n=total.hits)}
  
\NormalTok{fig6_long_genus =}\StringTok{ }\KeywordTok{melt}\NormalTok{(CI.krakamp.genus.fp, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"source"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"total.hits"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(CI.krakamp.genus.fp, id.vars = c("source", "mix.ID",
## "total.hits")): The melt generic in data.table has been passed a grouped_df
## and will attempt to redirect to the relevant reshape2 method; please note
## that reshape2 is deprecated, and this redirection is now deprecated as
## well. To continue using melt methods from reshape2 while both libraries
## are attached, e.g. melt.list, you can prepend the namespace like
## reshape2::melt(CI.krakamp.genus.fp). In the next version, this warning will
## become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{  fig6_long_genus =}\StringTok{ }\NormalTok{fig6_long_genus}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{total.hits)}
  
\NormalTok{fig6_long_species =}\StringTok{ }\KeywordTok{melt}\NormalTok{(CI.krakamp.species.fp, }\DataTypeTok{id.vars =} \KeywordTok{c}\NormalTok{(}\StringTok{"source"}\NormalTok{,}\StringTok{"mix.ID"}\NormalTok{, }\StringTok{"total.hits"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Warning in melt(CI.krakamp.species.fp, id.vars = c("source", "mix.ID",
## "total.hits")): The melt generic in data.table has been passed a grouped_df
## and will attempt to redirect to the relevant reshape2 method; please note
## that reshape2 is deprecated, and this redirection is now deprecated as
## well. To continue using melt methods from reshape2 while both libraries
## are attached, e.g. melt.list, you can prepend the namespace like
## reshape2::melt(CI.krakamp.species.fp). In the next version, this warning will
## become an error.
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{  fig6_long_species =}\StringTok{ }\NormalTok{fig6_long_species}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{rename}\NormalTok{(}\DataTypeTok{x=}\NormalTok{value, }\DataTypeTok{n=}\NormalTok{total.hits)}

\CommentTok{#group by source}
\CommentTok{# fig6_long_family = fig6_long_family %>%}
\CommentTok{#    group_by(source) %>%}
\CommentTok{#    summarize(n=mean(n),x=mean(x)) %>%}
\CommentTok{#    ungroup()}
  
\NormalTok{fig6_long_genus =}\StringTok{ }\NormalTok{fig6_long_genus }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{group_by}\NormalTok{(source) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fig6_long_species =}\StringTok{ }\NormalTok{fig6_long_species }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{group_by}\NormalTok{(source) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{n=}\KeywordTok{mean}\NormalTok{(n),}\DataTypeTok{x=}\KeywordTok{mean}\NormalTok{(x)) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{ungroup}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
 \CommentTok{#Calculate mean and binomial CI}

\CommentTok{#fig6_bin.family = binom.confint(fig6_long_family$x, fig6_long_family$n,methods="exact") %>%}
\CommentTok{#    select(-method)}
  
\NormalTok{fig6_bin.genus =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(fig6_long_genus}\OperatorTok{$}\NormalTok{x, fig6_long_genus}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}
  
\NormalTok{fig6_bin.species =}\StringTok{ }\KeywordTok{binom.confint}\NormalTok{(fig6_long_species}\OperatorTok{$}\NormalTok{x, fig6_long_species}\OperatorTok{$}\NormalTok{n,}\DataTypeTok{methods=}\StringTok{"exact"}\NormalTok{) }\OperatorTok{%>%}
\StringTok{    }\KeywordTok{select}\NormalTok{(}\OperatorTok{-}\NormalTok{method)}
  
  \CommentTok{#merge datasets}
  
\CommentTok{#  fig6_family.all = merge(fig6_bin.family, fig6_long_family, by = c("x", "n"))}
\CommentTok{#  fig6_family.all = unique(fig6_family.all)}
  
\NormalTok{  fig6_genus.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(fig6_bin.genus, fig6_long_genus, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{  fig6_genus.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(fig6_genus.all)}
  
\NormalTok{  fig6_species.all =}\StringTok{ }\KeywordTok{merge}\NormalTok{(fig6_bin.species, fig6_long_species, }\DataTypeTok{by =} \KeywordTok{c}\NormalTok{(}\StringTok{"x"}\NormalTok{, }\StringTok{"n"}\NormalTok{))}
\NormalTok{  fig6_species.all =}\StringTok{ }\KeywordTok{unique}\NormalTok{(fig6_species.all)}
  
\CommentTok{#Figure 6A: Family level}
\CommentTok{#fig6a=fig6_family.all%>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(source,mean))+}
\CommentTok{#  geom_errorbar(aes(source,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("source")+}
\CommentTok{#  ylab("proportion of false positives")+}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw()}

\CommentTok{#Figure 6B: Genus level}
\NormalTok{fig6b=fig6_genus.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"source"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure 6C: Species level}
\NormalTok{fig6c=fig6_species.all}\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(source,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"source"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}


\CommentTok{#Create panel for Figure 6}
\NormalTok{fig6=}\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#fig6a,}
\NormalTok{  fig6b,fig6c, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"fig6_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{fig6, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\includegraphics{fig6_combined.jpg} Note that the confidence intervals
here are tiny, relative to the false negatives. That is because of the
large counts of true and false positive reads in these data, as compared
with a single count (1) per sample if a taxon was found that was truly
present in that sample, in the false negatives analysis. In binomial
data, it is very intuitive that as counts increase confidence intervals
go down.

\hypertarget{data-prep-for-false-positives-by-sample-complexity}{%
\subsection{data prep for false positives by sample
complexity}\label{data-prep-for-false-positives-by-sample-complexity}}

approach:

\begin{itemize}
\tightlist
\item
  use the data we put together for fig 6 (\texttt{CI.krakamp.family.fp})
\item
  subset to include only the kraken data (i.e.~not the amplicon data)
\item
  merge these data with the \texttt{mixes} metadata
\item
  for rarity, create new \texttt{rarity} column based on binning
  \texttt{pollen.grain.proportion}
\end{itemize}

then follow the basic approach for e.g.~Fig 2:

\begin{itemize}
\tightlist
\item
  will need to alter how the \texttt{x} and \texttt{n} are calculated
  relative to the true positive / false negative analysis
\item
  here we have basically already calculated them, but need to
  \textbf{add} up hits of false positives and false negatives to
  summarize within each level (e.g.~for a given level of species
  richness)---there are already steps to do the summarizing, just need
  to make sure it happens via addition
\item
  then run \texttt{binom.confint} (from the \texttt{binomial} package)
  to get a mean and CI for each level
\item
  then plot
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# data prep}
\CommentTok{# family}
\CommentTok{#falsepos.plot.fam = CI.krakamp.family.fp %>% filter(source == "K")}
\CommentTok{#falsepos.plot.fam = merge(falsepos.plot.fam, mixes, by = "mix.ID", all.y = F)}
\CommentTok{#falsepos.plot.fam$rarity = cut(falsepos.plot.fam$pollen.grain.proportion, breaks = c(0, 0.25, 0.5, 0.75, 1), labels = c("<25%", "25-50%", "50-75%", ">75%"))}

\CommentTok{# genus}
\NormalTok{falsepos.plot.gen =}\StringTok{ }\NormalTok{CI.krakamp.genus.fp }\OperatorTok{%>%}\StringTok{ }\KeywordTok{filter}\NormalTok{(source }\OperatorTok{==}\StringTok{ "K"}\NormalTok{)}
\NormalTok{falsepos.plot.gen =}\StringTok{ }\KeywordTok{merge}\NormalTok{(falsepos.plot.gen, mixes, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{, }\DataTypeTok{all.y =}\NormalTok{ F)}
\NormalTok{falsepos.plot.gen}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{cut}\NormalTok{(falsepos.plot.gen}\OperatorTok{$}\NormalTok{pollen.grain.proportion, }\DataTypeTok{breaks =} \KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\FloatTok{0.25}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\FloatTok{0.75}\NormalTok{, }\DecValTok{1}\NormalTok{), }\DataTypeTok{labels =} \KeywordTok{c}\NormalTok{(}\StringTok{"<25%"}\NormalTok{, }\StringTok{"25-50%"}\NormalTok{, }\StringTok{"50-75%"}\NormalTok{, }\StringTok{">75%"}\NormalTok{))}

\CommentTok{# species}
\NormalTok{falsepos.plot.spp =}\StringTok{ }\NormalTok{CI.krakamp.species.fp }\OperatorTok{%>%}\StringTok{ }\KeywordTok{filter}\NormalTok{(source }\OperatorTok{==}\StringTok{ "K"}\NormalTok{)}
\NormalTok{falsepos.plot.spp =}\StringTok{ }\KeywordTok{merge}\NormalTok{(falsepos.plot.spp, mixes, }\DataTypeTok{by =} \StringTok{"mix.ID"}\NormalTok{, }\DataTypeTok{all.y =}\NormalTok{ F)}
\NormalTok{falsepos.plot.spp}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{cut}\NormalTok{(falsepos.plot.spp}\OperatorTok{$}\NormalTok{pollen.grain.proportion, }\DataTypeTok{breaks =} \KeywordTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\FloatTok{0.25}\NormalTok{, }\FloatTok{0.5}\NormalTok{, }\FloatTok{0.75}\NormalTok{, }\DecValTok{1}\NormalTok{), }\DataTypeTok{labels =} \KeywordTok{c}\NormalTok{(}\StringTok{"<25%"}\NormalTok{, }\StringTok{"25-50%"}\NormalTok{, }\StringTok{"50-75%"}\NormalTok{, }\StringTok{">75%"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\hypertarget{figure-i-false-positives-vs.-species-richness}{%
\subsection{Figure I: false positives vs.~species
richness}\label{figure-i-false-positives-vs.-species-richness}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# summarize for each level of species richness:}
\CommentTok{# family:}
\CommentTok{#falsepos.rich.plot.fam = falsepos.plot.fam %>% group_by(spp.rich) %>% summarize(total.hits = sum(total.hits), fp.hits = sum(fp.hits))}
\CommentTok{# genus}
\NormalTok{falsepos.rich.plot.gen =}\StringTok{ }\NormalTok{falsepos.plot.gen }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# species}
\NormalTok{falsepos.rich.plot.spp =}\StringTok{ }\NormalTok{falsepos.plot.spp }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(spp.rich) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# convert species richness to factor for better plotting:}
\CommentTok{#falsepos.rich.plot.fam$spp.rich = factor(falsepos.rich.plot.fam$spp.rich)}
\NormalTok{falsepos.rich.plot.gen}\OperatorTok{$}\NormalTok{spp.rich =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.rich.plot.gen}\OperatorTok{$}\NormalTok{spp.rich)}
\NormalTok{falsepos.rich.plot.spp}\OperatorTok{$}\NormalTok{spp.rich =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.rich.plot.spp}\OperatorTok{$}\NormalTok{spp.rich)}

\CommentTok{# binomial confidence intervals (and mean)}
\CommentTok{# family:}
\CommentTok{#falsepos.rich.plot.fam = data.frame(falsepos.rich.plot.fam, binom.confint(falsepos.rich.plot.fam$fp.hits, falsepos.rich.plot.fam$total.hits, methods = "exact"))}
\CommentTok{# genus:}
\NormalTok{falsepos.rich.plot.gen =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.rich.plot.gen, }\KeywordTok{binom.confint}\NormalTok{(falsepos.rich.plot.gen}\OperatorTok{$}\NormalTok{fp.hits, falsepos.rich.plot.gen}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}
\CommentTok{# species:}
\NormalTok{falsepos.rich.plot.spp =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.rich.plot.spp, }\KeywordTok{binom.confint}\NormalTok{(falsepos.rich.plot.spp}\OperatorTok{$}\NormalTok{fp.hits, falsepos.rich.plot.spp}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}

\CommentTok{#Figure IA: Family level}
\CommentTok{#figIa = falsepos.rich.plot.fam %>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(spp.rich,mean))+}
\CommentTok{#  geom_errorbar(aes(spp.rich,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("species richness") +}
\CommentTok{#  ylab("proportion of false positives") +}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches") +}
\CommentTok{#  theme_bw()}

\CommentTok{#Figure IB: Genus level}
\NormalTok{figIb =}\StringTok{ }\NormalTok{falsepos.rich.plot.gen }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"species richness"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Figure Ic: Species level}
\NormalTok{figIc =}\StringTok{ }\NormalTok{falsepos.rich.plot.spp }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(spp.rich,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"species richness"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}

\CommentTok{#Create panel for Figure I}
\NormalTok{figI =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figIa,}
\NormalTok{  figIb,figIc, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figI_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figI, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figI_combined.jpg}
\caption{Figure I}
\end{figure}

\hypertarget{figure-j-false-positives-vs.-relatedness}{%
\subsection{Figure J: false positives
vs.~relatedness}\label{figure-j-false-positives-vs.-relatedness}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# summarize for each level of relatedness:}
\CommentTok{# family:}
\CommentTok{#falsepos.relat.plot.fam = falsepos.plot.fam %>% group_by(relatedness) %>% summarize(total.hits = sum(total.hits), fp.hits = sum(fp.hits))}
\CommentTok{# genus}
\NormalTok{falsepos.relat.plot.gen =}\StringTok{ }\NormalTok{falsepos.plot.gen }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# species}
\NormalTok{falsepos.relat.plot.spp =}\StringTok{ }\NormalTok{falsepos.plot.spp }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(relatedness) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# convert relatedness to factor for better plotting:}
\CommentTok{#falsepos.relat.plot.fam$relatedness = factor(falsepos.relat.plot.fam$relatedness)}
\NormalTok{falsepos.relat.plot.gen}\OperatorTok{$}\NormalTok{relatedness =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.relat.plot.gen}\OperatorTok{$}\NormalTok{relatedness)}
\NormalTok{falsepos.relat.plot.spp}\OperatorTok{$}\NormalTok{relatedness =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.relat.plot.spp}\OperatorTok{$}\NormalTok{relatedness)}

\CommentTok{# binomial confidence intervals (and mean)}
\CommentTok{# family:}
\CommentTok{#falsepos.relat.plot.fam = data.frame(falsepos.relat.plot.fam, binom.confint(falsepos.relat.plot.fam$fp.hits, falsepos.relat.plot.fam$total.hits, methods = "exact"))}
\CommentTok{# genus:}
\NormalTok{falsepos.relat.plot.gen =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.relat.plot.gen, }\KeywordTok{binom.confint}\NormalTok{(falsepos.relat.plot.gen}\OperatorTok{$}\NormalTok{fp.hits, falsepos.relat.plot.gen}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}
\CommentTok{# species:}
\NormalTok{falsepos.relat.plot.spp =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.relat.plot.spp, }\KeywordTok{binom.confint}\NormalTok{(falsepos.relat.plot.spp}\OperatorTok{$}\NormalTok{fp.hits, falsepos.relat.plot.spp}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}


\CommentTok{#Figure JA: Family level}
\CommentTok{#figJa = falsepos.relat.plot.fam %>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(relatedness,mean))+}
\CommentTok{#  geom_errorbar(aes(relatedness,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("relatedness") +}
\CommentTok{#  ylab("proportion of false positives") +}
\CommentTok{#  ylim(0,1)+}
  \CommentTok{# scale_x_discrete(labels=paste(unique(falsepos.relat.plot.fam$relatedness))) +}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches")+}
\CommentTok{#  scale_x_discrete(labels=c("same genus", "same family", "same order", "dif. orders")) +}
\CommentTok{#  theme_bw()}
\CommentTok{#figJa = figJa + theme(axis.text.x = element_text(angle = 45, hjust = 1))}

\CommentTok{#Figure JB: Genus level}
\NormalTok{figJb =}\StringTok{ }\NormalTok{falsepos.relat.plot.gen }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_x_discrete}\NormalTok{(}\DataTypeTok{labels=}\KeywordTok{c}\NormalTok{(}\StringTok{"same genus"}\NormalTok{, }\StringTok{"same family"}\NormalTok{, }\StringTok{"same order"}\NormalTok{, }\StringTok{"dif. orders"}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figJb =}\StringTok{ }\NormalTok{figJb }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}


\CommentTok{#Figure Jc: Species level}
\NormalTok{figJc =}\StringTok{ }\NormalTok{falsepos.relat.plot.spp }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(relatedness,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"relatedness"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{scale_x_discrete}\NormalTok{(}\DataTypeTok{labels=}\KeywordTok{c}\NormalTok{(}\StringTok{"same genus"}\NormalTok{, }\StringTok{"same family"}\NormalTok{, }\StringTok{"same order"}\NormalTok{, }\StringTok{"dif. orders"}\NormalTok{)) }\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figJc =}\StringTok{ }\NormalTok{figJc }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Create panel for Figure J}
\NormalTok{figJ =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figJa,}
\NormalTok{  figJb,figJc, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figJ_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figJ, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figJ_combined.jpg}
\caption{Figure J}
\end{figure}

Again, the confidence intervals are tiny as explained in the previous
figure. We do not remark upon these more in subsequent figures.

\hypertarget{figure-k-false-positives-vs.-rarity-binned}{%
\subsection{Figure K: false positives vs.~rarity
(binned)}\label{figure-k-false-positives-vs.-rarity-binned}}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# summarize for each level of rarity:}
\CommentTok{# family:}
\CommentTok{#falsepos.rarity.plot.fam = falsepos.plot.fam %>% group_by(rarity) %>% summarize(total.hits = sum(total.hits), fp.hits = sum(fp.hits))}
\CommentTok{# genus}
\NormalTok{falsepos.rarity.plot.gen =}\StringTok{ }\NormalTok{falsepos.plot.gen }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# species}
\NormalTok{falsepos.rarity.plot.spp =}\StringTok{ }\NormalTok{falsepos.plot.spp }\OperatorTok{%>%}\StringTok{ }\KeywordTok{group_by}\NormalTok{(rarity) }\OperatorTok{%>%}\StringTok{ }\KeywordTok{summarize}\NormalTok{(}\DataTypeTok{total.hits =} \KeywordTok{sum}\NormalTok{(total.hits), }\DataTypeTok{fp.hits =} \KeywordTok{sum}\NormalTok{(fp.hits))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## `summarise()` ungrouping output (override with `.groups` argument)
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# convert rarity to factor for better plotting:}
\CommentTok{#falsepos.rarity.plot.fam$rarity = factor(falsepos.rarity.plot.fam$rarity)}
\NormalTok{falsepos.rarity.plot.gen}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.rarity.plot.gen}\OperatorTok{$}\NormalTok{rarity)}
\NormalTok{falsepos.rarity.plot.spp}\OperatorTok{$}\NormalTok{rarity =}\StringTok{ }\KeywordTok{factor}\NormalTok{(falsepos.rarity.plot.spp}\OperatorTok{$}\NormalTok{rarity)}

\CommentTok{# binomial confidence intervals (and mean)}
\CommentTok{# family:}
\CommentTok{#falsepos.rarity.plot.fam = data.frame(falsepos.rarity.plot.fam, binom.confint(falsepos.rarity.plot.fam$fp.hits, falsepos.rarity.plot.fam$total.hits, methods = "exact"))}
\CommentTok{# genus:}
\NormalTok{falsepos.rarity.plot.gen =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.rarity.plot.gen, }\KeywordTok{binom.confint}\NormalTok{(falsepos.rarity.plot.gen}\OperatorTok{$}\NormalTok{fp.hits, falsepos.rarity.plot.gen}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}
\CommentTok{# species:}
\NormalTok{falsepos.rarity.plot.spp =}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(falsepos.rarity.plot.spp, }\KeywordTok{binom.confint}\NormalTok{(falsepos.rarity.plot.spp}\OperatorTok{$}\NormalTok{fp.hits, falsepos.rarity.plot.spp}\OperatorTok{$}\NormalTok{total.hits, }\DataTypeTok{methods =} \StringTok{"exact"}\NormalTok{))}


\CommentTok{#Figure KA: Family level}
\CommentTok{#figKa = falsepos.rarity.plot.fam %>%}
\CommentTok{#  ggplot()+}
\CommentTok{#  geom_point(aes(rarity,mean))+}
\CommentTok{#  geom_errorbar(aes(rarity,ymin=lower, ymax=upper), width=0.2)+}
\CommentTok{#  xlab("minimum rarity") +}
\CommentTok{#  ylab("proportion of false positives") +}
\CommentTok{#  ylim(0,1)+}
\CommentTok{#  labs(title = NULL, subtitle = "A: Family matches")+}
\CommentTok{#  theme_bw()}
\CommentTok{# figKa = figKa + theme(axis.text.x = element_text(angle = 45, hjust = 1))}

\CommentTok{#Figure KB: Genus level}
\NormalTok{figKb =}\StringTok{ }\NormalTok{falsepos.rarity.plot.gen }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"minimum rarity"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"B: Genus matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figKb =}\StringTok{ }\NormalTok{figKb }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Figure Kc: Species level}
\NormalTok{figKc =}\StringTok{ }\NormalTok{falsepos.rarity.plot.spp }\OperatorTok{%>%}
\StringTok{  }\KeywordTok{ggplot}\NormalTok{()}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_point}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,mean))}\OperatorTok{+}
\StringTok{  }\KeywordTok{geom_errorbar}\NormalTok{(}\KeywordTok{aes}\NormalTok{(rarity,}\DataTypeTok{ymin=}\NormalTok{lower, }\DataTypeTok{ymax=}\NormalTok{upper), }\DataTypeTok{width=}\FloatTok{0.2}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{xlab}\NormalTok{(}\StringTok{"minimum rarity"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylab}\NormalTok{(}\StringTok{"proportion of false positives"}\NormalTok{) }\OperatorTok{+}
\StringTok{  }\KeywordTok{ylim}\NormalTok{(}\DecValTok{0}\NormalTok{,}\DecValTok{1}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"C: Species matches"}\NormalTok{)}\OperatorTok{+}
\StringTok{  }\KeywordTok{theme_bw}\NormalTok{()}
\NormalTok{figKc =}\StringTok{ }\NormalTok{figKc }\OperatorTok{+}\StringTok{ }\KeywordTok{theme}\NormalTok{(}\DataTypeTok{axis.text.x =} \KeywordTok{element_text}\NormalTok{(}\DataTypeTok{angle =} \DecValTok{45}\NormalTok{, }\DataTypeTok{hjust =} \DecValTok{1}\NormalTok{))}

\CommentTok{#Create panel for Figure K}
\NormalTok{figK =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figKa,}
\NormalTok{  figKb,figKc, }\DataTypeTok{nrow=}\DecValTok{1}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figK_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figK, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figK_combined.jpg}
\caption{Figure K}
\end{figure}

\hypertarget{figure-s2-false-positives-combined}{%
\subsection{Figure S2: false positives
combined}\label{figure-s2-false-positives-combined}}

Here we are putting together a figure comprised of panels we have
already built in the sections for Figs I, J, and K. This figure has 9
panels, with a row for each of the 3 facets of sample complexity and a
column for each of the 3 levels of taxonomic matching. Very
straightforward.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# re-label panels: }
\CommentTok{#figJa = figJa + labs(title = NULL, subtitle = "D: Family matches")}
\NormalTok{figJb =}\StringTok{ }\NormalTok{figJb }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"E: Genus matches"}\NormalTok{)}
\NormalTok{figJc =}\StringTok{ }\NormalTok{figJc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"F: Species matches"}\NormalTok{)}
\CommentTok{#figKa = figKa + labs(title = NULL, subtitle = "H: Family matches")}
\NormalTok{figKb =}\StringTok{ }\NormalTok{figKb }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"I: Genus matches"}\NormalTok{)}
\NormalTok{figKc =}\StringTok{ }\NormalTok{figKc }\OperatorTok{+}\StringTok{ }\KeywordTok{labs}\NormalTok{(}\DataTypeTok{title =} \OtherTok{NULL}\NormalTok{, }\DataTypeTok{subtitle =} \StringTok{"J: Species matches"}\NormalTok{)}

\CommentTok{#Create panel for Figure S2}
\NormalTok{figS2 =}\StringTok{ }\KeywordTok{grid.arrange}\NormalTok{(}\CommentTok{#figIa, }
\NormalTok{  figIb, figIc, }\CommentTok{#figJa, }
\NormalTok{  figJb, figJc, }\CommentTok{#figKa, }
\NormalTok{  figKb, figKc, }\DataTypeTok{nrow=}\DecValTok{3}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{suppressMessages}\NormalTok{(}\KeywordTok{ggsave}\NormalTok{(}\StringTok{"figS2_combined.jpg"}\NormalTok{, }\DataTypeTok{plot=}\NormalTok{figS2, }\DataTypeTok{device=}\StringTok{"jpeg"}\NormalTok{, }\DataTypeTok{width=}\DecValTok{169}\NormalTok{, }\DataTypeTok{height =} \DecValTok{250}\NormalTok{, }\DataTypeTok{units=}\StringTok{"mm"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{figure}
\centering
\includegraphics{figS2_combined.jpg}
\caption{Figure S2}
\end{figure}

\hypertarget{session-information}{%
\section{Session Information}\label{session-information}}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{sessionInfo}\NormalTok{()}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] broom.mixed_0.2.6 gridExtra_2.3     ggplot2_3.3.2     binom_1.1-1      
##  [5] data.table_1.13.4 tibble_3.0.4      tidyr_1.1.2       knitr_1.30       
##  [9] r2glmm_0.1.2      lmerTest_3.1-3    lme4_1.1-26       Matrix_1.2-18    
## [13] reshape2_1.4.4    stringr_1.4.0     dplyr_1.0.2       kableExtra_1.3.1 
## 
## loaded via a namespace (and not attached):
##  [1] statmod_1.4.35      TMB_1.7.18          tidyselect_1.1.0   
##  [4] xfun_0.19           purrr_0.3.4         splines_4.0.3      
##  [7] lattice_0.20-41     colorspace_2.0-0    vctrs_0.3.6        
## [10] generics_0.1.0      htmltools_0.5.0     viridisLite_0.3.0  
## [13] mgcv_1.8-33         yaml_2.2.1          rlang_0.4.9        
## [16] pillar_1.4.7        nloptr_1.2.2.2      glue_1.4.2         
## [19] withr_2.3.0         lifecycle_0.2.0     plyr_1.8.6         
## [22] munsell_0.5.0       gtable_0.3.0        rvest_0.3.6        
## [25] coda_0.19-4         evaluate_0.14       labeling_0.4.2     
## [28] broom_0.7.3         Rcpp_1.0.5          backports_1.2.1    
## [31] scales_1.1.1        webshot_0.5.2       farver_2.0.3       
## [34] digest_0.6.27       stringi_1.5.3       numDeriv_2016.8-1.1
## [37] grid_4.0.3          tools_4.0.3         magrittr_2.0.1     
## [40] crayon_1.3.4        pkgconfig_2.0.3     ellipsis_0.3.1     
## [43] MASS_7.3-53         xml2_1.3.2          minqa_1.2.4        
## [46] rmarkdown_2.6       httr_1.4.2          rstudioapi_0.13    
## [49] R6_2.5.0            boot_1.3-25         nlme_3.1-149       
## [52] compiler_4.0.3
\end{verbatim}

\end{document}
